// instantsearch.js@4.90.0/cjs/widgets/powered-by/powered-by.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _object_spread = require('@swc/helpers/_/_object_spread');
var _object_spread_props = require('@swc/helpers/_/_object_spread_props');
var instantsearchUiComponents = require('instantsearch-ui-components');
var preact = require('preact');
var PoweredBy = require('../../components/PoweredBy/PoweredBy.js');
var connectPoweredBy = require('../../connectors/powered-by/connectPoweredBy.js');
var suit$1 = require('../../lib/suit.js');
var documentation = require('../../lib/utils/documentation.js');
var getContainerNode = require('../../lib/utils/getContainerNode.js');

var suit = suit$1.component('PoweredBy');
var withUsage = documentation.createDocumentationMessageGenerator({
    name: 'powered-by'
});
var renderer = function renderer(param) {
    var containerNode = param.containerNode, cssClasses = param.cssClasses;
    return function(param, isFirstRendering) {
        var url = param.url, widgetParams = param.widgetParams;
        if (isFirstRendering) {
            var _widgetParams_theme = widgetParams.theme, theme = _widgetParams_theme === void 0 ? 'light' : _widgetParams_theme;
            preact.render(/*#__PURE__*/ preact.h(PoweredBy.default, {
                cssClasses: cssClasses,
                url: url,
                theme: theme
            }), containerNode);
            return;
        }
    };
};
var poweredBy = function poweredBy(widgetParams) {
    var _ref = widgetParams || {}, container = _ref.container, tmp = _ref.cssClasses, userCssClasses = tmp === void 0 ? {} : tmp, _ref_theme = _ref.theme, theme = _ref_theme === void 0 ? 'light' : _ref_theme;
    if (!container) {
        throw new Error(withUsage('The `container` option is required.'));
    }
    var containerNode = getContainerNode.getContainerNode(container);
    var cssClasses = {
        root: instantsearchUiComponents.cx(suit(), suit({
            modifierName: theme === 'dark' ? 'dark' : 'light'
        }), userCssClasses.root),
        link: instantsearchUiComponents.cx(suit({
            descendantName: 'link'
        }), userCssClasses.link),
        logo: instantsearchUiComponents.cx(suit({
            descendantName: 'logo'
        }), userCssClasses.logo)
    };
    var specializedRenderer = renderer({
        containerNode: containerNode,
        cssClasses: cssClasses
    });
    var makeWidget = connectPoweredBy.default(specializedRenderer, function() {
        return preact.render(null, containerNode);
    });
    return _object_spread_props._(_object_spread._({}, makeWidget({
        theme: theme
    })), {
        $$widgetType: 'ais.poweredBy'
    });
};

exports.default = poweredBy;
