// @rasenganjs/mdx@1.2.0-beta.7/lib/components/renderer.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { CodeBlock } from './codeblock.js';
import { Table } from './table.js';
import { Heading } from './heading.js';
import TableOfContents from './toc.js';
import createHeading from '../utils/create-heading.js';
/**
 * Renders an MDX content component with a custom code block component.
 *
 * @param {MDXRendererProps} props - The props for the MDX renderer.
 * @param {React.ReactNode} props.children - The MDX content to render.
 * @param {string} [props.className] - An optional CSS class name to apply to the rendered section.
 * @returns {React.ReactElement} - The rendered MDX content with the custom code block component.
 */
const MDXRenderer = ({ children: MDXContent, className, config, 
// TOC Extracted from the original MDX file
toc: originalTocData, }) => {
    const { components = {}, toc: customTocFunction = undefined } = config;
    return (_jsxs("section", { className: originalTocData ? 'rasengan-wrapper-with-toc' : 'rasengan-wrapper', children: [_jsx("section", { className: 'rasengan-markdown-body ' + className, children: _jsx(MDXContent, { components: {
                        ...components,
                        code: components.code ? components.code : CodeBlock,
                        table: components.table ? components.table : Table,
                        h1: components.h1
                            ? createHeading(components.h1)
                            : Heading({ variant: 'h1' }),
                        h2: components.h2
                            ? createHeading(components.h2)
                            : Heading({ variant: 'h2' }),
                        h3: components.h3
                            ? createHeading(components.h3)
                            : Heading({ variant: 'h3' }),
                        h4: components.h4
                            ? createHeading(components.h4)
                            : Heading({ variant: 'h4' }),
                        h5: components.h5
                            ? createHeading(components.h5)
                            : Heading({ variant: 'h5' }),
                        h6: components.h6
                            ? createHeading(components.h6)
                            : Heading({ variant: 'h6' }),
                    } }) }), originalTocData &&
                (customTocFunction ? (customTocFunction(originalTocData)) : (_jsx(TableOfContents, { items: originalTocData })))] }));
};
export { MDXRenderer };
