// instantsearch.js@4.90.0/cjs/widgets/rating-menu/defaultTemplates.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var instantsearchUiComponents = require('instantsearch-ui-components');
var preact = require('preact');
var formatNumber = require('../../lib/formatNumber.js');

function ItemWrapper(param) {
    var children = param.children, count = param.count, value = param.value, url = param.url, cssClasses = param.cssClasses;
    if (count) {
        return /*#__PURE__*/ preact.h("a", {
            className: instantsearchUiComponents.cx(cssClasses.link),
            "aria-label": "".concat(value, " & up"),
            href: url
        }, children);
    }
    return /*#__PURE__*/ preact.h("div", {
        className: instantsearchUiComponents.cx(cssClasses.link),
        "aria-label": "".concat(value, " & up"),
        disabled: true
    }, children);
}
var defaultTemplates = {
    item: function item(param) {
        var count = param.count, value = param.value, url = param.url, stars = param.stars, cssClasses = param.cssClasses;
        return /*#__PURE__*/ preact.h(ItemWrapper, {
            count: count,
            value: value,
            url: url,
            cssClasses: cssClasses
        }, stars.map(function(isFull, index) {
            return /*#__PURE__*/ preact.h("svg", {
                key: index,
                className: instantsearchUiComponents.cx(cssClasses.starIcon, isFull ? cssClasses.fullStarIcon : cssClasses.emptyStarIcon),
                "aria-hidden": "true",
                width: "24",
                height: "24"
            }, /*#__PURE__*/ preact.h("use", {
                xlinkHref: isFull ? '#ais-RatingMenu-starSymbol' : '#ais-RatingMenu-starEmptySymbol'
            }));
        }), /*#__PURE__*/ preact.h("span", {
            "aria-hidden": "true",
            className: instantsearchUiComponents.cx(cssClasses.label)
        }, "& Up"), count && /*#__PURE__*/ preact.h("span", {
            className: instantsearchUiComponents.cx(cssClasses.count)
        }, formatNumber.formatNumber(count)));
    }
};

exports.default = defaultTemplates;
