// instantsearch.js@4.90.0/cjs/widgets/rating-menu/rating-menu.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _object_spread = require('@swc/helpers/_/_object_spread');
var _object_spread_props = require('@swc/helpers/_/_object_spread_props');
var instantsearchUiComponents = require('instantsearch-ui-components');
var preact = require('preact');
var RefinementList = require('../../components/RefinementList/RefinementList.js');
var connectRatingMenu = require('../../connectors/rating-menu/connectRatingMenu.js');
var suit$1 = require('../../lib/suit.js');
var defaultTemplates = require('./defaultTemplates.js');
var documentation = require('../../lib/utils/documentation.js');
var getContainerNode = require('../../lib/utils/getContainerNode.js');
var prepareTemplateProps = require('../../lib/templating/prepareTemplateProps.js');

var withUsage = documentation.createDocumentationMessageGenerator({
    name: 'rating-menu'
});
var suit = suit$1.component('RatingMenu');
var renderer = function renderer(param) {
    var containerNode = param.containerNode, cssClasses = param.cssClasses, templates = param.templates, renderState = param.renderState;
    return function(param, isFirstRendering) {
        var refine = param.refine, items = param.items, createURL = param.createURL, instantSearchInstance = param.instantSearchInstance;
        if (isFirstRendering) {
            renderState.templateProps = prepareTemplateProps.prepareTemplateProps({
                defaultTemplates: defaultTemplates.default,
                templatesConfig: instantSearchInstance.templatesConfig,
                templates: templates
            });
            return;
        }
        preact.render(/*#__PURE__*/ preact.h(RefinementList.default, {
            createURL: createURL,
            cssClasses: cssClasses,
            facetValues: items,
            templateProps: renderState.templateProps,
            toggleRefinement: refine
        }, /*#__PURE__*/ preact.h("svg", {
            style: "display:none;"
        }, /*#__PURE__*/ preact.h("symbol", {
            id: suit({
                descendantName: 'starSymbol'
            }),
            viewBox: "0 0 24 24"
        }, /*#__PURE__*/ preact.h("path", {
            d: "M12 .288l2.833 8.718h9.167l-7.417 5.389 2.833 8.718-7.416-5.388-7.417 5.388 2.833-8.718-7.416-5.389h9.167z"
        })), /*#__PURE__*/ preact.h("symbol", {
            id: suit({
                descendantName: 'starEmptySymbol'
            }),
            viewBox: "0 0 24 24"
        }, /*#__PURE__*/ preact.h("path", {
            d: "M12 6.76l1.379 4.246h4.465l-3.612 2.625 1.379 4.246-3.611-2.625-3.612 2.625 1.379-4.246-3.612-2.625h4.465l1.38-4.246zm0-6.472l-2.833 8.718h-9.167l7.416 5.389-2.833 8.718 7.417-5.388 7.416 5.388-2.833-8.718 7.417-5.389h-9.167l-2.833-8.718z"
        })))), containerNode);
    };
};
var ratingMenu = function ratingMenu(widgetParams) {
    var _ref = widgetParams || {}, container = _ref.container, attribute = _ref.attribute, _ref_max = _ref.max, max = _ref_max === void 0 ? 5 : _ref_max, tmp = _ref.cssClasses, userCssClasses = tmp === void 0 ? {} : tmp, _ref_templates = _ref.templates, templates = _ref_templates === void 0 ? {} : _ref_templates;
    if (!container) {
        throw new Error(withUsage('The `container` option is required.'));
    }
    var containerNode = getContainerNode.getContainerNode(container);
    var cssClasses = {
        root: instantsearchUiComponents.cx(suit(), userCssClasses.root),
        noRefinementRoot: instantsearchUiComponents.cx(suit({
            modifierName: 'noRefinement'
        }), userCssClasses.noRefinementRoot),
        list: instantsearchUiComponents.cx(suit({
            descendantName: 'list'
        }), userCssClasses.list),
        item: instantsearchUiComponents.cx(suit({
            descendantName: 'item'
        }), userCssClasses.item),
        selectedItem: instantsearchUiComponents.cx(suit({
            descendantName: 'item',
            modifierName: 'selected'
        }), userCssClasses.selectedItem),
        disabledItem: instantsearchUiComponents.cx(suit({
            descendantName: 'item',
            modifierName: 'disabled'
        }), userCssClasses.disabledItem),
        link: instantsearchUiComponents.cx(suit({
            descendantName: 'link'
        }), userCssClasses.link),
        starIcon: instantsearchUiComponents.cx(suit({
            descendantName: 'starIcon'
        }), userCssClasses.starIcon),
        fullStarIcon: instantsearchUiComponents.cx(suit({
            descendantName: 'starIcon',
            modifierName: 'full'
        }), userCssClasses.fullStarIcon),
        emptyStarIcon: instantsearchUiComponents.cx(suit({
            descendantName: 'starIcon',
            modifierName: 'empty'
        }), userCssClasses.emptyStarIcon),
        label: instantsearchUiComponents.cx(suit({
            descendantName: 'label'
        }), userCssClasses.label),
        count: instantsearchUiComponents.cx(suit({
            descendantName: 'count'
        }), userCssClasses.count)
    };
    var specializedRenderer = renderer({
        containerNode: containerNode,
        cssClasses: cssClasses,
        renderState: {},
        templates: templates
    });
    var makeWidget = connectRatingMenu.default(specializedRenderer, function() {
        return preact.render(null, containerNode);
    });
    return _object_spread_props._(_object_spread._({}, makeWidget({
        attribute: attribute,
        max: max
    })), {
        $$widgetType: 'ais.ratingMenu'
    });
};

exports.default = ratingMenu;
