// instantsearch.js@4.90.0/cjs/components/MenuSelect/MenuSelect.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _object_spread = require('@swc/helpers/_/_object_spread');
var _object_spread_props = require('@swc/helpers/_/_object_spread_props');
var instantsearchUiComponents = require('instantsearch-ui-components');
var preact = require('preact');
var Template = require('../Template/Template.js');
var find = require('../../lib/utils/find.js');

function MenuSelect(param) {
    var cssClasses = param.cssClasses, templateProps = param.templateProps, items = param.items, refine = param.refine;
    var _ref = find.find(items, function(item) {
        return item.isRefined;
    }) || {
        value: ''
    }, selectedValue = _ref.value;
    return /*#__PURE__*/ preact.h("div", {
        className: instantsearchUiComponents.cx(cssClasses.root, items.length === 0 && cssClasses.noRefinementRoot)
    }, /*#__PURE__*/ preact.h("select", {
        className: cssClasses.select,
        value: selectedValue,
        onChange: function onChange(event) {
            refine(event.target.value);
        }
    }, /*#__PURE__*/ preact.h(Template.default, _object_spread_props._(_object_spread._({}, templateProps), {
        templateKey: "defaultOption",
        rootTagName: "option",
        rootProps: {
            value: '',
            className: cssClasses.option
        }
    })), items.map(function(item) {
        return /*#__PURE__*/ preact.h(Template.default, _object_spread_props._(_object_spread._({}, templateProps), {
            templateKey: "item",
            rootTagName: "option",
            rootProps: {
                value: item.value,
                className: cssClasses.option
            },
            key: item.value,
            data: item
        }));
    })));
}

exports.default = MenuSelect;
