// react-instantsearch@7.26.0/dist/es/ui/Stats.js
import { _ as _$2 } from '@swc/helpers/cjs/_object_spread.cjs';
import { _ as _$1 } from '@swc/helpers/cjs/_object_spread_props.cjs';
import { _ } from '@swc/helpers/cjs/_object_without_properties.cjs';
import { cx } from 'instantsearch-ui-components';
import React from 'react';

function Stats(_0) {
    var _0_classNames = _0.classNames, classNames = _0_classNames === void 0 ? {} : _0_classNames, nbHits = _0.nbHits, processingTimeMS = _0.processingTimeMS, nbSortedHits = _0.nbSortedHits, areHitsSorted = _0.areHitsSorted, translations = _0.translations, props = _(_0, [
        "classNames",
        "nbHits",
        "processingTimeMS",
        "nbSortedHits",
        "areHitsSorted",
        "translations"
    ]);
    var translationOptions = {
        nbHits: nbHits,
        processingTimeMS: processingTimeMS,
        nbSortedHits: nbSortedHits,
        areHitsSorted: areHitsSorted
    };
    return /*#__PURE__*/ React.createElement("div", _$1(_$2({}, props), {
        className: cx('ais-Stats', classNames.root, props.className)
    }), /*#__PURE__*/ React.createElement("span", {
        className: "ais-Stats-text"
    }, translations.rootElementText(translationOptions)));
}

export { Stats };
