// instantsearch.js@4.90.0/cjs/widgets/answers/answers.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _object_spread = require('@swc/helpers/_/_object_spread');
var _object_spread_props = require('@swc/helpers/_/_object_spread_props');
var instantsearchUiComponents = require('instantsearch-ui-components');
var preact = require('preact');
var Answers = require('../../components/Answers/Answers.js');
var connectAnswers = require('../../connectors/answers/connectAnswers.js');
var suit$1 = require('../../lib/suit.js');
var defaultTemplates = require('./defaultTemplates.js');
var getContainerNode = require('../../lib/utils/getContainerNode.js');
var prepareTemplateProps = require('../../lib/templating/prepareTemplateProps.js');
var documentation = require('../../lib/utils/documentation.js');
var logger = require('../../lib/utils/logger.js');

var withUsage = documentation.createDocumentationMessageGenerator({
    name: 'answers'
});
var suit = suit$1.component('Answers');
var renderer = function renderer(param) {
    var containerNode = param.containerNode, cssClasses = param.cssClasses, renderState = param.renderState, templates = param.templates;
    return function(param, isFirstRendering) {
        var hits = param.hits, isLoading = param.isLoading, instantSearchInstance = param.instantSearchInstance;
        if (isFirstRendering) {
            renderState.templateProps = prepareTemplateProps.prepareTemplateProps({
                defaultTemplates: defaultTemplates.default,
                templatesConfig: instantSearchInstance.templatesConfig,
                templates: templates
            });
            return;
        }
        preact.render(/*#__PURE__*/ preact.h(Answers.default, {
            cssClasses: cssClasses,
            hits: hits,
            isLoading: isLoading,
            templateProps: renderState.templateProps
        }), containerNode);
    };
};
/**
 * @deprecated the answers service is no longer offered, and this widget will be removed in InstantSearch.js v5
 */ var answersWidget = function answersWidget(widgetParams) {
    var _ref = widgetParams || {}, container = _ref.container, attributesForPrediction = _ref.attributesForPrediction, queryLanguages = _ref.queryLanguages, nbHits = _ref.nbHits, searchDebounceTime = _ref.searchDebounceTime, renderDebounceTime = _ref.renderDebounceTime, escapeHTML = _ref.escapeHTML, extraParameters = _ref.extraParameters, _ref_templates = _ref.templates, templates = _ref_templates === void 0 ? {} : _ref_templates, tmp = _ref.cssClasses, userCssClasses = tmp === void 0 ? {} : tmp;
    if (!container) {
        throw new Error(withUsage('The `container` option is required.'));
    }
    var containerNode = getContainerNode.getContainerNode(container);
    var cssClasses = {
        root: instantsearchUiComponents.cx(suit(), userCssClasses.root),
        emptyRoot: instantsearchUiComponents.cx(suit({
            modifierName: 'empty'
        }), userCssClasses.emptyRoot),
        header: instantsearchUiComponents.cx(suit({
            descendantName: 'header'
        }), userCssClasses.header),
        loader: instantsearchUiComponents.cx(suit({
            descendantName: 'loader'
        }), userCssClasses.loader),
        list: instantsearchUiComponents.cx(suit({
            descendantName: 'list'
        }), userCssClasses.list),
        item: instantsearchUiComponents.cx(suit({
            descendantName: 'item'
        }), userCssClasses.item)
    };
    var specializedRenderer = renderer({
        containerNode: containerNode,
        cssClasses: cssClasses,
        templates: templates,
        renderState: {}
    });
    var makeWidget = connectAnswers.default(specializedRenderer, function() {
        return preact.render(null, containerNode);
    });
    return _object_spread_props._(_object_spread._({}, makeWidget({
        attributesForPrediction: attributesForPrediction,
        queryLanguages: queryLanguages,
        nbHits: nbHits,
        searchDebounceTime: searchDebounceTime,
        renderDebounceTime: renderDebounceTime,
        escapeHTML: escapeHTML,
        extraParameters: extraParameters
    })), {
        $$widgetType: 'ais.answers'
    });
};
var answers = logger.deprecate(answersWidget);

exports.default = answers;
