// instantsearch.js@4.90.0/cjs/widgets/hits/hits.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _extends = require('@swc/helpers/_/_extends');
var _object_destructuring_empty = require('@swc/helpers/_/_object_destructuring_empty');
var _object_spread = require('@swc/helpers/_/_object_spread');
var _object_spread_props = require('@swc/helpers/_/_object_spread_props');
var _object_without_properties = require('@swc/helpers/_/_object_without_properties');
var instantsearchUiComponents = require('instantsearch-ui-components');
var preact = require('preact');
var Template = require('../../components/Template/Template.js');
var connectHits = require('../../connectors/hits/connectHits.js');
var listener = require('../../lib/insights/listener.js');
var defaultTemplates = require('./defaultTemplates.js');
var client = require('../../lib/insights/client.js');
var documentation = require('../../lib/utils/documentation.js');
var getContainerNode = require('../../lib/utils/getContainerNode.js');
var prepareTemplateProps = require('../../lib/templating/prepareTemplateProps.js');

var withUsage = documentation.createDocumentationMessageGenerator({
    name: 'hits'
});
var Hits = instantsearchUiComponents.createHitsComponent({
    createElement: preact.h,
    Fragment: preact.Fragment
});
var renderer = function renderer(param) {
    var renderState = param.renderState, cssClasses = param.cssClasses, containerNode = param.containerNode, templates = param.templates;
    return function(param, isFirstRendering) {
        var items = param.items, results = param.results, instantSearchInstance = param.instantSearchInstance, insights = param.insights, bindEvent = param.bindEvent, sendEvent = param.sendEvent, banner = param.banner;
        if (isFirstRendering) {
            renderState.templateProps = prepareTemplateProps.prepareTemplateProps({
                defaultTemplates: defaultTemplates.default,
                templatesConfig: instantSearchInstance.templatesConfig,
                templates: templates
            });
            return;
        }
        var handleInsightsClick = listener.createInsightsEventHandler({
            insights: insights,
            sendEvent: sendEvent
        });
        var emptyComponent = function emptyComponent(_0) {
            _object_destructuring_empty._(_0); var rootProps = _extends._({}, _0);
            return /*#__PURE__*/ preact.h(Template.default, _object_spread_props._(_object_spread._({}, renderState.templateProps), {
                rootProps: rootProps,
                templateKey: "empty",
                data: results,
                rootTagName: "fragment"
            }));
        };
        // @MAJOR: Move default hit component back to the UI library
        // once flavour specificities are erased
        var itemComponent = function itemComponent(_0) {
            var hit = _0.hit, index = _0.index, rootProps = _object_without_properties._(_0, [
                "hit",
                "index"
            ]);
            return /*#__PURE__*/ preact.h(Template.default, _object_spread_props._(_object_spread._({}, renderState.templateProps), {
                templateKey: "item",
                rootTagName: "li",
                rootProps: _object_spread_props._(_object_spread._({}, rootProps), {
                    onClick: function onClick(event) {
                        handleInsightsClick(event);
                        rootProps.onClick();
                    },
                    onAuxClick: function onAuxClick(event) {
                        handleInsightsClick(event);
                        rootProps.onAuxClick();
                    }
                }),
                data: _object_spread_props._(_object_spread._({}, hit), {
                    get __hitIndex () {
                        return index;
                    }
                }),
                bindEvent: bindEvent,
                sendEvent: sendEvent
            }));
        };
        var bannerComponent = function bannerComponent(props) {
            return /*#__PURE__*/ preact.h(Template.default, _object_spread_props._(_object_spread._({}, renderState.templateProps), {
                templateKey: "banner",
                data: props,
                rootTagName: "fragment"
            }));
        };
        preact.render(/*#__PURE__*/ preact.h(Hits, {
            hits: items,
            itemComponent: itemComponent,
            sendEvent: sendEvent,
            classNames: cssClasses,
            emptyComponent: emptyComponent,
            banner: banner,
            bannerComponent: templates.banner ? bannerComponent : undefined
        }), containerNode);
    };
};
var hits = (function hits(widgetParams) {
    var _ref = widgetParams || {}, container = _ref.container, escapeHTML = _ref.escapeHTML, transformItems = _ref.transformItems, _ref_templates = _ref.templates, templates = _ref_templates === void 0 ? {} : _ref_templates, _ref_cssClasses = _ref.cssClasses, cssClasses = _ref_cssClasses === void 0 ? {} : _ref_cssClasses;
    if (!container) {
        throw new Error(withUsage('The `container` option is required.'));
    }
    var containerNode = getContainerNode.getContainerNode(container);
    var specializedRenderer = renderer({
        containerNode: containerNode,
        cssClasses: cssClasses,
        renderState: {},
        templates: templates
    });
    var makeWidget = client.default(connectHits.default)(specializedRenderer, function() {
        return preact.render(null, containerNode);
    });
    return _object_spread_props._(_object_spread._({}, makeWidget({
        escapeHTML: escapeHTML,
        transformItems: transformItems
    })), {
        $$widgetType: 'ais.hits'
    });
});

exports.default = hits;
