// instantsearch.js@4.90.0/cjs/lib/templating/renderTemplate.js
'use strict';

var _define_property = require('@swc/helpers/_/_define_property');
var _object_spread = require('@swc/helpers/_/_object_spread');
var _object_spread_props = require('@swc/helpers/_/_object_spread_props');
var _type_of = require('@swc/helpers/_/_type_of');
var hogan = require('hogan.js');
var preact = require('htm/preact');
var Highlight = require('../../helpers/components/Highlight.js');
var ReverseHighlight = require('../../helpers/components/ReverseHighlight.js');
var Snippet = require('../../helpers/components/Snippet.js');
var ReverseSnippet = require('../../helpers/components/ReverseSnippet.js');

// We add all our template helper methods to the template as lambdas. Note
// that lambdas in Mustache are supposed to accept a second argument of
// `render` to get the rendered value, not the literal `{{value}}`. But
// this is currently broken (see https://github.com/twitter/hogan.js/issues/222).
function transformHelpersToHogan() {
    var helpers = arguments.length > 0 && arguments[0] !== void 0 ? arguments[0] : {}, compileOptions = arguments.length > 1 ? arguments[1] : void 0, data = arguments.length > 2 ? arguments[2] : void 0;
    return Object.keys(helpers).reduce(function(acc, helperKey) {
        return _object_spread_props._(_object_spread._({}, acc), _define_property._({}, helperKey, function() {
            var _this = this;
            return function(text) {
                var render = function render(value) {
                    return hogan.compile(value, compileOptions).render(_this);
                };
                return helpers[helperKey].call(data, text, render);
            };
        }));
    }, {});
}
function renderTemplate(param) {
    var templates = param.templates, templateKey = param.templateKey, compileOptions = param.compileOptions, helpers = param.helpers, data = param.data, bindEvent = param.bindEvent, sendEvent = param.sendEvent;
    var template = templates[templateKey];
    if (typeof template !== 'string' && typeof template !== 'function') {
        throw new Error("Template must be 'string' or 'function', was '".concat(typeof template === "undefined" ? "undefined" : _type_of._(template), "' (key: ").concat(templateKey, ")"));
    }
    if (typeof template === 'function') {
        // @MAJOR no longer pass bindEvent when string templates are removed
        var params = bindEvent || {};
        params.html = preact.html;
        params.sendEvent = sendEvent;
        params.components = {
            Highlight: Highlight.Highlight,
            ReverseHighlight: ReverseHighlight.ReverseHighlight,
            Snippet: Snippet.Snippet,
            ReverseSnippet: ReverseSnippet.ReverseSnippet
        };
        // @MAJOR remove the `as any` when string templates are removed
        // needed because not every template receives sendEvent
        return template(data, params);
    }
    var transformedHelpers = transformHelpersToHogan(helpers, compileOptions, data);
    return hogan.compile(template, compileOptions).render(_object_spread_props._(_object_spread._({}, data), {
        helpers: transformedHelpers
    })).replace(/[ \n\r\t\f\xA0]+/g, function(spaces) {
        return spaces.replace(/(^|\xA0+)[^\xA0]+/g, '$1 ');
    }).trim();
}

exports.renderTemplate = renderTemplate;
