// instantsearch.js@4.90.0/cjs/widgets/numeric-menu/numeric-menu.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _object_spread = require('@swc/helpers/_/_object_spread');
var _object_spread_props = require('@swc/helpers/_/_object_spread_props');
var instantsearchUiComponents = require('instantsearch-ui-components');
var preact = require('preact');
var RefinementList = require('../../components/RefinementList/RefinementList.js');
var connectNumericMenu = require('../../connectors/numeric-menu/connectNumericMenu.js');
var suit$1 = require('../../lib/suit.js');
var defaultTemplates = require('./defaultTemplates.js');
var documentation = require('../../lib/utils/documentation.js');
var getContainerNode = require('../../lib/utils/getContainerNode.js');
var prepareTemplateProps = require('../../lib/templating/prepareTemplateProps.js');

var withUsage = documentation.createDocumentationMessageGenerator({
    name: 'numeric-menu'
});
var suit = suit$1.component('NumericMenu');
var renderer = function renderer(param) {
    var containerNode = param.containerNode, attribute = param.attribute, cssClasses = param.cssClasses, renderState = param.renderState, templates = param.templates;
    return function(param, isFirstRendering) {
        var createURL = param.createURL, instantSearchInstance = param.instantSearchInstance, refine = param.refine, items = param.items;
        if (isFirstRendering) {
            renderState.templateProps = prepareTemplateProps.prepareTemplateProps({
                defaultTemplates: defaultTemplates.default,
                templatesConfig: instantSearchInstance.templatesConfig,
                templates: templates
            });
            return;
        }
        preact.render(/*#__PURE__*/ preact.h(RefinementList.default, {
            createURL: createURL,
            cssClasses: cssClasses,
            facetValues: items,
            templateProps: renderState.templateProps,
            toggleRefinement: refine,
            attribute: attribute
        }), containerNode);
    };
};
var numericMenu = function numericMenu(widgetParams) {
    var _ref = widgetParams || {}, container = _ref.container, attribute = _ref.attribute, items = _ref.items, tmp = _ref.cssClasses, userCssClasses = tmp === void 0 ? {} : tmp, _ref_templates = _ref.templates, templates = _ref_templates === void 0 ? {} : _ref_templates, transformItems = _ref.transformItems;
    if (!container) {
        throw new Error(withUsage('The `container` option is required.'));
    }
    var containerNode = getContainerNode.getContainerNode(container);
    var cssClasses = {
        root: instantsearchUiComponents.cx(suit(), userCssClasses.root),
        noRefinementRoot: instantsearchUiComponents.cx(suit({
            modifierName: 'noRefinement'
        }), userCssClasses.noRefinementRoot),
        list: instantsearchUiComponents.cx(suit({
            descendantName: 'list'
        }), userCssClasses.list),
        item: instantsearchUiComponents.cx(suit({
            descendantName: 'item'
        }), userCssClasses.item),
        selectedItem: instantsearchUiComponents.cx(suit({
            descendantName: 'item',
            modifierName: 'selected'
        }), userCssClasses.selectedItem),
        label: instantsearchUiComponents.cx(suit({
            descendantName: 'label'
        }), userCssClasses.label),
        radio: instantsearchUiComponents.cx(suit({
            descendantName: 'radio'
        }), userCssClasses.radio),
        labelText: instantsearchUiComponents.cx(suit({
            descendantName: 'labelText'
        }), userCssClasses.labelText)
    };
    var specializedRenderer = renderer({
        containerNode: containerNode,
        attribute: attribute,
        cssClasses: cssClasses,
        renderState: {},
        templates: templates
    });
    var makeWidget = connectNumericMenu.default(specializedRenderer, function() {
        return preact.render(null, containerNode);
    });
    return _object_spread_props._(_object_spread._({}, makeWidget({
        attribute: attribute,
        items: items,
        transformItems: transformItems
    })), {
        $$widgetType: 'ais.numericMenu'
    });
};

exports.default = numericMenu;
