// instantsearch.js@4.90.0/cjs/widgets/search-box/search-box.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _object_spread = require('@swc/helpers/_/_object_spread');
var _object_spread_props = require('@swc/helpers/_/_object_spread_props');
var instantsearchUiComponents = require('instantsearch-ui-components');
var preact = require('preact');
var SearchBox = require('../../components/SearchBox/SearchBox.js');
var connectSearchBox = require('../../connectors/search-box/connectSearchBox.js');
var suit$1 = require('../../lib/suit.js');
var defaultTemplates = require('./defaultTemplates.js');
var documentation = require('../../lib/utils/documentation.js');
var getContainerNode = require('../../lib/utils/getContainerNode.js');

var withUsage = documentation.createDocumentationMessageGenerator({
    name: 'search-box'
});
var suit = suit$1.component('SearchBox');
var renderer = function renderer(param) {
    var containerNode = param.containerNode, cssClasses = param.cssClasses, placeholder = param.placeholder, templates = param.templates, autofocus = param.autofocus, searchAsYouType = param.searchAsYouType, ignoreCompositionEvents = param.ignoreCompositionEvents, showReset = param.showReset, showSubmit = param.showSubmit, showLoadingIndicator = param.showLoadingIndicator;
    return function(param) {
        var refine = param.refine, query = param.query, isSearchStalled = param.isSearchStalled;
        preact.render(/*#__PURE__*/ preact.h(SearchBox.default, {
            query: query,
            placeholder: placeholder,
            autofocus: autofocus,
            refine: refine,
            searchAsYouType: searchAsYouType,
            ignoreCompositionEvents: ignoreCompositionEvents,
            templates: templates,
            showSubmit: showSubmit,
            showReset: showReset,
            showLoadingIndicator: showLoadingIndicator,
            isSearchStalled: isSearchStalled,
            cssClasses: cssClasses
        }), containerNode);
    };
};
var searchBox = function searchBox(widgetParams) {
    var _ref = widgetParams || {}, container = _ref.container, _ref_placeholder = _ref.placeholder, placeholder = _ref_placeholder === void 0 ? '' : _ref_placeholder, tmp = _ref.cssClasses, userCssClasses = tmp === void 0 ? {} : tmp, _ref_autofocus = _ref.autofocus, autofocus = _ref_autofocus === void 0 ? false : _ref_autofocus, _ref_searchAsYouType = _ref.searchAsYouType, searchAsYouType = _ref_searchAsYouType === void 0 ? true : _ref_searchAsYouType, _ref_ignoreCompositionEvents = _ref.ignoreCompositionEvents, ignoreCompositionEvents = _ref_ignoreCompositionEvents === void 0 ? false : _ref_ignoreCompositionEvents, _ref_showReset = _ref.showReset, showReset = _ref_showReset === void 0 ? true : _ref_showReset, _ref_showSubmit = _ref.showSubmit, showSubmit = _ref_showSubmit === void 0 ? true : _ref_showSubmit, _ref_showLoadingIndicator = _ref.showLoadingIndicator, showLoadingIndicator = _ref_showLoadingIndicator === void 0 ? true : _ref_showLoadingIndicator, queryHook = _ref.queryHook, tmp1 = _ref.templates, userTemplates = tmp1 === void 0 ? {} : tmp1;
    if (!container) {
        throw new Error(withUsage('The `container` option is required.'));
    }
    var containerNode = getContainerNode.getContainerNode(container);
    var cssClasses = {
        root: instantsearchUiComponents.cx(suit(), userCssClasses.root),
        form: instantsearchUiComponents.cx(suit({
            descendantName: 'form'
        }), userCssClasses.form),
        input: instantsearchUiComponents.cx(suit({
            descendantName: 'input'
        }), userCssClasses.input),
        submit: instantsearchUiComponents.cx(suit({
            descendantName: 'submit'
        }), userCssClasses.submit),
        submitIcon: instantsearchUiComponents.cx(suit({
            descendantName: 'submitIcon'
        }), userCssClasses.submitIcon),
        reset: instantsearchUiComponents.cx(suit({
            descendantName: 'reset'
        }), userCssClasses.reset),
        resetIcon: instantsearchUiComponents.cx(suit({
            descendantName: 'resetIcon'
        }), userCssClasses.resetIcon),
        loadingIndicator: instantsearchUiComponents.cx(suit({
            descendantName: 'loadingIndicator'
        }), userCssClasses.loadingIndicator),
        loadingIcon: instantsearchUiComponents.cx(suit({
            descendantName: 'loadingIcon'
        }), userCssClasses.loadingIcon)
    };
    var templates = _object_spread._({}, defaultTemplates.default, userTemplates);
    var specializedRenderer = renderer({
        containerNode: containerNode,
        cssClasses: cssClasses,
        placeholder: placeholder,
        templates: templates,
        autofocus: autofocus,
        searchAsYouType: searchAsYouType,
        ignoreCompositionEvents: ignoreCompositionEvents,
        showReset: showReset,
        showSubmit: showSubmit,
        showLoadingIndicator: showLoadingIndicator
    });
    var makeWidget = connectSearchBox.default(specializedRenderer, function() {
        return preact.render(null, containerNode);
    });
    return _object_spread_props._(_object_spread._({}, makeWidget({
        queryHook: queryHook
    })), {
        $$widgetType: 'ais.searchBox'
    });
};

exports.default = searchBox;
