// instantsearch.js@4.90.0/cjs/widgets/hits-per-page/hits-per-page.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _object_spread = require('@swc/helpers/_/_object_spread');
var _object_spread_props = require('@swc/helpers/_/_object_spread_props');
var instantsearchUiComponents = require('instantsearch-ui-components');
var preact = require('preact');
var Selector = require('../../components/Selector/Selector.js');
var connectHitsPerPage = require('../../connectors/hits-per-page/connectHitsPerPage.js');
var suit$1 = require('../../lib/suit.js');
var documentation = require('../../lib/utils/documentation.js');
var getContainerNode = require('../../lib/utils/getContainerNode.js');
var find = require('../../lib/utils/find.js');

var withUsage = documentation.createDocumentationMessageGenerator({
    name: 'hits-per-page'
});
var suit = suit$1.component('HitsPerPage');
var renderer = function renderer(param) {
    var containerNode = param.containerNode, cssClasses = param.cssClasses;
    return function(param, isFirstRendering) {
        var items = param.items, refine = param.refine;
        if (isFirstRendering) return;
        var _ref = find.find(items, function(param) {
            var isRefined = param.isRefined;
            return isRefined;
        }) || {}, currentValue = _ref.value;
        preact.render(/*#__PURE__*/ preact.h("div", {
            className: cssClasses.root
        }, /*#__PURE__*/ preact.h(Selector.default, {
            cssClasses: cssClasses,
            currentValue: currentValue,
            options: items,
            // @ts-expect-error: the refine function expects a number, but setValue will call it with a string. We don't want to change the type of the refine function because it's part of the connector API.
            setValue: refine
        })), containerNode);
    };
};
var hitsPerPage = function hitsPerPage(widgetParams) {
    var _ref = widgetParams || {}, container = _ref.container, items = _ref.items, tmp = _ref.cssClasses, userCssClasses = tmp === void 0 ? {} : tmp, transformItems = _ref.transformItems;
    if (!container) {
        throw new Error(withUsage('The `container` option is required.'));
    }
    var containerNode = getContainerNode.getContainerNode(container);
    var cssClasses = {
        root: instantsearchUiComponents.cx(suit(), userCssClasses.root),
        select: instantsearchUiComponents.cx(suit({
            descendantName: 'select'
        }), userCssClasses.select),
        option: instantsearchUiComponents.cx(suit({
            descendantName: 'option'
        }), userCssClasses.option)
    };
    var specializedRenderer = renderer({
        containerNode: containerNode,
        cssClasses: cssClasses
    });
    var makeWidget = connectHitsPerPage.default(specializedRenderer, function() {
        return preact.render(null, containerNode);
    });
    return _object_spread_props._(_object_spread._({}, makeWidget({
        items: items,
        transformItems: transformItems
    })), {
        $$widgetType: 'ais.hitsPerPage'
    });
};

exports.default = hitsPerPage;
