// instantsearch.js@4.90.0/cjs/components/CurrentRefinements/CurrentRefinements.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var instantsearchUiComponents = require('instantsearch-ui-components');
var preact = require('preact');
var capitalize = require('../../lib/utils/capitalize.js');
var isSpecialClick = require('../../lib/utils/isSpecialClick.js');

var createItemKey = function createItemKey(param) {
    var attribute = param.attribute, value = param.value, type = param.type, operator = param.operator;
    return [
        attribute,
        type,
        value,
        operator
    ].map(function(key) {
        return key;
    }).filter(Boolean).join(':');
};
var handleClick = function handleClick(callback) {
    return function(event) {
        if (isSpecialClick.isSpecialClick(event)) {
            return;
        }
        event.preventDefault();
        callback();
    };
};
var CurrentRefinements = function CurrentRefinements(param) {
    var items = param.items, cssClasses = param.cssClasses, canRefine = param.canRefine;
    return /*#__PURE__*/ preact.h("div", {
        className: instantsearchUiComponents.cx(cssClasses.root, !canRefine && cssClasses.noRefinementRoot)
    }, /*#__PURE__*/ preact.h("ul", {
        className: cssClasses.list
    }, items.map(function(item, index) {
        return /*#__PURE__*/ preact.h("li", {
            key: "".concat(item.indexName, "-").concat(item.attribute, "-").concat(index),
            className: cssClasses.item
        }, /*#__PURE__*/ preact.h("span", {
            className: cssClasses.label
        }, capitalize.capitalize(item.label), ": "), item.refinements.map(function(refinement) {
            return /*#__PURE__*/ preact.h("span", {
                key: createItemKey(refinement),
                className: cssClasses.category
            }, /*#__PURE__*/ preact.h("span", {
                className: cssClasses.categoryLabel
            }, refinement.attribute === 'query' ? /*#__PURE__*/ preact.h("q", null, refinement.label) : refinement.label), /*#__PURE__*/ preact.h("button", {
                className: cssClasses.delete,
                type: "button",
                onClick: handleClick(item.refine.bind(null, refinement))
            }, "✕"));
        }));
    })));
};

exports.default = CurrentRefinements;
