// react-instantsearch-core@7.26.0/dist/cjs/lib/useIndex.js
'use strict';

var index = require('instantsearch.js/cjs/widgets/index/index');
var React = require('react');
var useForceUpdate = require('./useForceUpdate.js');
var useIndexContext = require('./useIndexContext.js');
var useInstantSearchServerContext = require('./useInstantSearchServerContext.js');
var useInstantSearchSSRContext = require('./useInstantSearchSSRContext.js');
var useIsomorphicLayoutEffect = require('./useIsomorphicLayoutEffect.js');
var useStableValue = require('./useStableValue.js');
var useWidget = require('./useWidget.js');

function useIndex(props) {
    var serverContext = useInstantSearchServerContext.useInstantSearchServerContext();
    var ssrContext = useInstantSearchSSRContext.useInstantSearchSSRContext();
    var initialResults = ssrContext === null || ssrContext === void 0 ? void 0 : ssrContext.initialResults;
    var parentIndex = useIndexContext.useIndexContext();
    var stableProps = useStableValue.useStableValue(props);
    var indexWidget = React.useMemo(function() {
        return index(stableProps);
    }, [
        stableProps
    ]);
    var helper = indexWidget.getHelper();
    var forceUpdate = useForceUpdate.useForceUpdate();
    useIsomorphicLayoutEffect.useIsomorphicLayoutEffect(function() {
        forceUpdate();
    }, [
        helper,
        forceUpdate
    ]);
    useWidget.useWidget({
        widget: indexWidget,
        parentIndex: parentIndex,
        props: stableProps,
        shouldSsr: Boolean(serverContext || initialResults),
        skipSuspense: true
    });
    return indexWidget;
}

exports.useIndex = useIndex;
