// react-instantsearch-core@7.26.0/dist/cjs/lib/useSearchState.js
'use strict';

var _sliced_to_array = require('@swc/helpers/_/_sliced_to_array');
var React = require('react');
var useIndexContext = require('./useIndexContext.js');
var useInstantSearchContext = require('./useInstantSearchContext.js');

function useSearchState() {
    var search = useInstantSearchContext.useInstantSearchContext();
    var searchIndex = useIndexContext.useIndexContext();
    var indexId = searchIndex.getIndexId();
    var _useState = _sliced_to_array._(React.useState(function() {
        return search.getUiState();
    }), 2), uiState = _useState[0], setLocalUiState = _useState[1];
    var indexUiState = uiState[indexId];
    var _useState1 = _sliced_to_array._(React.useState(function() {
        return search.renderState;
    }), 2), renderState = _useState1[0], setRenderState = _useState1[1];
    var indexRenderState = renderState[indexId] || {};
    var setUiState = React.useCallback(function(nextUiState) {
        search.setUiState(nextUiState);
    }, [
        search
    ]);
    var setIndexUiState = React.useCallback(function(nextIndexUiState) {
        searchIndex.setIndexUiState(nextIndexUiState);
    }, [
        searchIndex
    ]);
    React.useEffect(function() {
        function handleRender() {
            setLocalUiState(search.getUiState());
            setRenderState(search.renderState);
        }
        search.addListener('render', handleRender);
        // Force setting state to mitigate potential race conditions where
        // render listener is added after search results have been returned.
        // This edge case is currently not covered by the tests.
        handleRender();
        return function() {
            search.removeListener('render', handleRender);
        };
    }, [
        search
    ]);
    return {
        uiState: uiState,
        setUiState: setUiState,
        indexUiState: indexUiState,
        setIndexUiState: setIndexUiState,
        renderState: renderState,
        indexRenderState: indexRenderState
    };
}

exports.useSearchState = useSearchState;
