// instantsearch.js@4.90.0/cjs/components/RefinementList/RefinementListItem.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _object_spread = require('@swc/helpers/_/_object_spread');
var _object_spread_props = require('@swc/helpers/_/_object_spread_props');
var preact = require('preact');
var Template = require('../Template/Template.js');

function RefinementListItem(param) {
    var className = param.className, handleClick = param.handleClick, facetValueToRefine = param.facetValueToRefine, isRefined = param.isRefined, templateProps = param.templateProps, templateKey = param.templateKey, templateData = param.templateData, subItems = param.subItems;
    return /*#__PURE__*/ preact.h("li", {
        className: className,
        onClick: function onClick(originalEvent) {
            handleClick({
                facetValueToRefine: facetValueToRefine,
                isRefined: isRefined,
                originalEvent: originalEvent
            });
        }
    }, /*#__PURE__*/ preact.h(Template.default, _object_spread_props._(_object_spread._({}, templateProps), {
        templateKey: templateKey,
        data: templateData
    })), subItems);
}

exports.default = RefinementListItem;
