// instantsearch.js@4.90.0/cjs/components/ToggleRefinement/ToggleRefinement.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _object_spread = require('@swc/helpers/_/_object_spread');
var _object_spread_props = require('@swc/helpers/_/_object_spread_props');
var preact = require('preact');
var Template = require('../Template/Template.js');

var ToggleRefinement = function ToggleRefinement(param) {
    var currentRefinement = param.currentRefinement, refine = param.refine, cssClasses = param.cssClasses, templateProps = param.templateProps;
    return /*#__PURE__*/ preact.h("div", {
        className: cssClasses.root
    }, /*#__PURE__*/ preact.h("label", {
        className: cssClasses.label
    }, /*#__PURE__*/ preact.h("input", {
        className: cssClasses.checkbox,
        type: "checkbox",
        checked: currentRefinement.isRefined,
        onChange: function onChange(event) {
            return refine({
                isRefined: !event.target.checked
            });
        }
    }), /*#__PURE__*/ preact.h(Template.default, _object_spread_props._(_object_spread._({}, templateProps), {
        rootTagName: "span",
        rootProps: {
            className: cssClasses.labelText
        },
        templateKey: "labelText",
        data: currentRefinement
    }))));
};

exports.default = ToggleRefinement;
