// instantsearch-ui-components@0.20.0/dist/cjs/index.js
'use strict';

var Autocomplete = require('./components/autocomplete/Autocomplete.js');
var AutocompleteDetachedContainer = require('./components/autocomplete/AutocompleteDetachedContainer.js');
var AutocompleteDetachedFormContainer = require('./components/autocomplete/AutocompleteDetachedFormContainer.js');
var AutocompleteDetachedOverlay = require('./components/autocomplete/AutocompleteDetachedOverlay.js');
var AutocompleteDetachedSearchButton = require('./components/autocomplete/AutocompleteDetachedSearchButton.js');
var AutocompleteIndex = require('./components/autocomplete/AutocompleteIndex.js');
var AutocompletePanel = require('./components/autocomplete/AutocompletePanel.js');
var AutocompleteRecentSearch = require('./components/autocomplete/AutocompleteRecentSearch.js');
var AutocompleteSearch = require('./components/autocomplete/AutocompleteSearch.js');
var AutocompleteSuggestion = require('./components/autocomplete/AutocompleteSuggestion.js');
var createAutocompletePropGetters = require('./components/autocomplete/createAutocompletePropGetters.js');
var createAutocompleteStorage = require('./components/autocomplete/createAutocompleteStorage.js');
var Button = require('./components/Button.js');
var Carousel = require('./components/Carousel.js');
var Chat = require('./components/chat/Chat.js');
var ChatHeader = require('./components/chat/ChatHeader.js');
var ChatMessage = require('./components/chat/ChatMessage.js');
var ChatMessages = require('./components/chat/ChatMessages.js');
var ChatMessageLoader = require('./components/chat/ChatMessageLoader.js');
var ChatMessageError = require('./components/chat/ChatMessageError.js');
var ChatPrompt = require('./components/chat/ChatPrompt.js');
var ChatPromptSuggestions = require('./components/chat/ChatPromptSuggestions.js');
var ChatToggleButton = require('./components/chat/ChatToggleButton.js');
var icons = require('./components/chat/icons.js');
var FrequentlyBoughtTogether = require('./components/FrequentlyBoughtTogether.js');
var Highlight = require('./components/Highlight.js');
var Hits = require('./components/Hits.js');
var LookingSimilar = require('./components/LookingSimilar.js');
var RelatedProducts = require('./components/RelatedProducts.js');
var TrendingItems = require('./components/TrendingItems.js');
var FilterSuggestions = require('./components/FilterSuggestions.js');
var cx = require('./lib/cx.js');
var stickToBottom = require('./lib/stickToBottom.js');
var find = require('./lib/utils/find.js');
var startsWith = require('./lib/utils/startsWith.js');



exports.createAutocompleteComponent = Autocomplete.createAutocompleteComponent;
exports.createAutocompleteDetachedContainerComponent = AutocompleteDetachedContainer.createAutocompleteDetachedContainerComponent;
exports.createAutocompleteDetachedFormContainerComponent = AutocompleteDetachedFormContainer.createAutocompleteDetachedFormContainerComponent;
exports.createAutocompleteDetachedOverlayComponent = AutocompleteDetachedOverlay.createAutocompleteDetachedOverlayComponent;
exports.createAutocompleteDetachedSearchButtonComponent = AutocompleteDetachedSearchButton.createAutocompleteDetachedSearchButtonComponent;
exports.createAutocompleteIndexComponent = AutocompleteIndex.createAutocompleteIndexComponent;
exports.createAutocompletePanelComponent = AutocompletePanel.createAutocompletePanelComponent;
exports.createAutocompleteRecentSearchComponent = AutocompleteRecentSearch.createAutocompleteRecentSearchComponent;
exports.createAutocompleteSearchComponent = AutocompleteSearch.createAutocompleteSearchComponent;
exports.createAutocompleteSuggestionComponent = AutocompleteSuggestion.createAutocompleteSuggestionComponent;
exports.createAutocompletePropGetters = createAutocompletePropGetters.createAutocompletePropGetters;
exports.createAutocompleteStorage = createAutocompleteStorage.createAutocompleteStorage;
exports.createStorage = createAutocompleteStorage.createStorage;
exports.createButtonComponent = Button.createButtonComponent;
exports.createCarouselComponent = Carousel.createCarouselComponent;
exports.generateCarouselId = Carousel.generateCarouselId;
exports.createChatComponent = Chat.createChatComponent;
exports.createChatHeaderComponent = ChatHeader.createChatHeaderComponent;
exports.createChatMessageComponent = ChatMessage.createChatMessageComponent;
exports.createChatMessagesComponent = ChatMessages.createChatMessagesComponent;
exports.createChatMessageLoaderComponent = ChatMessageLoader.createChatMessageLoaderComponent;
exports.createChatMessageErrorComponent = ChatMessageError.createChatMessageErrorComponent;
exports.createChatPromptComponent = ChatPrompt.createChatPromptComponent;
exports.createChatPromptSuggestionsComponent = ChatPromptSuggestions.createChatPromptSuggestionsComponent;
exports.createChatToggleButtonComponent = ChatToggleButton.createChatToggleButtonComponent;
exports.ArrowRightIcon = icons.ArrowRightIcon;
exports.ArrowUpIcon = icons.ArrowUpIcon;
exports.ChevronDownIcon = icons.ChevronDownIcon;
exports.ChevronLeftIcon = icons.ChevronLeftIcon;
exports.ChevronRightIcon = icons.ChevronRightIcon;
exports.ChevronUpIcon = icons.ChevronUpIcon;
exports.CloseIcon = icons.CloseIcon;
exports.CopyIcon = icons.CopyIcon;
exports.LoadingSpinnerIcon = icons.LoadingSpinnerIcon;
exports.MaximizeIcon = icons.MaximizeIcon;
exports.MenuIcon = icons.MenuIcon;
exports.MinimizeIcon = icons.MinimizeIcon;
exports.ReloadIcon = icons.ReloadIcon;
exports.SparklesIcon = icons.SparklesIcon;
exports.StopIcon = icons.StopIcon;
exports.createFrequentlyBoughtTogetherComponent = FrequentlyBoughtTogether.createFrequentlyBoughtTogetherComponent;
exports.createHighlightComponent = Highlight.createHighlightComponent;
exports.createHitsComponent = Hits.createHitsComponent;
exports.createLookingSimilarComponent = LookingSimilar.createLookingSimilarComponent;
exports.createRelatedProductsComponent = RelatedProducts.createRelatedProductsComponent;
exports.createTrendingItemsComponent = TrendingItems.createTrendingItemsComponent;
exports.createFilterSuggestionsComponent = FilterSuggestions.createFilterSuggestionsComponent;
exports.cx = cx.cx;
exports.createStickToBottom = stickToBottom.createStickToBottom;
exports.find = find.find;
exports.startsWith = startsWith.startsWith;
