// instantsearch.js@4.90.0/cjs/widgets/filter-suggestions/filter-suggestions.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _object_spread = require('@swc/helpers/_/_object_spread');
var _object_spread_props = require('@swc/helpers/_/_object_spread_props');
var instantsearchUiComponents = require('instantsearch-ui-components');
var preact = require('preact');
var Template = require('../../components/Template/Template.js');
var connectFilterSuggestions = require('../../connectors/filter-suggestions/connectFilterSuggestions.js');
var documentation = require('../../lib/utils/documentation.js');
var getContainerNode = require('../../lib/utils/getContainerNode.js');
var prepareTemplateProps = require('../../lib/templating/prepareTemplateProps.js');

var withUsage = documentation.createDocumentationMessageGenerator({
    name: 'filter-suggestions'
});
var FilterSuggestions = instantsearchUiComponents.createFilterSuggestionsComponent({
    createElement: preact.h,
    Fragment: 'fragment'
});
var createRenderer = function createRenderer(param) {
    var renderState = param.renderState, cssClasses = param.cssClasses, containerNode = param.containerNode, templates = param.templates, maxSuggestions = param.maxSuggestions;
    return function(props, isFirstRendering) {
        var suggestions = props.suggestions, isLoading = props.isLoading, refine = props.refine, instantSearchInstance = props.instantSearchInstance;
        var headerTemplate = templates.header === false ? undefined : templates.header;
        if (isFirstRendering) {
            renderState.templateProps = prepareTemplateProps.prepareTemplateProps({
                defaultTemplates: {},
                templatesConfig: instantSearchInstance.templatesConfig,
                templates: {
                    header: headerTemplate,
                    item: templates.item,
                    empty: templates.empty
                }
            });
            return;
        }
        var headerComponent;
        if (templates.header === false) {
            headerComponent = false;
        } else if (headerTemplate) {
            headerComponent = function headerComponent(headerProps) {
                return /*#__PURE__*/ preact.h(Template.default, _object_spread_props._(_object_spread._({}, renderState.templateProps), {
                    templateKey: "header",
                    rootTagName: "div",
                    data: headerProps
                }));
            };
        }
        var itemComponent = templates.item ? function(itemProps) {
            return /*#__PURE__*/ preact.h(Template.default, _object_spread_props._(_object_spread._({}, renderState.templateProps), {
                templateKey: "item",
                rootTagName: "fragment",
                data: itemProps
            }));
        } : undefined;
        var emptyComponent = templates.empty ? function(emptyProps) {
            return /*#__PURE__*/ preact.h(Template.default, _object_spread_props._(_object_spread._({}, renderState.templateProps), {
                templateKey: "empty",
                rootTagName: "div",
                data: emptyProps
            }));
        } : undefined;
        var uiProps = {
            suggestions: suggestions,
            isLoading: isLoading,
            refine: refine,
            skeletonCount: maxSuggestions,
            itemComponent: itemComponent,
            headerComponent: headerComponent,
            emptyComponent: emptyComponent
        };
        preact.render(/*#__PURE__*/ preact.h(FilterSuggestions, _object_spread._({
            classNames: cssClasses
        }, uiProps)), containerNode);
    };
};
var filterSuggestions = (function filterSuggestions(widgetParams) {
    var _ref = widgetParams || {}, container = _ref.container, _ref_templates = _ref.templates, templates = _ref_templates === void 0 ? {} : _ref_templates, _ref_cssClasses = _ref.cssClasses, cssClasses = _ref_cssClasses === void 0 ? {} : _ref_cssClasses, agentId = _ref.agentId, attributes = _ref.attributes, maxSuggestions = _ref.maxSuggestions, debounceMs = _ref.debounceMs, hitsToSample = _ref.hitsToSample, transformItems = _ref.transformItems, transport = _ref.transport;
    if (!container) {
        throw new Error(withUsage('The `container` option is required.'));
    }
    var containerNode = getContainerNode.getContainerNode(container);
    var specializedRenderer = createRenderer({
        containerNode: containerNode,
        cssClasses: cssClasses,
        renderState: {},
        templates: templates,
        maxSuggestions: maxSuggestions
    });
    var makeWidget = connectFilterSuggestions.default(specializedRenderer, function() {
        return preact.render(null, containerNode);
    });
    return _object_spread_props._(_object_spread._({}, makeWidget({
        agentId: agentId,
        attributes: attributes,
        maxSuggestions: maxSuggestions,
        debounceMs: debounceMs,
        hitsToSample: hitsToSample,
        transformItems: transformItems,
        transport: transport
    })), {
        $$widgetType: 'ais.filterSuggestions'
    });
});

exports.default = filterSuggestions;
