// instantsearch-ui-components@0.20.0/dist/cjs/components/autocomplete/AutocompleteIndex.js
'use strict';

var _object_spread = require('@swc/helpers/_/_object_spread');
var _object_spread_props = require('@swc/helpers/_/_object_spread_props');
var _object_without_properties = require('@swc/helpers/_/_object_without_properties');
var cx = require('../../lib/cx.js');

function createAutocompleteIndexComponent(param) {
    var createElement = param.createElement;
    return function AutocompleteIndex(userProps) {
        var items = userProps.items, HeaderComponent = userProps.HeaderComponent, ItemComponent = userProps.ItemComponent, getItemProps = userProps.getItemProps, _userProps_classNames = userProps.classNames, classNames = _userProps_classNames === void 0 ? {} : _userProps_classNames;
        if (items.length === 0) {
            return null;
        }
        return /*#__PURE__*/ createElement("div", {
            className: cx.cx('ais-AutocompleteIndex', classNames.root)
        }, HeaderComponent && /*#__PURE__*/ createElement("div", {
            className: cx.cx('ais-AutocompleteIndexHeader', classNames.header)
        }, /*#__PURE__*/ createElement(HeaderComponent, {
            items: items
        })), /*#__PURE__*/ createElement("ol", {
            className: cx.cx('ais-AutocompleteIndexList', classNames.list)
        }, items.map(function(item, index) {
            var _getItemProps = getItemProps(item, index), className = _getItemProps.className, onSelect = _getItemProps.onSelect, onApply = _getItemProps.onApply, itemProps = _object_without_properties._(_getItemProps, [
                "className",
                "onSelect",
                "onApply"
            ]);
            return /*#__PURE__*/ createElement("li", _object_spread_props._(_object_spread._({
                key: "".concat(itemProps.id, ":").concat(item.objectID)
            }, itemProps), {
                className: cx.cx('ais-AutocompleteIndexItem', classNames.item, className)
            }), /*#__PURE__*/ createElement(ItemComponent, {
                item: item,
                onSelect: onSelect,
                onApply: onApply
            }));
        })));
    };
}

exports.createAutocompleteIndexComponent = createAutocompleteIndexComponent;
