// instantsearch-ui-components@0.20.0/dist/cjs/components/autocomplete/AutocompleteSuggestion.js
'use strict';

var icons = require('./icons.js');
var cx = require('../../lib/cx.js');

function createAutocompleteSuggestionComponent(param) {
    var createElement = param.createElement;
    return function AutocompleteSuggestion(userProps) {
        var item = userProps.item, children = userProps.children, onSelect = userProps.onSelect, onApply = userProps.onApply, _userProps_classNames = userProps.classNames, classNames = _userProps_classNames === void 0 ? {} : _userProps_classNames;
        return /*#__PURE__*/ createElement("div", {
            onClick: onSelect,
            className: cx.cx('ais-AutocompleteItemWrapper', 'ais-AutocompleteSuggestionWrapper', classNames.root)
        }, /*#__PURE__*/ createElement("div", {
            className: cx.cx('ais-AutocompleteItemContent', 'ais-AutocompleteSuggestionItemContent', classNames.content)
        }, /*#__PURE__*/ createElement("div", {
            className: cx.cx('ais-AutocompleteItemIcon', 'ais-AutocompleteSuggestionItemIcon', classNames.content)
        }, /*#__PURE__*/ createElement(icons.SubmitIcon, {
            createElement: createElement
        })), /*#__PURE__*/ createElement("div", {
            className: cx.cx('ais-AutocompleteItemContentBody', 'ais-AutocompleteSuggestionItemContentBody', classNames.content)
        }, children)), /*#__PURE__*/ createElement("div", {
            className: cx.cx('ais-AutocompleteItemActions', 'ais-AutocompleteSuggestionItemActions', classNames.actions)
        }, /*#__PURE__*/ createElement("button", {
            className: cx.cx('ais-AutocompleteItemActionButton', 'ais-AutocompleteSuggestionItemApplyButton', classNames.applyButton),
            type: "button",
            title: "Apply ".concat(item.query, " as search"),
            onClick: function onClick(evt) {
                evt.stopPropagation();
                onApply();
            }
        }, /*#__PURE__*/ createElement(icons.ApplyIcon, {
            createElement: createElement
        }))));
    };
}

exports.createAutocompleteSuggestionComponent = createAutocompleteSuggestionComponent;
