// instantsearch-ui-components@0.20.0/dist/cjs/components/autocomplete/AutocompleteDetachedFormContainer.js
'use strict';

var cx = require('../../lib/cx.js');
var Button = require('../Button.js');

function createAutocompleteDetachedFormContainerComponent(param) {
    var createElement = param.createElement;
    var Button$1 = Button.createButtonComponent({
        createElement: createElement
    });
    return function AutocompleteDetachedFormContainer(userProps) {
        var children = userProps.children, _userProps_classNames = userProps.classNames, classNames = _userProps_classNames === void 0 ? {} : _userProps_classNames, onCancel = userProps.onCancel, translations = userProps.translations;
        return /*#__PURE__*/ createElement("div", {
            className: cx.cx('ais-AutocompleteDetachedFormContainer', classNames.detachedFormContainer)
        }, children, /*#__PURE__*/ createElement(Button$1, {
            variant: "ghost",
            className: cx.cx('ais-AutocompleteDetachedCancelButton', classNames.detachedCancelButton),
            onClick: onCancel
        }, translations.detachedCancelButtonText));
    };
}

exports.createAutocompleteDetachedFormContainerComponent = createAutocompleteDetachedFormContainerComponent;
