// instantsearch.js@4.90.0/cjs/widgets/sort-by/sort-by.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _object_spread = require('@swc/helpers/_/_object_spread');
var _object_spread_props = require('@swc/helpers/_/_object_spread_props');
var instantsearchUiComponents = require('instantsearch-ui-components');
var preact = require('preact');
var Selector = require('../../components/Selector/Selector.js');
var connectSortBy = require('../../connectors/sort-by/connectSortBy.js');
var suit$1 = require('../../lib/suit.js');
var documentation = require('../../lib/utils/documentation.js');
var getContainerNode = require('../../lib/utils/getContainerNode.js');

var withUsage = documentation.createDocumentationMessageGenerator({
    name: 'sort-by'
});
var suit = suit$1.component('SortBy');
var renderer = function renderer(param) {
    var containerNode = param.containerNode, cssClasses = param.cssClasses;
    return function(param, isFirstRendering) {
        var currentRefinement = param.currentRefinement, options = param.options, refine = param.refine;
        if (isFirstRendering) {
            return;
        }
        preact.render(/*#__PURE__*/ preact.h("div", {
            className: cssClasses.root
        }, /*#__PURE__*/ preact.h(Selector.default, {
            cssClasses: cssClasses,
            currentValue: currentRefinement,
            options: options,
            setValue: refine,
            ariaLabel: "Sort results by"
        })), containerNode);
    };
};
/**
 * Sort by selector is a widget used for letting the user choose between different
 * indices that contains the same data with a different order / ranking formula.
 */ var sortBy = function sortBy(widgetParams) {
    var _ref = widgetParams || {}, container = _ref.container, items = _ref.items, tmp = _ref.cssClasses, userCssClasses = tmp === void 0 ? {} : tmp, transformItems = _ref.transformItems;
    if (!container) {
        throw new Error(withUsage('The `container` option is required.'));
    }
    var containerNode = getContainerNode.getContainerNode(container);
    var cssClasses = {
        root: instantsearchUiComponents.cx(suit(), userCssClasses.root),
        select: instantsearchUiComponents.cx(suit({
            descendantName: 'select'
        }), userCssClasses.select),
        option: instantsearchUiComponents.cx(suit({
            descendantName: 'option'
        }), userCssClasses.option)
    };
    var specializedRenderer = renderer({
        containerNode: containerNode,
        cssClasses: cssClasses
    });
    var makeWidget = connectSortBy.default(specializedRenderer, function() {
        return preact.render(null, containerNode);
    });
    return _object_spread_props._(_object_spread._({}, makeWidget({
        container: containerNode,
        items: items,
        transformItems: transformItems
    })), {
        $$widgetType: 'ais.sortBy'
    });
};

exports.default = sortBy;
