// instantsearch.js@4.90.0/cjs/templates/carousel/carousel.js
'use strict';

var _object_spread = require('@swc/helpers/_/_object_spread');
var _sliced_to_array = require('@swc/helpers/_/_sliced_to_array');
var preact$1 = require('htm/preact');
var instantsearchUiComponents = require('instantsearch-ui-components');
var preact = require('preact');
var hooks = require('preact/hooks');

var Carousel = instantsearchUiComponents.createCarouselComponent({
    createElement: preact.h,
    Fragment: preact.Fragment
});
function CarouselWithRefs(props) {
    var _useState = _sliced_to_array._(hooks.useState(false), 2), canScrollLeft = _useState[0], setCanScrollLeft = _useState[1];
    var _useState1 = _sliced_to_array._(hooks.useState(true), 2), canScrollRight = _useState1[0], setCanScrollRight = _useState1[1];
    var carouselRefs = {
        listRef: hooks.useRef(null),
        nextButtonRef: hooks.useRef(null),
        previousButtonRef: hooks.useRef(null),
        carouselIdRef: hooks.useRef(instantsearchUiComponents.generateCarouselId()),
        canScrollLeft: canScrollLeft,
        canScrollRight: canScrollRight,
        setCanScrollLeft: setCanScrollLeft,
        setCanScrollRight: setCanScrollRight
    };
    return /*#__PURE__*/ preact.h(Carousel, _object_spread._({}, carouselRefs, props));
}
function carousel() {
    var _ref = arguments.length > 0 && arguments[0] !== void 0 ? arguments[0] : {}, cssClasses = _ref.cssClasses, _ref_templates = _ref.templates, templates = _ref_templates === void 0 ? {} : _ref_templates, _ref_showNavigation = _ref.showNavigation, showNavigation = _ref_showNavigation === void 0 ? true : _ref_showNavigation;
    return function CarouselTemplate(param) {
        var items = param.items, widgetTemplates = param.templates, tmp = param.cssClasses, widgetCssClasses = tmp === void 0 ? {} : tmp, _param_sendEvent = param.sendEvent, sendEvent = _param_sendEvent === void 0 ? function() {} : _param_sendEvent;
        var previous = templates.previous, next = templates.next, header = templates.header;
        return /*#__PURE__*/ preact.h(CarouselWithRefs, {
            items: items,
            sendEvent: sendEvent,
            itemComponent: widgetTemplates.item,
            headerComponent: header ? function(props) {
                return header(_object_spread._({
                    html: preact$1.html
                }, props));
            } : undefined,
            previousIconComponent: previous ? function() {
                return previous({
                    html: preact$1.html
                });
            } : undefined,
            nextIconComponent: next ? function() {
                return next({
                    html: preact$1.html
                });
            } : undefined,
            classNames: _object_spread._({}, cssClasses, {
                list: instantsearchUiComponents.cx(cssClasses === null || cssClasses === void 0 ? void 0 : cssClasses.list, widgetCssClasses === null || widgetCssClasses === void 0 ? void 0 : widgetCssClasses.list),
                item: instantsearchUiComponents.cx(cssClasses === null || cssClasses === void 0 ? void 0 : cssClasses.item, widgetCssClasses === null || widgetCssClasses === void 0 ? void 0 : widgetCssClasses.item)
            }),
            showNavigation: showNavigation
        });
    };
}

exports.carousel = carousel;
