// react-instantsearch@7.26.0/dist/cjs/index.js
'use strict';

var reactInstantsearchCore = require('react-instantsearch-core');
var useStickToBottom = require('./lib/useStickToBottom.js');
var Autocomplete = require('./widgets/Autocomplete.js');
var Breadcrumb = require('./widgets/Breadcrumb.js');
var chat = require('instantsearch.js/cjs/lib/chat');
var Chat = require('./widgets/Chat.js');
var ClearRefinements = require('./widgets/ClearRefinements.js');
var CurrentRefinements = require('./widgets/CurrentRefinements.js');
var FrequentlyBoughtTogether = require('./widgets/FrequentlyBoughtTogether.js');
var HierarchicalMenu = require('./widgets/HierarchicalMenu.js');
var Highlight = require('./widgets/Highlight.js');
var Hits = require('./widgets/Hits.js');
var HitsPerPage = require('./widgets/HitsPerPage.js');
var InfiniteHits = require('./widgets/InfiniteHits.js');
var LookingSimilar = require('./widgets/LookingSimilar.js');
var Menu = require('./widgets/Menu.js');
var Pagination = require('./widgets/Pagination.js');
var PoweredBy = require('./widgets/PoweredBy.js');
var RangeInput = require('./widgets/RangeInput.js');
var RefinementList = require('./widgets/RefinementList.js');
var RelatedProducts = require('./widgets/RelatedProducts.js');
var ReverseHighlight = require('./widgets/ReverseHighlight.js');
var SearchBox = require('./widgets/SearchBox.js');
var Snippet = require('./widgets/Snippet.js');
var SortBy = require('./widgets/SortBy.js');
var Stats = require('./widgets/Stats.js');
var ToggleRefinement = require('./widgets/ToggleRefinement.js');
var TrendingItems = require('./widgets/TrendingItems.js');
var FilterSuggestions = require('./widgets/FilterSuggestions.js');
var Carousel = require('./components/Carousel.js');



exports.useStickToBottom = useStickToBottom.useStickToBottom;
exports.EXPERIMENTAL_Autocomplete = Autocomplete.EXPERIMENTAL_Autocomplete;
exports.Breadcrumb = Breadcrumb.Breadcrumb;
Object.defineProperty(exports, "MemorizeToolType", {
	enumerable: true,
	get: function () { return chat.MemorizeToolType; }
});
Object.defineProperty(exports, "MemorySearchToolType", {
	enumerable: true,
	get: function () { return chat.MemorySearchToolType; }
});
Object.defineProperty(exports, "PonderToolType", {
	enumerable: true,
	get: function () { return chat.PonderToolType; }
});
Object.defineProperty(exports, "RecommendToolType", {
	enumerable: true,
	get: function () { return chat.RecommendToolType; }
});
Object.defineProperty(exports, "SearchIndexToolType", {
	enumerable: true,
	get: function () { return chat.SearchIndexToolType; }
});
exports.Chat = Chat.Chat;
exports.createDefaultTools = Chat.createDefaultTools;
exports.ClearRefinements = ClearRefinements.ClearRefinements;
exports.CurrentRefinements = CurrentRefinements.CurrentRefinements;
exports.FrequentlyBoughtTogether = FrequentlyBoughtTogether.FrequentlyBoughtTogether;
exports.HierarchicalMenu = HierarchicalMenu.HierarchicalMenu;
exports.Highlight = Highlight.Highlight;
exports.Hits = Hits.Hits;
exports.HitsPerPage = HitsPerPage.HitsPerPage;
exports.InfiniteHits = InfiniteHits.InfiniteHits;
exports.LookingSimilar = LookingSimilar.LookingSimilar;
exports.Menu = Menu.Menu;
exports.Pagination = Pagination.Pagination;
exports.PoweredBy = PoweredBy.PoweredBy;
exports.RangeInput = RangeInput.RangeInput;
exports.RefinementList = RefinementList.RefinementList;
exports.RelatedProducts = RelatedProducts.RelatedProducts;
exports.ReverseHighlight = ReverseHighlight.ReverseHighlight;
exports.SearchBox = SearchBox.SearchBox;
exports.Snippet = Snippet.Snippet;
exports.SortBy = SortBy.SortBy;
exports.Stats = Stats.Stats;
exports.ToggleRefinement = ToggleRefinement.ToggleRefinement;
exports.TrendingItems = TrendingItems.TrendingItems;
exports.FilterSuggestions = FilterSuggestions.FilterSuggestions;
exports.Carousel = Carousel.Carousel;
Object.keys(reactInstantsearchCore).forEach(function (k) {
	if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
		enumerable: true,
		get: function () { return reactInstantsearchCore[k]; }
	});
});
