// react-instantsearch@7.26.0/dist/cjs/ui/HierarchicalMenu.js
'use strict';

var _object_spread = require('@swc/helpers/_/_object_spread');
var _object_spread_props = require('@swc/helpers/_/_object_spread_props');
var _object_without_properties = require('@swc/helpers/_/_object_without_properties');
var instantsearchUiComponents = require('instantsearch-ui-components');
var React = require('react');
var isModifierClick = require('./lib/isModifierClick.js');
var ShowMoreButton = require('./ShowMoreButton.js');

function HierarchicalList(param) {
    var className = param.className, _param_classNames = param.classNames, classNames = _param_classNames === void 0 ? {} : _param_classNames, items = param.items, createURL = param.createURL, onNavigate = param.onNavigate;
    if (items.length === 0) {
        return null;
    }
    return /*#__PURE__*/ React.createElement("ul", {
        className: instantsearchUiComponents.cx('ais-HierarchicalMenu-list', classNames.list, className)
    }, items.map(function(item) {
        return /*#__PURE__*/ React.createElement("li", {
            key: item.value,
            className: instantsearchUiComponents.cx('ais-HierarchicalMenu-item', classNames.item, item.isRefined && instantsearchUiComponents.cx('ais-HierarchicalMenu-item--selected', classNames.selectedItem), item.data && item.data.length > 0 && instantsearchUiComponents.cx('ais-HierarchicalMenu-item--parent', classNames.parentItem))
        }, /*#__PURE__*/ React.createElement("a", {
            className: instantsearchUiComponents.cx('ais-HierarchicalMenu-link', classNames.link, item.isRefined && instantsearchUiComponents.cx('ais-HierarchicalMenu-link--selected', classNames.selectedItemLink)),
            href: createURL(item.value),
            onClick: function onClick(event) {
                if (isModifierClick.isModifierClick(event)) {
                    return;
                }
                event.preventDefault();
                onNavigate(item.value);
            }
        }, /*#__PURE__*/ React.createElement("span", {
            className: instantsearchUiComponents.cx('ais-HierarchicalMenu-label', classNames.label)
        }, item.label), /*#__PURE__*/ React.createElement("span", {
            className: instantsearchUiComponents.cx('ais-HierarchicalMenu-count', classNames.count)
        }, item.count)), item.data && /*#__PURE__*/ React.createElement(HierarchicalList, {
            className: instantsearchUiComponents.cx('ais-HierarchicalMenu-list--child', classNames.childList),
            classNames: classNames,
            items: item.data,
            onNavigate: onNavigate,
            createURL: createURL
        }));
    }));
}
function HierarchicalMenu(_0) {
    var _0_classNames = _0.classNames, classNames = _0_classNames === void 0 ? {} : _0_classNames, items = _0.items, hasItems = _0.hasItems, onNavigate = _0.onNavigate, createURL = _0.createURL, showMore = _0.showMore, canToggleShowMore = _0.canToggleShowMore, onToggleShowMore = _0.onToggleShowMore, isShowingMore = _0.isShowingMore, translations = _0.translations, props = _object_without_properties._(_0, [
        "classNames",
        "items",
        "hasItems",
        "onNavigate",
        "createURL",
        "showMore",
        "canToggleShowMore",
        "onToggleShowMore",
        "isShowingMore",
        "translations"
    ]);
    return /*#__PURE__*/ React.createElement("div", _object_spread_props._(_object_spread._({}, props), {
        className: instantsearchUiComponents.cx('ais-HierarchicalMenu', classNames.root, !hasItems && instantsearchUiComponents.cx('ais-HierarchicalMenu--noRefinement', classNames.noRefinementRoot), props.className)
    }), /*#__PURE__*/ React.createElement(HierarchicalList, {
        classNames: classNames,
        items: items,
        onNavigate: onNavigate,
        createURL: createURL
    }), showMore && /*#__PURE__*/ React.createElement(ShowMoreButton.ShowMoreButton, {
        className: instantsearchUiComponents.cx('ais-HierarchicalMenu-showMore', classNames.showMore, !canToggleShowMore && instantsearchUiComponents.cx('ais-HierarchicalMenu-showMore--disabled', classNames.disabledShowMore)),
        disabled: !canToggleShowMore,
        onClick: onToggleShowMore,
        isShowingMore: isShowingMore,
        translations: translations
    }));
}

exports.HierarchicalMenu = HierarchicalMenu;
