// react-instantsearch@7.26.0/dist/cjs/ui/Menu.js
'use strict';

var _object_spread = require('@swc/helpers/_/_object_spread');
var _object_spread_props = require('@swc/helpers/_/_object_spread_props');
var _object_without_properties = require('@swc/helpers/_/_object_without_properties');
var instantsearchUiComponents = require('instantsearch-ui-components');
var React = require('react');
var isModifierClick = require('./lib/isModifierClick.js');
var ShowMoreButton = require('./ShowMoreButton.js');

function Menu(_0) {
    var items = _0.items, _0_classNames = _0.classNames, classNames = _0_classNames === void 0 ? {} : _0_classNames, showMore = _0.showMore, canToggleShowMore = _0.canToggleShowMore, onToggleShowMore = _0.onToggleShowMore, isShowingMore = _0.isShowingMore, createURL = _0.createURL, onRefine = _0.onRefine, translations = _0.translations, props = _object_without_properties._(_0, [
        "items",
        "classNames",
        "showMore",
        "canToggleShowMore",
        "onToggleShowMore",
        "isShowingMore",
        "createURL",
        "onRefine",
        "translations"
    ]);
    return /*#__PURE__*/ React.createElement("div", _object_spread_props._(_object_spread._({}, props), {
        className: instantsearchUiComponents.cx('ais-Menu', classNames.root, items.length === 0 && instantsearchUiComponents.cx('ais-Menu--noRefinement', classNames.noRefinementRoot), props.className)
    }), /*#__PURE__*/ React.createElement("ul", {
        className: instantsearchUiComponents.cx('ais-Menu-list', classNames.list)
    }, items.map(function(item) {
        return /*#__PURE__*/ React.createElement("li", {
            key: item.label,
            className: instantsearchUiComponents.cx('ais-Menu-item', classNames.item, item.isRefined && instantsearchUiComponents.cx('ais-Menu-item--selected', classNames.selectedItem))
        }, /*#__PURE__*/ React.createElement("a", {
            className: instantsearchUiComponents.cx('ais-Menu-link', classNames.link),
            href: createURL(item.value),
            onClick: function onClick(event) {
                if (isModifierClick.isModifierClick(event)) {
                    return;
                }
                event.preventDefault();
                onRefine(item);
            }
        }, /*#__PURE__*/ React.createElement("span", {
            className: instantsearchUiComponents.cx('ais-Menu-label', classNames.label)
        }, item.label), /*#__PURE__*/ React.createElement("span", {
            className: instantsearchUiComponents.cx('ais-Menu-count', classNames.count)
        }, item.count)));
    })), showMore && /*#__PURE__*/ React.createElement(ShowMoreButton.ShowMoreButton, {
        className: instantsearchUiComponents.cx('ais-Menu-showMore', classNames.showMore, !canToggleShowMore && instantsearchUiComponents.cx('ais-Menu-showMore--disabled', classNames.disabledShowMore)),
        disabled: !canToggleShowMore,
        onClick: onToggleShowMore,
        isShowingMore: isShowingMore,
        translations: translations
    }));
}

exports.Menu = Menu;
