// react-instantsearch@7.26.0/dist/cjs/ui/RefinementList.js
'use strict';

var _object_spread = require('@swc/helpers/_/_object_spread');
var _object_spread_props = require('@swc/helpers/_/_object_spread_props');
var _object_without_properties = require('@swc/helpers/_/_object_without_properties');
var instantsearchUiComponents = require('instantsearch-ui-components');
var utils = require('instantsearch.js/cjs/lib/utils');
var React = require('react');
var Highlight = require('./Highlight.js');
var ShowMoreButton = require('./ShowMoreButton.js');

function RefinementList(_0) {
    _0.canRefine; var items = _0.items, onRefine = _0.onRefine, query = _0.query, searchBox = _0.searchBox, noResults = _0.noResults, showMore = _0.showMore, canToggleShowMore = _0.canToggleShowMore, onToggleShowMore = _0.onToggleShowMore, isShowingMore = _0.isShowingMore, className = _0.className, _0_classNames = _0.classNames, classNames = _0_classNames === void 0 ? {} : _0_classNames, translations = _0.translations, props = _object_without_properties._(_0, [
        "canRefine",
        "items",
        "onRefine",
        "query",
        "searchBox",
        "noResults",
        "showMore",
        "canToggleShowMore",
        "onToggleShowMore",
        "isShowingMore",
        "className",
        "classNames",
        "translations"
    ]);
    return /*#__PURE__*/ React.createElement("div", _object_spread_props._(_object_spread._({}, props), {
        className: instantsearchUiComponents.cx('ais-RefinementList', classNames.root, items.length === 0 && instantsearchUiComponents.cx('ais-RefinementList--noRefinement', classNames.noRefinementRoot), className)
    }), searchBox && /*#__PURE__*/ React.createElement("div", {
        className: instantsearchUiComponents.cx('ais-RefinementList-searchBox', classNames.searchBox)
    }, searchBox), noResults ? /*#__PURE__*/ React.createElement("div", {
        className: instantsearchUiComponents.cx('ais-RefinementList-noResults', classNames.noResults)
    }, noResults) : /*#__PURE__*/ React.createElement("ul", {
        className: instantsearchUiComponents.cx('ais-RefinementList-list', classNames.list)
    }, items.map(function(item) {
        return /*#__PURE__*/ React.createElement("li", {
            key: item.value,
            className: instantsearchUiComponents.cx('ais-RefinementList-item', classNames.item, item.isRefined && instantsearchUiComponents.cx('ais-RefinementList-item--selected', classNames.selectedItem))
        }, /*#__PURE__*/ React.createElement("label", {
            className: instantsearchUiComponents.cx('ais-RefinementList-label', classNames.label)
        }, /*#__PURE__*/ React.createElement("input", {
            checked: item.isRefined,
            className: instantsearchUiComponents.cx('ais-RefinementList-checkbox', classNames.checkbox),
            type: "checkbox",
            value: item.value,
            onChange: function onChange() {
                onRefine(item);
            }
        }), /*#__PURE__*/ React.createElement("span", {
            className: instantsearchUiComponents.cx('ais-RefinementList-labelText', classNames.labelText)
        }, query.length > 0 ? /*#__PURE__*/ React.createElement(Highlight.Highlight, {
            parts: [
                utils.getHighlightedParts(utils.unescape(item.highlighted || ''))
            ]
        }) : item.label), /*#__PURE__*/ React.createElement("span", {
            className: instantsearchUiComponents.cx('ais-RefinementList-count', classNames.count)
        }, item.count)));
    })), showMore && /*#__PURE__*/ React.createElement(ShowMoreButton.ShowMoreButton, {
        className: instantsearchUiComponents.cx('ais-RefinementList-showMore', classNames.showMore, !canToggleShowMore && instantsearchUiComponents.cx('ais-RefinementList-showMore--disabled', classNames.disabledShowMore)),
        disabled: !canToggleShowMore,
        onClick: onToggleShowMore,
        isShowingMore: isShowingMore,
        translations: translations
    }));
}

exports.RefinementList = RefinementList;
