// react-instantsearch@7.26.0/dist/cjs/ui/Snippet.js
'use strict';

var _object_spread = require('@swc/helpers/_/_object_spread');
var _object_without_properties = require('@swc/helpers/_/_object_without_properties');
var instantsearchUiComponents = require('instantsearch-ui-components');
var React = require('react');
var InternalHighlight = require('./InternalHighlight.js');

function Snippet(_0) {
    var _0_classNames = _0.classNames, classNames = _0_classNames === void 0 ? {} : _0_classNames, props = _object_without_properties._(_0, [
        "classNames"
    ]);
    return /*#__PURE__*/ React.createElement(InternalHighlight.InternalHighlight, _object_spread._({
        classNames: {
            root: instantsearchUiComponents.cx('ais-Snippet', classNames.root),
            highlighted: instantsearchUiComponents.cx('ais-Snippet-highlighted', classNames.highlighted),
            nonHighlighted: instantsearchUiComponents.cx('ais-Snippet-nonHighlighted', classNames.nonHighlighted),
            separator: instantsearchUiComponents.cx('ais-Snippet-separator', classNames.separator)
        }
    }, props));
}

exports.Snippet = Snippet;
