// react-instantsearch@7.26.0/dist/cjs/ui/Stats.js
'use strict';

var _object_spread = require('@swc/helpers/_/_object_spread');
var _object_spread_props = require('@swc/helpers/_/_object_spread_props');
var _object_without_properties = require('@swc/helpers/_/_object_without_properties');
var instantsearchUiComponents = require('instantsearch-ui-components');
var React = require('react');

function Stats(_0) {
    var _0_classNames = _0.classNames, classNames = _0_classNames === void 0 ? {} : _0_classNames, nbHits = _0.nbHits, processingTimeMS = _0.processingTimeMS, nbSortedHits = _0.nbSortedHits, areHitsSorted = _0.areHitsSorted, translations = _0.translations, props = _object_without_properties._(_0, [
        "classNames",
        "nbHits",
        "processingTimeMS",
        "nbSortedHits",
        "areHitsSorted",
        "translations"
    ]);
    var translationOptions = {
        nbHits: nbHits,
        processingTimeMS: processingTimeMS,
        nbSortedHits: nbSortedHits,
        areHitsSorted: areHitsSorted
    };
    return /*#__PURE__*/ React.createElement("div", _object_spread_props._(_object_spread._({}, props), {
        className: instantsearchUiComponents.cx('ais-Stats', classNames.root, props.className)
    }), /*#__PURE__*/ React.createElement("span", {
        className: "ais-Stats-text"
    }, translations.rootElementText(translationOptions)));
}

exports.Stats = Stats;
