// react-instantsearch@7.26.0/dist/cjs/ui/ToggleRefinement.js
'use strict';

var _object_spread = require('@swc/helpers/_/_object_spread');
var _object_spread_props = require('@swc/helpers/_/_object_spread_props');
var _object_without_properties = require('@swc/helpers/_/_object_without_properties');
var instantsearchUiComponents = require('instantsearch-ui-components');
var React = require('react');

function ToggleRefinement(_0) {
    var _0_classNames = _0.classNames, classNames = _0_classNames === void 0 ? {} : _0_classNames, checked = _0.checked, onChange = _0.onChange, label = _0.label, props = _object_without_properties._(_0, [
        "classNames",
        "checked",
        "onChange",
        "label"
    ]);
    return /*#__PURE__*/ React.createElement("div", _object_spread_props._(_object_spread._({}, props), {
        className: instantsearchUiComponents.cx('ais-ToggleRefinement', classNames.root, props.className)
    }), /*#__PURE__*/ React.createElement("label", {
        className: instantsearchUiComponents.cx('ais-ToggleRefinement-label', classNames.label)
    }, /*#__PURE__*/ React.createElement("input", {
        className: instantsearchUiComponents.cx('ais-ToggleRefinement-checkbox', classNames.checkbox),
        type: "checkbox",
        checked: checked,
        onChange: function onChange1(event) {
            onChange(event.target.checked);
        }
    }), /*#__PURE__*/ React.createElement("span", {
        className: instantsearchUiComponents.cx('ais-ToggleRefinement-labelText', classNames.labelText)
    }, label)));
}

exports.ToggleRefinement = ToggleRefinement;
