// instantsearch.js@4.90.0/cjs/widgets/refinement-list/defaultTemplates.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var instantsearchUiComponents = require('instantsearch-ui-components');
var preact = require('preact');
var formatNumber = require('../../lib/formatNumber.js');

var defaultTemplates = {
    item: function item(param) {
        var cssClasses = param.cssClasses, count = param.count, value = param.value, highlighted = param.highlighted, isRefined = param.isRefined, isFromSearch = param.isFromSearch;
        return /*#__PURE__*/ preact.h("label", {
            className: instantsearchUiComponents.cx(cssClasses.label)
        }, /*#__PURE__*/ preact.h("input", {
            type: "checkbox",
            className: instantsearchUiComponents.cx(cssClasses.checkbox),
            value: value,
            defaultChecked: isRefined
        }), /*#__PURE__*/ preact.h("span", {
            className: instantsearchUiComponents.cx(cssClasses.labelText),
            dangerouslySetInnerHTML: isFromSearch ? {
                __html: highlighted
            } : undefined
        }, !isFromSearch && highlighted), /*#__PURE__*/ preact.h("span", {
            className: instantsearchUiComponents.cx(cssClasses.count)
        }, formatNumber.formatNumber(count)));
    },
    showMoreText: function showMoreText(param) {
        var isShowingMore = param.isShowingMore;
        return isShowingMore ? 'Show less' : 'Show more';
    },
    searchableNoResults: function searchableNoResults() {
        return 'No results';
    }
};

exports.default = defaultTemplates;
