// instantsearch.js@4.90.0/cjs/components/Panel/Panel.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _sliced_to_array = require('@swc/helpers/_/_sliced_to_array');
var instantsearchUiComponents = require('instantsearch-ui-components');
var preact = require('preact');
var hooks = require('preact/hooks');
var Template = require('../Template/Template.js');

function Panel(props) {
    var _useState = _sliced_to_array._(hooks.useState(props.isCollapsed), 2), isCollapsed = _useState[0], setIsCollapsed = _useState[1];
    var _useState1 = _sliced_to_array._(hooks.useState(false), 2), isControlled = _useState1[0], setIsControlled = _useState1[1];
    var bodyRef = hooks.useRef(null);
    hooks.useEffect(function() {
        var node = bodyRef.current;
        if (!node) {
            return undefined;
        }
        node.appendChild(props.bodyElement);
        return function() {
            node.removeChild(props.bodyElement);
        };
    }, [
        bodyRef,
        props.bodyElement
    ]);
    if (!isControlled && props.isCollapsed !== isCollapsed) {
        setIsCollapsed(props.isCollapsed);
    }
    return /*#__PURE__*/ preact.h("div", {
        className: instantsearchUiComponents.cx(props.cssClasses.root, props.hidden && props.cssClasses.noRefinementRoot, props.collapsible && props.cssClasses.collapsibleRoot, isCollapsed && props.cssClasses.collapsedRoot),
        hidden: props.hidden
    }, props.templates.header && /*#__PURE__*/ preact.h("div", {
        className: props.cssClasses.header
    }, /*#__PURE__*/ preact.h(Template.default, {
        templates: props.templates,
        templateKey: "header",
        rootTagName: "span",
        data: props.data
    }), props.collapsible && /*#__PURE__*/ preact.h("button", {
        className: props.cssClasses.collapseButton,
        "aria-expanded": !isCollapsed,
        onClick: function onClick(event) {
            event.preventDefault();
            setIsControlled(true);
            setIsCollapsed(function(prevIsCollapsed) {
                return !prevIsCollapsed;
            });
        }
    }, /*#__PURE__*/ preact.h(Template.default, {
        templates: props.templates,
        templateKey: "collapseButtonText",
        rootTagName: "span",
        data: {
            collapsed: isCollapsed
        }
    }))), /*#__PURE__*/ preact.h("div", {
        className: props.cssClasses.body,
        ref: bodyRef
    }), props.templates.footer && /*#__PURE__*/ preact.h(Template.default, {
        templates: props.templates,
        templateKey: "footer",
        rootProps: {
            className: props.cssClasses.footer
        },
        data: props.data
    }));
}

exports.default = Panel;
