// instantsearch.js@4.90.0/cjs/lib/utils/index.js
'use strict';

var addWidgetId = require('./addWidgetId.js');
var capitalize = require('./capitalize.js');
var checkIndexUiState = require('./checkIndexUiState.js');
var checkRendering = require('./checkRendering.js');
var clearRefinements = require('./clearRefinements.js');
var concatHighlightedParts = require('./concatHighlightedParts.js');
var createConcurrentSafePromise = require('./createConcurrentSafePromise.js');
var createSendEventForFacet = require('./createSendEventForFacet.js');
var createSendEventForHits = require('./createSendEventForHits.js');
var setIndexHelperState = require('./setIndexHelperState.js');
var isIndexWidget = require('./isIndexWidget.js');
var debounce = require('./debounce.js');
var defer = require('./defer.js');
var documentation = require('./documentation.js');
var escapeHighlight = require('./escape-highlight.js');
var escapeHtml = require('./escape-html.js');
var escapeFacetValue = require('./escapeFacetValue.js');
var find = require('./find.js');
var findIndex = require('./findIndex.js');
var geoSearch = require('./geo-search.js');
var getAlgoliaAgent = require('./getAlgoliaAgent.js');
var getAppIdAndApiKey = require('./getAppIdAndApiKey.js');
var getContainerNode = require('./getContainerNode.js');
var getHighlightedParts = require('./getHighlightedParts.js');
var getHighlightFromSiblings = require('./getHighlightFromSiblings.js');
var getObjectType = require('./getObjectType.js');
var getPropertyByPath = require('./getPropertyByPath.js');
var getRefinements = require('./getRefinements.js');
var getWidgetAttribute = require('./getWidgetAttribute.js');
var hitsAbsolutePosition = require('./hits-absolute-position.js');
var hitsQueryId = require('./hits-query-id.js');
var hydrateRecommendCache = require('./hydrateRecommendCache.js');
var hydrateSearchClient = require('./hydrateSearchClient.js');
var isDomElement = require('./isDomElement.js');
var isEqual = require('./isEqual.js');
var isFacetRefined = require('./isFacetRefined.js');
var isFiniteNumber = require('./isFiniteNumber.js');
var isPlainObject = require('./isPlainObject.js');
var isSpecialClick = require('./isSpecialClick.js');
var walkIndex = require('./walkIndex.js');
var logger = require('./logger.js');
var mergeSearchParameters = require('./mergeSearchParameters.js');
var omit = require('./omit.js');
var noop = require('./noop.js');
var range = require('./range.js');
var renderArgs = require('./render-args.js');
var resolveSearchParameters = require('./resolveSearchParameters.js');
var reverseHighlightedParts = require('./reverseHighlightedParts.js');
var safelyRunOnBrowser = require('./safelyRunOnBrowser.js');
var serializer = require('./serializer.js');
var toArray = require('./toArray.js');
var uniq = require('./uniq.js');



exports.addWidgetId = addWidgetId.addWidgetId;
exports.resetWidgetId = addWidgetId.resetWidgetId;
exports.capitalize = capitalize.capitalize;
exports.checkIndexUiState = checkIndexUiState.checkIndexUiState;
exports.checkRendering = checkRendering.checkRendering;
exports.clearRefinements = clearRefinements.clearRefinements;
exports.concatHighlightedParts = concatHighlightedParts.concatHighlightedParts;
exports.createConcurrentSafePromise = createConcurrentSafePromise.createConcurrentSafePromise;
exports.createSendEventForFacet = createSendEventForFacet.createSendEventForFacet;
exports._buildEventPayloadsForHits = createSendEventForHits._buildEventPayloadsForHits;
exports.createBindEventForHits = createSendEventForHits.createBindEventForHits;
exports.createSendEventForHits = createSendEventForHits.createSendEventForHits;
exports.setIndexHelperState = setIndexHelperState.setIndexHelperState;
exports.isIndexWidget = isIndexWidget.isIndexWidget;
exports.debounce = debounce.debounce;
exports.defer = defer.defer;
exports.createDocumentationLink = documentation.createDocumentationLink;
exports.createDocumentationMessageGenerator = documentation.createDocumentationMessageGenerator;
exports.TAG_PLACEHOLDER = escapeHighlight.TAG_PLACEHOLDER;
exports.TAG_REPLACEMENT = escapeHighlight.TAG_REPLACEMENT;
exports.escapeFacets = escapeHighlight.escapeFacets;
exports.escapeHits = escapeHighlight.escapeHits;
exports.escape = escapeHtml.escape;
exports.unescape = escapeHtml.unescape;
exports.escapeFacetValue = escapeFacetValue.escapeFacetValue;
exports.unescapeFacetValue = escapeFacetValue.unescapeFacetValue;
exports.find = find.find;
exports.findIndex = findIndex.findIndex;
exports.aroundLatLngToPosition = geoSearch.aroundLatLngToPosition;
exports.insideBoundingBoxToBoundingBox = geoSearch.insideBoundingBoxToBoundingBox;
exports.getAlgoliaAgent = getAlgoliaAgent.getAlgoliaAgent;
exports.getAppIdAndApiKey = getAppIdAndApiKey.getAppIdAndApiKey;
exports.getContainerNode = getContainerNode.getContainerNode;
exports.getHighlightedParts = getHighlightedParts.getHighlightedParts;
exports.getHighlightFromSiblings = getHighlightFromSiblings.getHighlightFromSiblings;
exports.getObjectType = getObjectType.getObjectType;
exports.getPropertyByPath = getPropertyByPath.getPropertyByPath;
exports.getRefinements = getRefinements.getRefinements;
exports.getWidgetAttribute = getWidgetAttribute.getWidgetAttribute;
exports.addAbsolutePosition = hitsAbsolutePosition.addAbsolutePosition;
exports.addQueryID = hitsQueryId.addQueryID;
exports.hydrateRecommendCache = hydrateRecommendCache.hydrateRecommendCache;
exports.hydrateSearchClient = hydrateSearchClient.hydrateSearchClient;
exports.isDomElement = isDomElement.isDomElement;
exports.isEqual = isEqual.isEqual;
exports.isFacetRefined = isFacetRefined.isFacetRefined;
exports.isFiniteNumber = isFiniteNumber.isFiniteNumber;
exports.isPlainObject = isPlainObject.isPlainObject;
exports.isSpecialClick = isSpecialClick.isSpecialClick;
exports.walkIndex = walkIndex.walkIndex;
exports.deprecate = logger.deprecate;
exports.warn = logger.warn;
exports.warning = logger.warning;
exports.mergeSearchParameters = mergeSearchParameters.mergeSearchParameters;
exports.omit = omit.omit;
exports.noop = noop.noop;
exports.range = range.range;
exports.createInitArgs = renderArgs.createInitArgs;
exports.createRenderArgs = renderArgs.createRenderArgs;
exports.resolveSearchParameters = resolveSearchParameters.resolveSearchParameters;
exports.reverseHighlightedParts = reverseHighlightedParts.reverseHighlightedParts;
exports.safelyRunOnBrowser = safelyRunOnBrowser.safelyRunOnBrowser;
exports.deserializePayload = serializer.deserializePayload;
exports.serializePayload = serializer.serializePayload;
exports.toArray = toArray.toArray;
exports.uniq = uniq.uniq;
