// instantsearch.js@4.90.0/cjs/widgets/voice-search/voice-search.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _object_spread = require('@swc/helpers/_/_object_spread');
var _object_spread_props = require('@swc/helpers/_/_object_spread_props');
var instantsearchUiComponents = require('instantsearch-ui-components');
var preact = require('preact');
var VoiceSearch = require('../../components/VoiceSearch/VoiceSearch.js');
var connectVoiceSearch = require('../../connectors/voice-search/connectVoiceSearch.js');
var suit$1 = require('../../lib/suit.js');
var defaultTemplates = require('./defaultTemplates.js');
var documentation = require('../../lib/utils/documentation.js');
var getContainerNode = require('../../lib/utils/getContainerNode.js');

var withUsage = documentation.createDocumentationMessageGenerator({
    name: 'voice-search'
});
var suit = suit$1.component('VoiceSearch');
var renderer = function renderer(param) {
    var containerNode = param.containerNode, cssClasses = param.cssClasses, templates = param.templates;
    return function(param) {
        var isBrowserSupported = param.isBrowserSupported, isListening = param.isListening, toggleListening = param.toggleListening, voiceListeningState = param.voiceListeningState;
        preact.render(/*#__PURE__*/ preact.h(VoiceSearch.default, {
            cssClasses: cssClasses,
            templates: templates,
            isBrowserSupported: isBrowserSupported,
            isListening: isListening,
            toggleListening: toggleListening,
            voiceListeningState: voiceListeningState
        }), containerNode);
    };
};
var voiceSearch = function voiceSearch(widgetParams) {
    var _ref = widgetParams || {}, container = _ref.container, tmp = _ref.cssClasses, userCssClasses = tmp === void 0 ? {} : tmp, tmp1 = _ref.templates, userTemplates = tmp1 === void 0 ? {} : tmp1, _ref_searchAsYouSpeak = _ref.searchAsYouSpeak, searchAsYouSpeak = _ref_searchAsYouSpeak === void 0 ? false : _ref_searchAsYouSpeak, language = _ref.language, additionalQueryParameters = _ref.additionalQueryParameters, createVoiceSearchHelper = _ref.createVoiceSearchHelper;
    if (!container) {
        throw new Error(withUsage('The `container` option is required.'));
    }
    var containerNode = getContainerNode.getContainerNode(container);
    var cssClasses = {
        root: instantsearchUiComponents.cx(suit(), userCssClasses.root),
        button: instantsearchUiComponents.cx(suit({
            descendantName: 'button'
        }), userCssClasses.button),
        status: instantsearchUiComponents.cx(suit({
            descendantName: 'status'
        }), userCssClasses.status)
    };
    var templates = _object_spread._({}, defaultTemplates.default, userTemplates);
    var specializedRenderer = renderer({
        containerNode: containerNode,
        cssClasses: cssClasses,
        templates: templates
    });
    var makeWidget = connectVoiceSearch.default(specializedRenderer, function() {
        return preact.render(null, containerNode);
    });
    return _object_spread_props._(_object_spread._({}, makeWidget({
        container: containerNode,
        cssClasses: cssClasses,
        templates: templates,
        searchAsYouSpeak: searchAsYouSpeak,
        language: language,
        additionalQueryParameters: additionalQueryParameters,
        createVoiceSearchHelper: createVoiceSearchHelper
    })), {
        $$widgetType: 'ais.voiceSearch'
    });
};

exports.default = voiceSearch;
