// instantsearch.js@4.90.0/cjs/components/ClearRefinements/ClearRefinements.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _object_spread = require('@swc/helpers/_/_object_spread');
var _object_spread_props = require('@swc/helpers/_/_object_spread_props');
var instantsearchUiComponents = require('instantsearch-ui-components');
var preact = require('preact');
var Template = require('../Template/Template.js');

var ClearRefinements = function ClearRefinements(param) {
    var hasRefinements = param.hasRefinements, refine = param.refine, cssClasses = param.cssClasses, templateProps = param.templateProps;
    return /*#__PURE__*/ preact.h("div", {
        className: cssClasses.root
    }, /*#__PURE__*/ preact.h(Template.default, _object_spread_props._(_object_spread._({}, templateProps), {
        templateKey: "resetLabel",
        rootTagName: "button",
        rootProps: {
            className: instantsearchUiComponents.cx(cssClasses.button, !hasRefinements && cssClasses.disabledButton),
            onClick: refine,
            disabled: !hasRefinements
        },
        data: {
            hasRefinements: hasRefinements
        }
    })));
};

exports.default = ClearRefinements;
