// instantsearch.js@4.90.0/cjs/widgets/clear-refinements/clear-refinements.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _object_spread = require('@swc/helpers/_/_object_spread');
var _object_spread_props = require('@swc/helpers/_/_object_spread_props');
var instantsearchUiComponents = require('instantsearch-ui-components');
var preact = require('preact');
var ClearRefinements = require('../../components/ClearRefinements/ClearRefinements.js');
var connectClearRefinements = require('../../connectors/clear-refinements/connectClearRefinements.js');
var suit$1 = require('../../lib/suit.js');
var defaultTemplates = require('./defaultTemplates.js');
var documentation = require('../../lib/utils/documentation.js');
var getContainerNode = require('../../lib/utils/getContainerNode.js');
var prepareTemplateProps = require('../../lib/templating/prepareTemplateProps.js');

var withUsage = documentation.createDocumentationMessageGenerator({
    name: 'clear-refinements'
});
var suit = suit$1.component('ClearRefinements');
var renderer = function renderer(param) {
    var containerNode = param.containerNode, cssClasses = param.cssClasses, renderState = param.renderState, templates = param.templates;
    return function(param, isFirstRendering) {
        var refine = param.refine, canRefine = param.canRefine, instantSearchInstance = param.instantSearchInstance;
        if (isFirstRendering) {
            renderState.templateProps = prepareTemplateProps.prepareTemplateProps({
                defaultTemplates: defaultTemplates.default,
                templatesConfig: instantSearchInstance.templatesConfig,
                templates: templates
            });
            return;
        }
        preact.render(/*#__PURE__*/ preact.h(ClearRefinements.default, {
            refine: refine,
            cssClasses: cssClasses,
            hasRefinements: canRefine,
            templateProps: renderState.templateProps
        }), containerNode);
    };
};
var clearRefinements = function clearRefinements(widgetParams) {
    var _ref = widgetParams || {}, container = _ref.container, _ref_templates = _ref.templates, templates = _ref_templates === void 0 ? {} : _ref_templates, includedAttributes = _ref.includedAttributes, excludedAttributes = _ref.excludedAttributes, transformItems = _ref.transformItems, tmp = _ref.cssClasses, userCssClasses = tmp === void 0 ? {} : tmp;
    if (!container) {
        throw new Error(withUsage('The `container` option is required.'));
    }
    var containerNode = getContainerNode.getContainerNode(container);
    var cssClasses = {
        root: instantsearchUiComponents.cx(suit(), userCssClasses.root),
        button: instantsearchUiComponents.cx(suit({
            descendantName: 'button'
        }), userCssClasses.button),
        disabledButton: instantsearchUiComponents.cx(suit({
            descendantName: 'button',
            modifierName: 'disabled'
        }), userCssClasses.disabledButton)
    };
    var specializedRenderer = renderer({
        containerNode: containerNode,
        cssClasses: cssClasses,
        renderState: {},
        templates: templates
    });
    var makeWidget = connectClearRefinements.default(specializedRenderer, function() {
        return preact.render(null, containerNode);
    });
    return _object_spread_props._(_object_spread._({}, makeWidget({
        includedAttributes: includedAttributes,
        excludedAttributes: excludedAttributes,
        transformItems: transformItems
    })), {
        $$widgetType: 'ais.clearRefinements'
    });
};

exports.default = clearRefinements;
