// react-instantsearch-core@7.26.0/dist/cjs/components/DynamicWidgets.js
'use strict';

var _object_without_properties = require('@swc/helpers/_/_object_without_properties');
var _type_of = require('@swc/helpers/_/_type_of');
var React = require('react');
var useDynamicWidgets = require('../connectors/useDynamicWidgets.js');
var invariant = require('../lib/invariant.js');

function DefaultFallbackComponent() {
    return null;
}
function DynamicWidgets(_0) {
    var children = _0.children, tmp = _0.fallbackComponent, Fallback = tmp === void 0 ? DefaultFallbackComponent : tmp, props = _object_without_properties._(_0, [
        "children",
        "fallbackComponent"
    ]);
    var FallbackComponent = React.useRef(Fallback);
    var attributesToRender = useDynamicWidgets.useDynamicWidgets(props, {
        $$widgetType: 'ais.dynamicWidgets'
    }).attributesToRender;
    var widgets = new Map();
    React.Children.forEach(children, function(child) {
        var attribute = getWidgetAttribute(child);
        invariant.invariant(attribute !== undefined);
        widgets.set(attribute, child);
    });
    return /*#__PURE__*/ React.createElement(React.Fragment, null, attributesToRender.map(function(attribute) {
        return /*#__PURE__*/ React.createElement(React.Fragment, {
            key: attribute
        }, widgets.get(attribute) || /*#__PURE__*/ React.createElement(FallbackComponent.current, {
            attribute: attribute
        }));
    }));
}
function isReactElement(element) {
    return (typeof element === "undefined" ? "undefined" : _type_of._(element)) === 'object' && element.props;
}
function getWidgetAttribute(element) {
    if (!isReactElement(element)) {
        return undefined;
    }
    if (element.props.attribute) {
        return element.props.attribute;
    }
    if (Array.isArray(element.props.attributes)) {
        return element.props.attributes[0];
    }
    if (element.props.children) {
        invariant.invariant(React.Children.count(element.props.children) === 1);
        return getWidgetAttribute(React.Children.only(element.props.children));
    }
    return undefined;
}

exports.DynamicWidgets = DynamicWidgets;
