// instantsearch.js@4.90.0/cjs/widgets/relevant-sort/relevant-sort.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _object_spread = require('@swc/helpers/_/_object_spread');
var _object_spread_props = require('@swc/helpers/_/_object_spread_props');
var instantsearchUiComponents = require('instantsearch-ui-components');
var preact = require('preact');
var RelevantSort = require('../../components/RelevantSort/RelevantSort.js');
var connectRelevantSort = require('../../connectors/relevant-sort/connectRelevantSort.js');
var suit$1 = require('../../lib/suit.js');
var defaultTemplates = require('./defaultTemplates.js');
var documentation = require('../../lib/utils/documentation.js');
var getContainerNode = require('../../lib/utils/getContainerNode.js');

var withUsage = documentation.createDocumentationMessageGenerator({
    name: 'relevant-sort'
});
var suit = suit$1.component('RelevantSort');
var renderer = function renderer(param) {
    var containerNode = param.containerNode, cssClasses = param.cssClasses, templates = param.templates;
    return function(param) {
        var isRelevantSorted = param.isRelevantSorted, isVirtualReplica = param.isVirtualReplica, refine = param.refine;
        preact.render(/*#__PURE__*/ preact.h(RelevantSort.default, {
            cssClasses: cssClasses,
            templates: templates,
            isRelevantSorted: isRelevantSorted,
            isVirtualReplica: isVirtualReplica,
            refine: refine
        }), containerNode);
    };
};
var relevantSort = function relevantSort(widgetParams) {
    var container = widgetParams.container, tmp = widgetParams.templates, userTemplates = tmp === void 0 ? {} : tmp, tmp1 = widgetParams.cssClasses, userCssClasses = tmp1 === void 0 ? {} : tmp1;
    if (!container) {
        throw new Error(withUsage('The `container` option is required.'));
    }
    var containerNode = getContainerNode.getContainerNode(container);
    var cssClasses = {
        root: instantsearchUiComponents.cx(suit(), userCssClasses.root),
        text: instantsearchUiComponents.cx(suit({
            descendantName: 'text'
        }), userCssClasses.text),
        button: instantsearchUiComponents.cx(suit({
            descendantName: 'button'
        }), userCssClasses.button)
    };
    var templates = _object_spread._({}, defaultTemplates.default, userTemplates);
    var specializedRenderer = renderer({
        containerNode: containerNode,
        cssClasses: cssClasses,
        renderState: {},
        templates: templates
    });
    var makeWidget = connectRelevantSort.default(specializedRenderer, function() {
        preact.render(null, containerNode);
    });
    return _object_spread_props._(_object_spread._({}, makeWidget({})), {
        $$widgetType: 'ais.relevantSort'
    });
};

exports.default = relevantSort;
