// instantsearch.js@4.90.0/cjs/widgets/stats/stats.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _object_spread = require('@swc/helpers/_/_object_spread');
var _object_spread_props = require('@swc/helpers/_/_object_spread_props');
var instantsearchUiComponents = require('instantsearch-ui-components');
var preact = require('preact');
var Stats = require('../../components/Stats/Stats.js');
var connectStats = require('../../connectors/stats/connectStats.js');
var formatNumber = require('../../lib/formatNumber.js');
var suit$1 = require('../../lib/suit.js');
var documentation = require('../../lib/utils/documentation.js');
var getContainerNode = require('../../lib/utils/getContainerNode.js');
var prepareTemplateProps = require('../../lib/templating/prepareTemplateProps.js');

var withUsage = documentation.createDocumentationMessageGenerator({
    name: 'stats'
});
var suit = suit$1.component('Stats');
var defaultTemplates = {
    text: function text(props) {
        return "".concat(props.areHitsSorted ? getSortedResultsSentence(props) : getResultsSentence(props), " found in ").concat(props.processingTimeMS, "ms");
    }
};
function getSortedResultsSentence(param) {
    var nbHits = param.nbHits, hasNoSortedResults = param.hasNoSortedResults, hasOneSortedResults = param.hasOneSortedResults, hasManySortedResults = param.hasManySortedResults, nbSortedHits = param.nbSortedHits;
    var suffix = "sorted out of ".concat(formatNumber.formatNumber(nbHits));
    if (hasNoSortedResults) {
        return "No relevant results ".concat(suffix);
    }
    if (hasOneSortedResults) {
        return "1 relevant result ".concat(suffix);
    }
    if (hasManySortedResults) {
        return "".concat(formatNumber.formatNumber(nbSortedHits || 0), " relevant results ").concat(suffix);
    }
    return '';
}
function getResultsSentence(param) {
    var nbHits = param.nbHits, hasNoResults = param.hasNoResults, hasOneResult = param.hasOneResult, hasManyResults = param.hasManyResults;
    if (hasNoResults) {
        return 'No results';
    }
    if (hasOneResult) {
        return '1 result';
    }
    if (hasManyResults) {
        return "".concat(formatNumber.formatNumber(nbHits), " results");
    }
    return '';
}
var renderer = function renderer(param) {
    var renderState = param.renderState, cssClasses = param.cssClasses, containerNode = param.containerNode, templates = param.templates;
    return function(param, isFirstRendering) {
        var hitsPerPage = param.hitsPerPage, nbHits = param.nbHits, nbSortedHits = param.nbSortedHits, areHitsSorted = param.areHitsSorted, nbPages = param.nbPages, page = param.page, processingTimeMS = param.processingTimeMS, query = param.query, instantSearchInstance = param.instantSearchInstance;
        if (isFirstRendering) {
            renderState.templateProps = prepareTemplateProps.prepareTemplateProps({
                defaultTemplates: defaultTemplates,
                templatesConfig: instantSearchInstance.templatesConfig,
                templates: templates
            });
            return;
        }
        preact.render(/*#__PURE__*/ preact.h(Stats.default, {
            cssClasses: cssClasses,
            hitsPerPage: hitsPerPage,
            nbHits: nbHits,
            nbSortedHits: nbSortedHits,
            areHitsSorted: areHitsSorted,
            nbPages: nbPages,
            page: page,
            processingTimeMS: processingTimeMS,
            query: query,
            templateProps: renderState.templateProps
        }), containerNode);
    };
};
/**
 * The `stats` widget is used to display useful insights about the current results.
 *
 * By default, it will display the **number of hits** and the time taken to compute the
 * results inside the engine.
 */ var stats = function stats(widgetParams) {
    var _ref = widgetParams || {}, container = _ref.container, tmp = _ref.cssClasses, userCssClasses = tmp === void 0 ? {} : tmp, _ref_templates = _ref.templates, templates = _ref_templates === void 0 ? {} : _ref_templates;
    if (!container) {
        throw new Error(withUsage('The `container` option is required.'));
    }
    var containerNode = getContainerNode.getContainerNode(container);
    var cssClasses = {
        root: instantsearchUiComponents.cx(suit(), userCssClasses.root),
        text: instantsearchUiComponents.cx(suit({
            descendantName: 'text'
        }), userCssClasses.text)
    };
    var specializedRenderer = renderer({
        containerNode: containerNode,
        cssClasses: cssClasses,
        templates: templates,
        renderState: {}
    });
    var makeWidget = connectStats.default(specializedRenderer, function() {
        return preact.render(null, containerNode);
    });
    return _object_spread_props._(_object_spread._({}, makeWidget({})), {
        $$widgetType: 'ais.stats'
    });
};

exports.default = stats;
exports.defaultTemplates = defaultTemplates;
