// instantsearch.js@4.90.0/cjs/widgets/infinite-hits/infinite-hits.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _object_spread = require('@swc/helpers/_/_object_spread');
var _object_spread_props = require('@swc/helpers/_/_object_spread_props');
var instantsearchUiComponents = require('instantsearch-ui-components');
var preact = require('preact');
var InfiniteHits = require('../../components/InfiniteHits/InfiniteHits.js');
var connectInfiniteHits = require('../../connectors/infinite-hits/connectInfiniteHits.js');
var suit$1 = require('../../lib/suit.js');
var defaultTemplates = require('./defaultTemplates.js');
var client = require('../../lib/insights/client.js');
var documentation = require('../../lib/utils/documentation.js');
var getContainerNode = require('../../lib/utils/getContainerNode.js');
var prepareTemplateProps = require('../../lib/templating/prepareTemplateProps.js');

var withUsage = documentation.createDocumentationMessageGenerator({
    name: 'infinite-hits'
});
var suit = suit$1.component('InfiniteHits');
var renderer = function renderer(param) {
    var containerNode = param.containerNode, cssClasses = param.cssClasses, renderState = param.renderState, templates = param.templates, hasShowPrevious = param.showPrevious;
    return function(param, isFirstRendering) {
        var items = param.items, results = param.results, showMore = param.showMore, showPrevious = param.showPrevious, isFirstPage = param.isFirstPage, isLastPage = param.isLastPage, instantSearchInstance = param.instantSearchInstance, insights = param.insights, bindEvent = param.bindEvent, sendEvent = param.sendEvent, banner = param.banner;
        if (isFirstRendering) {
            renderState.templateProps = prepareTemplateProps.prepareTemplateProps({
                defaultTemplates: defaultTemplates.default,
                templatesConfig: instantSearchInstance.templatesConfig,
                templates: templates
            });
            return;
        }
        preact.render(/*#__PURE__*/ preact.h(InfiniteHits.default, {
            cssClasses: cssClasses,
            hits: items,
            results: results,
            hasShowPrevious: hasShowPrevious,
            showPrevious: showPrevious,
            showMore: showMore,
            templateProps: renderState.templateProps,
            isFirstPage: isFirstPage,
            isLastPage: isLastPage,
            insights: insights,
            sendEvent: sendEvent,
            bindEvent: bindEvent,
            banner: banner
        }), containerNode);
    };
};
var infiniteHits = (function infiniteHits(widgetParams) {
    var _ref = widgetParams || {}, container = _ref.container, escapeHTML = _ref.escapeHTML, transformItems = _ref.transformItems, _ref_templates = _ref.templates, templates = _ref_templates === void 0 ? {} : _ref_templates, tmp = _ref.cssClasses, userCssClasses = tmp === void 0 ? {} : tmp, showPrevious = _ref.showPrevious, cache = _ref.cache;
    if (!container) {
        throw new Error(withUsage('The `container` option is required.'));
    }
    var containerNode = getContainerNode.getContainerNode(container);
    var cssClasses = {
        root: instantsearchUiComponents.cx(suit(), userCssClasses.root),
        emptyRoot: instantsearchUiComponents.cx(suit({
            modifierName: 'empty'
        }), userCssClasses.emptyRoot),
        item: instantsearchUiComponents.cx(suit({
            descendantName: 'item'
        }), userCssClasses.item),
        list: instantsearchUiComponents.cx(suit({
            descendantName: 'list'
        }), userCssClasses.list),
        loadPrevious: instantsearchUiComponents.cx(suit({
            descendantName: 'loadPrevious'
        }), userCssClasses.loadPrevious),
        disabledLoadPrevious: instantsearchUiComponents.cx(suit({
            descendantName: 'loadPrevious',
            modifierName: 'disabled'
        }), userCssClasses.disabledLoadPrevious),
        loadMore: instantsearchUiComponents.cx(suit({
            descendantName: 'loadMore'
        }), userCssClasses.loadMore),
        disabledLoadMore: instantsearchUiComponents.cx(suit({
            descendantName: 'loadMore',
            modifierName: 'disabled'
        }), userCssClasses.disabledLoadMore),
        bannerRoot: instantsearchUiComponents.cx(suit({
            descendantName: 'banner'
        }), userCssClasses.bannerRoot),
        bannerImage: instantsearchUiComponents.cx(suit({
            descendantName: 'banner-image'
        }), userCssClasses.bannerImage),
        bannerLink: instantsearchUiComponents.cx(suit({
            descendantName: 'banner-link'
        }), userCssClasses.bannerLink)
    };
    var specializedRenderer = renderer({
        containerNode: containerNode,
        cssClasses: cssClasses,
        templates: templates,
        showPrevious: showPrevious,
        renderState: {}
    });
    var makeWidget = client.default(connectInfiniteHits.default)(specializedRenderer, function() {
        return preact.render(null, containerNode);
    });
    return _object_spread_props._(_object_spread._({}, makeWidget({
        escapeHTML: escapeHTML,
        transformItems: transformItems,
        showPrevious: showPrevious,
        cache: cache
    })), {
        $$widgetType: 'ais.infiniteHits'
    });
});

exports.default = infiniteHits;
