// instantsearch.js@4.90.0/cjs/connectors/index.umd.js
'use strict';

var connectAnswers = require('./answers/connectAnswers.js');
var connectConfigureRelatedItems = require('./configure-related-items/connectConfigureRelatedItems.js');
var connectDynamicWidgets = require('./dynamic-widgets/connectDynamicWidgets.js');
var connectClearRefinements = require('./clear-refinements/connectClearRefinements.js');
var connectCurrentRefinements = require('./current-refinements/connectCurrentRefinements.js');
var connectHierarchicalMenu = require('./hierarchical-menu/connectHierarchicalMenu.js');
var connectHits = require('./hits/connectHits.js');
var connectHitsWithInsights = require('./hits/connectHitsWithInsights.js');
var connectHitsPerPage = require('./hits-per-page/connectHitsPerPage.js');
var connectInfiniteHits = require('./infinite-hits/connectInfiniteHits.js');
var connectInfiniteHitsWithInsights = require('./infinite-hits/connectInfiniteHitsWithInsights.js');
var connectMenu = require('./menu/connectMenu.js');
var connectNumericMenu = require('./numeric-menu/connectNumericMenu.js');
var connectPagination = require('./pagination/connectPagination.js');
var connectRange = require('./range/connectRange.js');
var connectRefinementList = require('./refinement-list/connectRefinementList.js');
var connectRelatedProducts = require('./related-products/connectRelatedProducts.js');
var connectSearchBox = require('./search-box/connectSearchBox.js');
var connectSortBy = require('./sort-by/connectSortBy.js');
var connectRatingMenu = require('./rating-menu/connectRatingMenu.js');
var connectStats = require('./stats/connectStats.js');
var connectToggleRefinement = require('./toggle-refinement/connectToggleRefinement.js');
var connectTrendingItems = require('./trending-items/connectTrendingItems.js');
var connectBreadcrumb = require('./breadcrumb/connectBreadcrumb.js');
var connectGeoSearch = require('./geo-search/connectGeoSearch.js');
var connectPoweredBy = require('./powered-by/connectPoweredBy.js');
var connectConfigure = require('./configure/connectConfigure.js');
var connectAutocomplete = require('./autocomplete/connectAutocomplete.js');
var connectQueryRules = require('./query-rules/connectQueryRules.js');
var connectVoiceSearch = require('./voice-search/connectVoiceSearch.js');
var connectRelevantSort = require('./relevant-sort/connectRelevantSort.js');
var connectFrequentlyBoughtTogether = require('./frequently-bought-together/connectFrequentlyBoughtTogether.js');
var connectLookingSimilar = require('./looking-similar/connectLookingSimilar.js');
var connectFilterSuggestions = require('./filter-suggestions/connectFilterSuggestions.js');
var logger = require('../lib/utils/logger.js');

/** @deprecated answers is no longer supported */ var EXPERIMENTAL_connectAnswers = logger.deprecate(connectAnswers.default);
/** @deprecated use connectRelatedItems instead */ var EXPERIMENTAL_connectConfigureRelatedItems = logger.deprecate(connectConfigureRelatedItems.default);
/** @deprecated use connectDynamicWidgets */ var EXPERIMENTAL_connectDynamicWidgets = logger.deprecate(connectDynamicWidgets.default);
var connectChat = function connectChat() {
    throw new Error('"connectChat" is not available from the UMD build.\n\nPlease use InstantSearch.js with a packaging system:\nhttps://www.algolia.com/doc/guides/building-search-ui/installation/js/#with-a-packaging-system');
};

exports.connectDynamicWidgets = connectDynamicWidgets.default;
exports.connectClearRefinements = connectClearRefinements.default;
exports.connectCurrentRefinements = connectCurrentRefinements.default;
exports.connectHierarchicalMenu = connectHierarchicalMenu.default;
exports.connectHits = connectHits.default;
exports.connectHitsWithInsights = connectHitsWithInsights.default;
exports.connectHitsPerPage = connectHitsPerPage.default;
exports.connectInfiniteHits = connectInfiniteHits.default;
exports.connectInfiniteHitsWithInsights = connectInfiniteHitsWithInsights.default;
exports.connectMenu = connectMenu.default;
exports.connectNumericMenu = connectNumericMenu.default;
exports.connectPagination = connectPagination.default;
exports.connectRange = connectRange.default;
exports.connectRefinementList = connectRefinementList.default;
exports.connectRelatedProducts = connectRelatedProducts.default;
exports.connectSearchBox = connectSearchBox.default;
exports.connectSortBy = connectSortBy.default;
exports.connectRatingMenu = connectRatingMenu.default;
exports.connectStats = connectStats.default;
exports.connectToggleRefinement = connectToggleRefinement.default;
exports.connectTrendingItems = connectTrendingItems.default;
exports.connectBreadcrumb = connectBreadcrumb.default;
exports.connectGeoSearch = connectGeoSearch.default;
exports.connectPoweredBy = connectPoweredBy.default;
exports.connectConfigure = connectConfigure.default;
exports.connectAutocomplete = connectAutocomplete.default;
exports.connectQueryRules = connectQueryRules.default;
exports.connectVoiceSearch = connectVoiceSearch.default;
exports.connectRelevantSort = connectRelevantSort.default;
exports.connectFrequentlyBoughtTogether = connectFrequentlyBoughtTogether.default;
exports.connectLookingSimilar = connectLookingSimilar.default;
exports.connectFilterSuggestions = connectFilterSuggestions.default;
exports.EXPERIMENTAL_connectAnswers = EXPERIMENTAL_connectAnswers;
exports.EXPERIMENTAL_connectConfigureRelatedItems = EXPERIMENTAL_connectConfigureRelatedItems;
exports.EXPERIMENTAL_connectDynamicWidgets = EXPERIMENTAL_connectDynamicWidgets;
exports.connectChat = connectChat;
