// instantsearch.js@4.90.0/cjs/widgets/index.umd.js
'use strict';

var answers = require('./answers/answers.js');
var configureRelatedItems = require('./configure-related-items/configure-related-items.js');
var dynamicWidgets = require('./dynamic-widgets/dynamic-widgets.js');
var analytics = require('./analytics/analytics.js');
var autocomplete = require('./autocomplete/autocomplete.js');
var breadcrumb = require('./breadcrumb/breadcrumb.js');
var clearRefinements = require('./clear-refinements/clear-refinements.js');
var configure = require('./configure/configure.js');
var currentRefinements = require('./current-refinements/current-refinements.js');
var geoSearch = require('./geo-search/geo-search.js');
var hierarchicalMenu = require('./hierarchical-menu/hierarchical-menu.js');
var hits = require('./hits/hits.js');
var hitsPerPage = require('./hits-per-page/hits-per-page.js');
var index = require('./index/index.js');
var infiniteHits = require('./infinite-hits/infinite-hits.js');
var menu = require('./menu/menu.js');
var menuSelect = require('./menu-select/menu-select.js');
var numericMenu = require('./numeric-menu/numeric-menu.js');
var pagination = require('./pagination/pagination.js');
var panel = require('./panel/panel.js');
var places = require('./places/places.js');
var poweredBy = require('./powered-by/powered-by.js');
var queryRuleContext = require('./query-rule-context/query-rule-context.js');
var queryRuleCustomData = require('./query-rule-custom-data/query-rule-custom-data.js');
var relatedProducts = require('./related-products/related-products.js');
var rangeInput = require('./range-input/range-input.js');
var rangeSlider = require('./range-slider/range-slider.js');
var ratingMenu = require('./rating-menu/rating-menu.js');
var refinementList = require('./refinement-list/refinement-list.js');
var relevantSort = require('./relevant-sort/relevant-sort.js');
var searchBox = require('./search-box/search-box.js');
var sortBy = require('./sort-by/sort-by.js');
var stats = require('./stats/stats.js');
var toggleRefinement = require('./toggle-refinement/toggle-refinement.js');
var trendingItems = require('./trending-items/trending-items.js');
var voiceSearch = require('./voice-search/voice-search.js');
var frequentlyBoughtTogether = require('./frequently-bought-together/frequently-bought-together.js');
var lookingSimilar = require('./looking-similar/looking-similar.js');
var filterSuggestions = require('./filter-suggestions/filter-suggestions.js');
var logger = require('../lib/utils/logger.js');

/** @deprecated answers is no longer supported */ var EXPERIMENTAL_answers = logger.deprecate(answers.default);
/** @deprecated use relatedItems instead */ var EXPERIMENTAL_configureRelatedItems = logger.deprecate(configureRelatedItems.default);
/** @deprecated use dynamicWidgets */ var EXPERIMENTAL_dynamicWidgets = logger.deprecate(dynamicWidgets.default);
var chat = function chat() {
    throw new Error('"chat" is not available from the UMD build.\n\nPlease use InstantSearch.js with a packaging system:\nhttps://www.algolia.com/doc/guides/building-search-ui/installation/js/#with-a-packaging-system');
};

exports.dynamicWidgets = dynamicWidgets.default;
exports.analytics = analytics.default;
exports.EXPERIMENTAL_autocomplete = autocomplete.EXPERIMENTAL_autocomplete;
exports.breadcrumb = breadcrumb.default;
exports.clearRefinements = clearRefinements.default;
exports.configure = configure.default;
exports.currentRefinements = currentRefinements.default;
exports.geoSearch = geoSearch.default;
exports.hierarchicalMenu = hierarchicalMenu.default;
exports.hits = hits.default;
exports.hitsPerPage = hitsPerPage.default;
exports.index = index.default;
exports.infiniteHits = infiniteHits.default;
exports.menu = menu.default;
exports.menuSelect = menuSelect.default;
exports.numericMenu = numericMenu.default;
exports.pagination = pagination.default;
exports.panel = panel.default;
exports.places = places.default;
exports.poweredBy = poweredBy.default;
exports.queryRuleContext = queryRuleContext.default;
exports.queryRuleCustomData = queryRuleCustomData.default;
exports.relatedProducts = relatedProducts.default;
exports.rangeInput = rangeInput.default;
exports.rangeSlider = rangeSlider.default;
exports.ratingMenu = ratingMenu.default;
exports.refinementList = refinementList.default;
exports.relevantSort = relevantSort.default;
exports.searchBox = searchBox.default;
exports.sortBy = sortBy.default;
exports.stats = stats.default;
exports.toggleRefinement = toggleRefinement.default;
exports.trendingItems = trendingItems.default;
exports.voiceSearch = voiceSearch.default;
exports.frequentlyBoughtTogether = frequentlyBoughtTogether.default;
exports.lookingSimilar = lookingSimilar.default;
exports.filterSuggestions = filterSuggestions.default;
exports.EXPERIMENTAL_answers = EXPERIMENTAL_answers;
exports.EXPERIMENTAL_configureRelatedItems = EXPERIMENTAL_configureRelatedItems;
exports.EXPERIMENTAL_dynamicWidgets = EXPERIMENTAL_dynamicWidgets;
exports.chat = chat;
