// instantsearch.js@4.90.0/cjs/widgets/breadcrumb/breadcrumb.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _object_spread = require('@swc/helpers/_/_object_spread');
var _object_spread_props = require('@swc/helpers/_/_object_spread_props');
var instantsearchUiComponents = require('instantsearch-ui-components');
var preact = require('preact');
var Breadcrumb = require('../../components/Breadcrumb/Breadcrumb.js');
var connectBreadcrumb = require('../../connectors/breadcrumb/connectBreadcrumb.js');
var suit$1 = require('../../lib/suit.js');
var defaultTemplates = require('./defaultTemplates.js');
var documentation = require('../../lib/utils/documentation.js');
var getContainerNode = require('../../lib/utils/getContainerNode.js');
var prepareTemplateProps = require('../../lib/templating/prepareTemplateProps.js');

var withUsage = documentation.createDocumentationMessageGenerator({
    name: 'breadcrumb'
});
var suit = suit$1.component('Breadcrumb');
var renderer = function renderer(param) {
    var containerNode = param.containerNode, cssClasses = param.cssClasses, renderState = param.renderState, templates = param.templates;
    return function(param, isFirstRendering) {
        var canRefine = param.canRefine, createURL = param.createURL, instantSearchInstance = param.instantSearchInstance, items = param.items, refine = param.refine;
        if (isFirstRendering) {
            renderState.templateProps = prepareTemplateProps.prepareTemplateProps({
                defaultTemplates: defaultTemplates.default,
                templatesConfig: instantSearchInstance.templatesConfig,
                templates: templates
            });
            return;
        }
        preact.render(/*#__PURE__*/ preact.h(Breadcrumb.default, {
            canRefine: canRefine,
            cssClasses: cssClasses,
            createURL: createURL,
            items: items,
            refine: refine,
            templateProps: renderState.templateProps
        }), containerNode);
    };
};
var breadcrumb = function breadcrumb(widgetParams) {
    var _ref = widgetParams || {}, container = _ref.container, attributes = _ref.attributes, separator = _ref.separator, rootPath = _ref.rootPath, transformItems = _ref.transformItems, _ref_templates = _ref.templates, templates = _ref_templates === void 0 ? {} : _ref_templates, tmp = _ref.cssClasses, userCssClasses = tmp === void 0 ? {} : tmp;
    if (!container) {
        throw new Error(withUsage('The `container` option is required.'));
    }
    var containerNode = getContainerNode.getContainerNode(container);
    var cssClasses = {
        root: instantsearchUiComponents.cx(suit(), userCssClasses.root),
        noRefinementRoot: instantsearchUiComponents.cx(suit({
            modifierName: 'noRefinement'
        }), userCssClasses.noRefinementRoot),
        list: instantsearchUiComponents.cx(suit({
            descendantName: 'list'
        }), userCssClasses.list),
        item: instantsearchUiComponents.cx(suit({
            descendantName: 'item'
        }), userCssClasses.item),
        selectedItem: instantsearchUiComponents.cx(suit({
            descendantName: 'item',
            modifierName: 'selected'
        }), userCssClasses.selectedItem),
        separator: instantsearchUiComponents.cx(suit({
            descendantName: 'separator'
        }), userCssClasses.separator),
        link: instantsearchUiComponents.cx(suit({
            descendantName: 'link'
        }), userCssClasses.link)
    };
    var specializedRenderer = renderer({
        containerNode: containerNode,
        cssClasses: cssClasses,
        renderState: {},
        templates: templates
    });
    var makeWidget = connectBreadcrumb.default(specializedRenderer, function() {
        return preact.render(null, containerNode);
    });
    return _object_spread_props._(_object_spread._({}, makeWidget({
        attributes: attributes,
        separator: separator,
        rootPath: rootPath,
        transformItems: transformItems
    })), {
        $$widgetType: 'ais.breadcrumb'
    });
};

exports.default = breadcrumb;
