// instantsearch-ui-components@0.20.0/dist/cjs/components/index.js
'use strict';

var Button = require('./Button.js');
var Carousel = require('./Carousel.js');
var Chat = require('./chat/Chat.js');
var ChatHeader = require('./chat/ChatHeader.js');
var ChatMessage = require('./chat/ChatMessage.js');
var ChatMessages = require('./chat/ChatMessages.js');
var ChatMessageLoader = require('./chat/ChatMessageLoader.js');
var ChatMessageError = require('./chat/ChatMessageError.js');
var ChatPrompt = require('./chat/ChatPrompt.js');
var ChatPromptSuggestions = require('./chat/ChatPromptSuggestions.js');
var ChatToggleButton = require('./chat/ChatToggleButton.js');
var icons = require('./chat/icons.js');
var FrequentlyBoughtTogether = require('./FrequentlyBoughtTogether.js');
var Highlight = require('./Highlight.js');
var Hits = require('./Hits.js');
var LookingSimilar = require('./LookingSimilar.js');
var RelatedProducts = require('./RelatedProducts.js');
var TrendingItems = require('./TrendingItems.js');
var FilterSuggestions = require('./FilterSuggestions.js');
var Autocomplete = require('./autocomplete/Autocomplete.js');
var AutocompleteDetachedContainer = require('./autocomplete/AutocompleteDetachedContainer.js');
var AutocompleteDetachedFormContainer = require('./autocomplete/AutocompleteDetachedFormContainer.js');
var AutocompleteDetachedOverlay = require('./autocomplete/AutocompleteDetachedOverlay.js');
var AutocompleteDetachedSearchButton = require('./autocomplete/AutocompleteDetachedSearchButton.js');
var AutocompleteIndex = require('./autocomplete/AutocompleteIndex.js');
var AutocompletePanel = require('./autocomplete/AutocompletePanel.js');
var AutocompleteRecentSearch = require('./autocomplete/AutocompleteRecentSearch.js');
var AutocompleteSearch = require('./autocomplete/AutocompleteSearch.js');
var AutocompleteSuggestion = require('./autocomplete/AutocompleteSuggestion.js');
var createAutocompletePropGetters = require('./autocomplete/createAutocompletePropGetters.js');
var createAutocompleteStorage = require('./autocomplete/createAutocompleteStorage.js');



exports.createButtonComponent = Button.createButtonComponent;
exports.createCarouselComponent = Carousel.createCarouselComponent;
exports.generateCarouselId = Carousel.generateCarouselId;
exports.createChatComponent = Chat.createChatComponent;
exports.createChatHeaderComponent = ChatHeader.createChatHeaderComponent;
exports.createChatMessageComponent = ChatMessage.createChatMessageComponent;
exports.createChatMessagesComponent = ChatMessages.createChatMessagesComponent;
exports.createChatMessageLoaderComponent = ChatMessageLoader.createChatMessageLoaderComponent;
exports.createChatMessageErrorComponent = ChatMessageError.createChatMessageErrorComponent;
exports.createChatPromptComponent = ChatPrompt.createChatPromptComponent;
exports.createChatPromptSuggestionsComponent = ChatPromptSuggestions.createChatPromptSuggestionsComponent;
exports.createChatToggleButtonComponent = ChatToggleButton.createChatToggleButtonComponent;
exports.ArrowRightIcon = icons.ArrowRightIcon;
exports.ArrowUpIcon = icons.ArrowUpIcon;
exports.ChevronDownIcon = icons.ChevronDownIcon;
exports.ChevronLeftIcon = icons.ChevronLeftIcon;
exports.ChevronRightIcon = icons.ChevronRightIcon;
exports.ChevronUpIcon = icons.ChevronUpIcon;
exports.CloseIcon = icons.CloseIcon;
exports.CopyIcon = icons.CopyIcon;
exports.LoadingSpinnerIcon = icons.LoadingSpinnerIcon;
exports.MaximizeIcon = icons.MaximizeIcon;
exports.MenuIcon = icons.MenuIcon;
exports.MinimizeIcon = icons.MinimizeIcon;
exports.ReloadIcon = icons.ReloadIcon;
exports.SparklesIcon = icons.SparklesIcon;
exports.StopIcon = icons.StopIcon;
exports.createFrequentlyBoughtTogetherComponent = FrequentlyBoughtTogether.createFrequentlyBoughtTogetherComponent;
exports.createHighlightComponent = Highlight.createHighlightComponent;
exports.createHitsComponent = Hits.createHitsComponent;
exports.createLookingSimilarComponent = LookingSimilar.createLookingSimilarComponent;
exports.createRelatedProductsComponent = RelatedProducts.createRelatedProductsComponent;
exports.createTrendingItemsComponent = TrendingItems.createTrendingItemsComponent;
exports.createFilterSuggestionsComponent = FilterSuggestions.createFilterSuggestionsComponent;
exports.createAutocompleteComponent = Autocomplete.createAutocompleteComponent;
exports.createAutocompleteDetachedContainerComponent = AutocompleteDetachedContainer.createAutocompleteDetachedContainerComponent;
exports.createAutocompleteDetachedFormContainerComponent = AutocompleteDetachedFormContainer.createAutocompleteDetachedFormContainerComponent;
exports.createAutocompleteDetachedOverlayComponent = AutocompleteDetachedOverlay.createAutocompleteDetachedOverlayComponent;
exports.createAutocompleteDetachedSearchButtonComponent = AutocompleteDetachedSearchButton.createAutocompleteDetachedSearchButtonComponent;
exports.createAutocompleteIndexComponent = AutocompleteIndex.createAutocompleteIndexComponent;
exports.createAutocompletePanelComponent = AutocompletePanel.createAutocompletePanelComponent;
exports.createAutocompleteRecentSearchComponent = AutocompleteRecentSearch.createAutocompleteRecentSearchComponent;
exports.createAutocompleteSearchComponent = AutocompleteSearch.createAutocompleteSearchComponent;
exports.createAutocompleteSuggestionComponent = AutocompleteSuggestion.createAutocompleteSuggestionComponent;
exports.createAutocompletePropGetters = createAutocompletePropGetters.createAutocompletePropGetters;
exports.createAutocompleteStorage = createAutocompleteStorage.createAutocompleteStorage;
exports.createStorage = createAutocompleteStorage.createStorage;
