// instantsearch-ui-components@0.20.0/dist/cjs/components/RelatedProducts.js
'use strict';

var _object_spread = require('@swc/helpers/_/_object_spread');
var _object_spread_props = require('@swc/helpers/_/_object_spread_props');
var _object_without_properties = require('@swc/helpers/_/_object_without_properties');
var DefaultEmpty = require('./recommend-shared/DefaultEmpty.js');
var DefaultHeader = require('./recommend-shared/DefaultHeader.js');
var DefaultItem = require('./recommend-shared/DefaultItem.js');
var List = require('./recommend-shared/List.js');
var cx = require('../lib/cx.js');

function createRelatedProductsComponent(param) {
    var createElement = param.createElement, Fragment = param.Fragment;
    return function RelatedProducts(userProps) {
        var _userProps_classNames = userProps.classNames, classNames = _userProps_classNames === void 0 ? {} : _userProps_classNames, tmp = userProps.emptyComponent, EmptyComponent = tmp === void 0 ? DefaultEmpty.createDefaultEmptyComponent({
            createElement: createElement,
            Fragment: Fragment
        }) : tmp, tmp1 = userProps.headerComponent, HeaderComponent = tmp1 === void 0 ? DefaultHeader.createDefaultHeaderComponent({
            createElement: createElement,
            Fragment: Fragment
        }) : tmp1, tmp2 = userProps.itemComponent, ItemComponent = tmp2 === void 0 ? DefaultItem.createDefaultItemComponent({
            createElement: createElement,
            Fragment: Fragment
        }) : tmp2, tmp3 = userProps.layout, Layout = tmp3 === void 0 ? List.createListComponent({
            createElement: createElement,
            Fragment: Fragment
        }) : tmp3, items = userProps.items, status = userProps.status, userTranslations = userProps.translations, sendEvent = userProps.sendEvent, props = _object_without_properties._(userProps, [
            "classNames",
            "emptyComponent",
            "headerComponent",
            "itemComponent",
            "layout",
            "items",
            "status",
            "translations",
            "sendEvent"
        ]);
        var translations = _object_spread._({
            title: 'Related products',
            sliderLabel: 'Related products'
        }, userTranslations);
        var cssClasses = {
            root: cx.cx('ais-RelatedProducts', classNames.root),
            emptyRoot: cx.cx('ais-RelatedProducts', classNames.root, 'ais-RelatedProducts--empty', classNames.emptyRoot, props.className),
            title: cx.cx('ais-RelatedProducts-title', classNames.title),
            container: cx.cx('ais-RelatedProducts-container', classNames.container),
            list: cx.cx('ais-RelatedProducts-list', classNames.list),
            item: cx.cx('ais-RelatedProducts-item', classNames.item)
        };
        if (items.length === 0 && status === 'idle') {
            return /*#__PURE__*/ createElement("section", _object_spread_props._(_object_spread._({}, props), {
                className: cssClasses.emptyRoot
            }), /*#__PURE__*/ createElement(EmptyComponent, null));
        }
        return /*#__PURE__*/ createElement("section", _object_spread_props._(_object_spread._({}, props), {
            className: cssClasses.root
        }), /*#__PURE__*/ createElement(HeaderComponent, {
            classNames: cssClasses,
            items: items,
            translations: translations
        }), /*#__PURE__*/ createElement(Layout, {
            classNames: cssClasses,
            itemComponent: ItemComponent,
            items: items,
            sendEvent: sendEvent
        }));
    };
}

exports.createRelatedProductsComponent = createRelatedProductsComponent;
