// instantsearch.js@4.90.0/cjs/widgets/query-rule-custom-data/query-rule-custom-data.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _object_spread = require('@swc/helpers/_/_object_spread');
var _object_spread_props = require('@swc/helpers/_/_object_spread_props');
var instantsearchUiComponents = require('instantsearch-ui-components');
var preact = require('preact');
var QueryRuleCustomData = require('../../components/QueryRuleCustomData/QueryRuleCustomData.js');
var connectQueryRules = require('../../connectors/query-rules/connectQueryRules.js');
var suit$1 = require('../../lib/suit.js');
var documentation = require('../../lib/utils/documentation.js');
var getContainerNode = require('../../lib/utils/getContainerNode.js');

var defaultTemplates = {
    default: function _default(param) {
        var items = param.items;
        return JSON.stringify(items, null, 2);
    }
};
var withUsage = documentation.createDocumentationMessageGenerator({
    name: 'query-rule-custom-data'
});
var suit = suit$1.component('QueryRuleCustomData');
var renderer = function renderer(param) {
    var containerNode = param.containerNode, cssClasses = param.cssClasses, templates = param.templates;
    return function(param) {
        var items = param.items;
        preact.render(/*#__PURE__*/ preact.h(QueryRuleCustomData.default, {
            cssClasses: cssClasses,
            templates: templates,
            items: items
        }), containerNode);
    };
};
var queryRuleCustomData = function queryRuleCustomData(widgetParams) {
    var _ref = widgetParams || {}, container = _ref.container, tmp = _ref.cssClasses, userCssClasses = tmp === void 0 ? {} : tmp, tmp1 = _ref.templates, userTemplates = tmp1 === void 0 ? {} : tmp1, _ref_transformItems = _ref.transformItems, transformItems = _ref_transformItems === void 0 ? function(items) {
        return items;
    } : _ref_transformItems;
    if (!container) {
        throw new Error(withUsage('The `container` option is required.'));
    }
    var cssClasses = {
        root: instantsearchUiComponents.cx(suit(), userCssClasses.root)
    };
    var containerNode = getContainerNode.getContainerNode(container);
    var templates = _object_spread._({}, defaultTemplates, userTemplates);
    var specializedRenderer = renderer({
        containerNode: containerNode,
        cssClasses: cssClasses,
        renderState: {},
        templates: templates
    });
    var makeWidget = connectQueryRules.default(specializedRenderer, function() {
        preact.render(null, containerNode);
    });
    return _object_spread_props._(_object_spread._({}, makeWidget({
        transformItems: transformItems
    })), {
        $$widgetType: 'ais.queryRuleCustomData'
    });
};

exports.default = queryRuleCustomData;
exports.defaultTemplates = defaultTemplates;
