// react-instantsearch@7.26.0/dist/cjs/widgets/Breadcrumb.js
'use strict';

var _object_spread = require('@swc/helpers/_/_object_spread');
var _object_without_properties = require('@swc/helpers/_/_object_without_properties');
var React = require('react');
var reactInstantsearchCore = require('react-instantsearch-core');
var Breadcrumb$1 = require('../ui/Breadcrumb.js');

function Breadcrumb(_0) {
    var attributes = _0.attributes, rootPath = _0.rootPath; _0.separator; var transformItems = _0.transformItems, translations = _0.translations, props = _object_without_properties._(_0, [
        "attributes",
        "rootPath",
        "separator",
        "transformItems",
        "translations"
    ]);
    var _useBreadcrumb = reactInstantsearchCore.useBreadcrumb({
        attributes: attributes,
        rootPath: rootPath,
        transformItems: transformItems
    }, {
        $$widgetType: 'ais.breadcrumb'
    }), canRefine = _useBreadcrumb.canRefine, createURL = _useBreadcrumb.createURL, items = _useBreadcrumb.items, refine = _useBreadcrumb.refine;
    var uiProps = {
        items: items,
        hasItems: canRefine,
        createURL: createURL,
        onNavigate: refine,
        translations: _object_spread._({
            rootElementText: 'Home'
        }, translations)
    };
    return /*#__PURE__*/ React.createElement(Breadcrumb$1.Breadcrumb, _object_spread._({}, props, uiProps));
}

exports.Breadcrumb = Breadcrumb;
