// react-instantsearch@7.26.0/dist/cjs/widgets/ClearRefinements.js
'use strict';

var _object_spread = require('@swc/helpers/_/_object_spread');
var _object_without_properties = require('@swc/helpers/_/_object_without_properties');
var React = require('react');
var reactInstantsearchCore = require('react-instantsearch-core');
var ClearRefinements$1 = require('../ui/ClearRefinements.js');

function ClearRefinements(_0) {
    var includedAttributes = _0.includedAttributes, excludedAttributes = _0.excludedAttributes, transformItems = _0.transformItems, translations = _0.translations, props = _object_without_properties._(_0, [
        "includedAttributes",
        "excludedAttributes",
        "transformItems",
        "translations"
    ]);
    var _useClearRefinements = reactInstantsearchCore.useClearRefinements({
        includedAttributes: includedAttributes,
        excludedAttributes: excludedAttributes,
        transformItems: transformItems
    }, {
        $$widgetType: 'ais.clearRefinements'
    }), canRefine = _useClearRefinements.canRefine, refine = _useClearRefinements.refine;
    var uiProps = {
        onClick: refine,
        disabled: !canRefine,
        translations: _object_spread._({
            resetButtonText: 'Clear refinements'
        }, translations)
    };
    return /*#__PURE__*/ React.createElement(ClearRefinements$1.ClearRefinements, _object_spread._({}, props, uiProps));
}

exports.ClearRefinements = ClearRefinements;
