// react-instantsearch@7.26.0/dist/cjs/widgets/index.js
'use strict';

var Autocomplete = require('./Autocomplete.js');
var Breadcrumb = require('./Breadcrumb.js');
var Chat = require('./Chat.js');
var ClearRefinements = require('./ClearRefinements.js');
var CurrentRefinements = require('./CurrentRefinements.js');
var FrequentlyBoughtTogether = require('./FrequentlyBoughtTogether.js');
var HierarchicalMenu = require('./HierarchicalMenu.js');
var Highlight = require('./Highlight.js');
var Hits = require('./Hits.js');
var HitsPerPage = require('./HitsPerPage.js');
var InfiniteHits = require('./InfiniteHits.js');
var LookingSimilar = require('./LookingSimilar.js');
var Menu = require('./Menu.js');
var Pagination = require('./Pagination.js');
var PoweredBy = require('./PoweredBy.js');
var RangeInput = require('./RangeInput.js');
var RefinementList = require('./RefinementList.js');
var RelatedProducts = require('./RelatedProducts.js');
var ReverseHighlight = require('./ReverseHighlight.js');
var SearchBox = require('./SearchBox.js');
var Snippet = require('./Snippet.js');
var SortBy = require('./SortBy.js');
var Stats = require('./Stats.js');
var ToggleRefinement = require('./ToggleRefinement.js');
var TrendingItems = require('./TrendingItems.js');
var FilterSuggestions = require('./FilterSuggestions.js');
var chat = require('instantsearch.js/cjs/lib/chat');



exports.EXPERIMENTAL_Autocomplete = Autocomplete.EXPERIMENTAL_Autocomplete;
exports.Breadcrumb = Breadcrumb.Breadcrumb;
exports.Chat = Chat.Chat;
exports.createDefaultTools = Chat.createDefaultTools;
exports.ClearRefinements = ClearRefinements.ClearRefinements;
exports.CurrentRefinements = CurrentRefinements.CurrentRefinements;
exports.FrequentlyBoughtTogether = FrequentlyBoughtTogether.FrequentlyBoughtTogether;
exports.HierarchicalMenu = HierarchicalMenu.HierarchicalMenu;
exports.Highlight = Highlight.Highlight;
exports.Hits = Hits.Hits;
exports.HitsPerPage = HitsPerPage.HitsPerPage;
exports.InfiniteHits = InfiniteHits.InfiniteHits;
exports.LookingSimilar = LookingSimilar.LookingSimilar;
exports.Menu = Menu.Menu;
exports.Pagination = Pagination.Pagination;
exports.PoweredBy = PoweredBy.PoweredBy;
exports.RangeInput = RangeInput.RangeInput;
exports.RefinementList = RefinementList.RefinementList;
exports.RelatedProducts = RelatedProducts.RelatedProducts;
exports.ReverseHighlight = ReverseHighlight.ReverseHighlight;
exports.SearchBox = SearchBox.SearchBox;
exports.Snippet = Snippet.Snippet;
exports.SortBy = SortBy.SortBy;
exports.Stats = Stats.Stats;
exports.ToggleRefinement = ToggleRefinement.ToggleRefinement;
exports.TrendingItems = TrendingItems.TrendingItems;
exports.FilterSuggestions = FilterSuggestions.FilterSuggestions;
Object.defineProperty(exports, "MemorizeToolType", {
	enumerable: true,
	get: function () { return chat.MemorizeToolType; }
});
Object.defineProperty(exports, "MemorySearchToolType", {
	enumerable: true,
	get: function () { return chat.MemorySearchToolType; }
});
Object.defineProperty(exports, "PonderToolType", {
	enumerable: true,
	get: function () { return chat.PonderToolType; }
});
Object.defineProperty(exports, "RecommendToolType", {
	enumerable: true,
	get: function () { return chat.RecommendToolType; }
});
Object.defineProperty(exports, "SearchIndexToolType", {
	enumerable: true,
	get: function () { return chat.SearchIndexToolType; }
});
