// react-instantsearch@7.26.0/dist/cjs/widgets/index.umd.js
'use strict';

var Autocomplete = require('./Autocomplete.js');
var Breadcrumb = require('./Breadcrumb.js');
var FilterSuggestions = require('./FilterSuggestions.js');
var ClearRefinements = require('./ClearRefinements.js');
var CurrentRefinements = require('./CurrentRefinements.js');
var FrequentlyBoughtTogether = require('./FrequentlyBoughtTogether.js');
var HierarchicalMenu = require('./HierarchicalMenu.js');
var Highlight = require('./Highlight.js');
var Hits = require('./Hits.js');
var HitsPerPage = require('./HitsPerPage.js');
var InfiniteHits = require('./InfiniteHits.js');
var LookingSimilar = require('./LookingSimilar.js');
var Menu = require('./Menu.js');
var Pagination = require('./Pagination.js');
var PoweredBy = require('./PoweredBy.js');
var RangeInput = require('./RangeInput.js');
var RefinementList = require('./RefinementList.js');
var RelatedProducts = require('./RelatedProducts.js');
var ReverseHighlight = require('./ReverseHighlight.js');
var SearchBox = require('./SearchBox.js');
var Snippet = require('./Snippet.js');
var SortBy = require('./SortBy.js');
var Stats = require('./Stats.js');
var ToggleRefinement = require('./ToggleRefinement.js');
var TrendingItems = require('./TrendingItems.js');

var Chat = function Chat() {
    throw new Error('"<Chat>" is not available from the UMD build.\n\nPlease use React InstantSearch with a packaging system:\nhttps://www.algolia.com/doc/guides/building-search-ui/installation/react/#install-react-instantsearch-as-an-npm-package');
};
var SearchIndexToolType = undefined;
var RecommendToolType = undefined;
var MemorizeToolType = undefined;
var MemorySearchToolType = undefined;
var PonderToolType = undefined;
var createDefaultTools = function createDefaultTools() {};

exports.EXPERIMENTAL_Autocomplete = Autocomplete.EXPERIMENTAL_Autocomplete;
exports.Breadcrumb = Breadcrumb.Breadcrumb;
exports.FilterSuggestions = FilterSuggestions.FilterSuggestions;
exports.ClearRefinements = ClearRefinements.ClearRefinements;
exports.CurrentRefinements = CurrentRefinements.CurrentRefinements;
exports.FrequentlyBoughtTogether = FrequentlyBoughtTogether.FrequentlyBoughtTogether;
exports.HierarchicalMenu = HierarchicalMenu.HierarchicalMenu;
exports.Highlight = Highlight.Highlight;
exports.Hits = Hits.Hits;
exports.HitsPerPage = HitsPerPage.HitsPerPage;
exports.InfiniteHits = InfiniteHits.InfiniteHits;
exports.LookingSimilar = LookingSimilar.LookingSimilar;
exports.Menu = Menu.Menu;
exports.Pagination = Pagination.Pagination;
exports.PoweredBy = PoweredBy.PoweredBy;
exports.RangeInput = RangeInput.RangeInput;
exports.RefinementList = RefinementList.RefinementList;
exports.RelatedProducts = RelatedProducts.RelatedProducts;
exports.ReverseHighlight = ReverseHighlight.ReverseHighlight;
exports.SearchBox = SearchBox.SearchBox;
exports.Snippet = Snippet.Snippet;
exports.SortBy = SortBy.SortBy;
exports.Stats = Stats.Stats;
exports.ToggleRefinement = ToggleRefinement.ToggleRefinement;
exports.TrendingItems = TrendingItems.TrendingItems;
exports.Chat = Chat;
exports.MemorizeToolType = MemorizeToolType;
exports.MemorySearchToolType = MemorySearchToolType;
exports.PonderToolType = PonderToolType;
exports.RecommendToolType = RecommendToolType;
exports.SearchIndexToolType = SearchIndexToolType;
exports.createDefaultTools = createDefaultTools;
