// react-instantsearch@7.26.0/dist/cjs/widgets/ReverseHighlight.js
'use strict';

var _object_spread = require('@swc/helpers/_/_object_spread');
var _object_spread_props = require('@swc/helpers/_/_object_spread_props');
var _object_without_properties = require('@swc/helpers/_/_object_without_properties');
var utils = require('instantsearch.js/cjs/lib/utils');
var React = require('react');
var ReverseHighlight$1 = require('../ui/ReverseHighlight.js');

function ReverseHighlight(_0) {
    var hit = _0.hit, attribute = _0.attribute, highlightedTagName = _0.highlightedTagName, nonHighlightedTagName = _0.nonHighlightedTagName, separator = _0.separator, props = _object_without_properties._(_0, [
        "hit",
        "attribute",
        "highlightedTagName",
        "nonHighlightedTagName",
        "separator"
    ]);
    var property = utils.getPropertyByPath(hit._highlightResult, attribute) || [];
    var properties = Array.isArray(property) ? property : [
        property
    ];
    var parts = properties.map(function(singleValue) {
        return utils.getHighlightedParts(utils.unescape(singleValue.value || '')).map(function(_0) {
            var isHighlighted = _0.isHighlighted, rest = _object_without_properties._(_0, [
                "isHighlighted"
            ]);
            return _object_spread_props._(_object_spread._({}, rest), {
                isHighlighted: !isHighlighted
            });
        });
    });
    return /*#__PURE__*/ React.createElement(ReverseHighlight$1.ReverseHighlight, _object_spread_props._(_object_spread._({}, props), {
        parts: parts,
        highlightedTagName: highlightedTagName,
        nonHighlightedTagName: nonHighlightedTagName,
        separator: separator
    }));
}

exports.ReverseHighlight = ReverseHighlight;
