// react-instantsearch@7.26.0/dist/cjs/widgets/ToggleRefinement.js
'use strict';

var _object_spread = require('@swc/helpers/_/_object_spread');
var _object_without_properties = require('@swc/helpers/_/_object_without_properties');
var React = require('react');
var reactInstantsearchCore = require('react-instantsearch-core');
var ToggleRefinement$1 = require('../ui/ToggleRefinement.js');

function ToggleRefinement(_0) {
    var attribute = _0.attribute, on = _0.on, off = _0.off, props = _object_without_properties._(_0, [
        "attribute",
        "on",
        "off"
    ]);
    var _useToggleRefinement = reactInstantsearchCore.useToggleRefinement({
        attribute: attribute,
        on: on,
        off: off
    }, {
        $$widgetType: 'ais.toggleRefinement'
    }), refine = _useToggleRefinement.refine, value = _useToggleRefinement.value;
    var uiProps = {
        checked: value.isRefined,
        onChange: function onChange(isChecked) {
            return refine({
                isRefined: !isChecked
            });
        }
    };
    return /*#__PURE__*/ React.createElement(ToggleRefinement$1.ToggleRefinement, _object_spread._({
        label: value.name
    }, props, uiProps));
}

exports.ToggleRefinement = ToggleRefinement;
