// instantsearch.js@4.90.0/cjs/components/InfiniteHits/InfiniteHits.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _object_spread = require('@swc/helpers/_/_object_spread');
var _object_spread_props = require('@swc/helpers/_/_object_spread_props');
var instantsearchUiComponents = require('instantsearch-ui-components');
var preact = require('preact');
var listener = require('../../lib/insights/listener.js');
var Template = require('../Template/Template.js');

var DefaultBanner = function DefaultBanner(param) {
    var banner = param.banner, classNames = param.classNames;
    if (!banner.image.urls[0].url) {
        return null;
    }
    return /*#__PURE__*/ preact.h("aside", {
        className: instantsearchUiComponents.cx(classNames.bannerRoot)
    }, banner.link ? /*#__PURE__*/ preact.h("a", {
        className: instantsearchUiComponents.cx(classNames.bannerLink),
        href: banner.link.url,
        target: banner.link.target
    }, /*#__PURE__*/ preact.h("img", {
        className: instantsearchUiComponents.cx(classNames.bannerImage),
        src: banner.image.urls[0].url,
        alt: banner.image.title
    })) : /*#__PURE__*/ preact.h("img", {
        className: instantsearchUiComponents.cx(classNames.bannerImage),
        src: banner.image.urls[0].url,
        alt: banner.image.title
    }));
};
var InfiniteHits = function InfiniteHits(param) {
    var results = param.results, hits = param.hits, insights = param.insights, bindEvent = param.bindEvent, sendEvent = param.sendEvent, hasShowPrevious = param.hasShowPrevious, showPrevious = param.showPrevious, showMore = param.showMore, isFirstPage = param.isFirstPage, isLastPage = param.isLastPage, cssClasses = param.cssClasses, templateProps = param.templateProps, banner = param.banner;
    var handleInsightsClick = listener.createInsightsEventHandler({
        insights: insights,
        sendEvent: sendEvent
    });
    if (results.hits.length === 0) {
        return /*#__PURE__*/ preact.h("div", {
            className: instantsearchUiComponents.cx(cssClasses.root, cssClasses.emptyRoot),
            onClick: handleInsightsClick
        }, banner && (templateProps.templates.banner ? /*#__PURE__*/ preact.h(Template.default, _object_spread_props._(_object_spread._({}, templateProps), {
            templateKey: "banner",
            rootTagName: "fragment",
            data: {
                banner: banner,
                className: cssClasses.bannerRoot
            }
        })) : /*#__PURE__*/ preact.h(DefaultBanner, {
            banner: banner,
            classNames: cssClasses
        })), /*#__PURE__*/ preact.h(Template.default, _object_spread_props._(_object_spread._({}, templateProps), {
            templateKey: "empty",
            rootTagName: "fragment",
            data: results
        })));
    }
    return /*#__PURE__*/ preact.h("div", {
        className: cssClasses.root
    }, hasShowPrevious && /*#__PURE__*/ preact.h(Template.default, _object_spread_props._(_object_spread._({}, templateProps), {
        templateKey: "showPreviousText",
        rootTagName: "button",
        rootProps: {
            className: instantsearchUiComponents.cx(cssClasses.loadPrevious, isFirstPage && cssClasses.disabledLoadPrevious),
            disabled: isFirstPage,
            onClick: showPrevious
        }
    })), banner && (templateProps.templates.banner ? /*#__PURE__*/ preact.h(Template.default, _object_spread_props._(_object_spread._({}, templateProps), {
        templateKey: "banner",
        rootTagName: "fragment",
        data: {
            banner: banner,
            className: cssClasses.bannerRoot
        }
    })) : /*#__PURE__*/ preact.h(DefaultBanner, {
        banner: banner,
        classNames: cssClasses
    })), /*#__PURE__*/ preact.h("ol", {
        className: cssClasses.list
    }, hits.map(function(hit, index) {
        return /*#__PURE__*/ preact.h(Template.default, _object_spread_props._(_object_spread._({}, templateProps), {
            templateKey: "item",
            rootTagName: "li",
            rootProps: {
                className: cssClasses.item,
                onClick: function onClick(event) {
                    handleInsightsClick(event);
                    sendEvent('click:internal', hit, 'Hit Clicked');
                },
                onAuxClick: function onAuxClick(event) {
                    handleInsightsClick(event);
                    sendEvent('click:internal', hit, 'Hit Clicked');
                }
            },
            key: hit.objectID,
            data: _object_spread_props._(_object_spread._({}, hit), {
                get __hitIndex () {
                    return index;
                }
            }),
            bindEvent: bindEvent,
            sendEvent: sendEvent
        }));
    })), /*#__PURE__*/ preact.h(Template.default, _object_spread_props._(_object_spread._({}, templateProps), {
        templateKey: "showMoreText",
        rootTagName: "button",
        rootProps: {
            className: instantsearchUiComponents.cx(cssClasses.loadMore, isLastPage && cssClasses.disabledLoadMore),
            disabled: isLastPage,
            onClick: showMore
        }
    })));
};

exports.default = InfiniteHits;
