// instantsearch.js@4.90.0/cjs/components/Answers/Answers.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _object_spread = require('@swc/helpers/_/_object_spread');
var _object_spread_props = require('@swc/helpers/_/_object_spread_props');
var instantsearchUiComponents = require('instantsearch-ui-components');
var preact = require('preact');
var Template = require('../Template/Template.js');

var Answers = function Answers(param) {
    var hits = param.hits, isLoading = param.isLoading, cssClasses = param.cssClasses, templateProps = param.templateProps;
    return /*#__PURE__*/ preact.h("div", {
        className: instantsearchUiComponents.cx(cssClasses.root, hits.length === 0 && cssClasses.emptyRoot)
    }, /*#__PURE__*/ preact.h(Template.default, _object_spread_props._(_object_spread._({}, templateProps), {
        templateKey: "header",
        rootProps: {
            className: cssClasses.header
        },
        data: {
            hits: hits,
            isLoading: isLoading
        }
    })), isLoading ? /*#__PURE__*/ preact.h(Template.default, _object_spread_props._(_object_spread._({}, templateProps), {
        templateKey: "loader",
        rootProps: {
            className: cssClasses.loader
        }
    })) : /*#__PURE__*/ preact.h("ul", {
        className: cssClasses.list
    }, hits.map(function(hit, index) {
        return /*#__PURE__*/ preact.h(Template.default, _object_spread_props._(_object_spread._({}, templateProps), {
            templateKey: "item",
            rootTagName: "li",
            rootProps: {
                className: cssClasses.item
            },
            key: hit.objectID,
            data: _object_spread_props._(_object_spread._({}, hit), {
                get __hitIndex () {
                    return index;
                }
            })
        }));
    })));
};

exports.default = Answers;
