// instantsearch.js@4.90.0/cjs/components/Breadcrumb/Breadcrumb.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _object_spread = require('@swc/helpers/_/_object_spread');
var _object_spread_props = require('@swc/helpers/_/_object_spread_props');
var instantsearchUiComponents = require('instantsearch-ui-components');
var preact = require('preact');
var Template = require('../Template/Template.js');
var isSpecialClick = require('../../lib/utils/isSpecialClick.js');

var Breadcrumb = function Breadcrumb(param) {
    var items = param.items, cssClasses = param.cssClasses, templateProps = param.templateProps, createURL = param.createURL, refine = param.refine;
    return /*#__PURE__*/ preact.h("div", {
        className: instantsearchUiComponents.cx(cssClasses.root, items.length === 0 && cssClasses.noRefinementRoot)
    }, /*#__PURE__*/ preact.h("ul", {
        className: cssClasses.list
    }, /*#__PURE__*/ preact.h("li", {
        className: instantsearchUiComponents.cx(cssClasses.item, items.length === 0 && cssClasses.selectedItem)
    }, /*#__PURE__*/ preact.h(Template.default, _object_spread_props._(_object_spread._({}, templateProps), {
        templateKey: "home",
        rootTagName: "a",
        rootProps: {
            className: cssClasses.link,
            href: createURL(null),
            onClick: function onClick(event) {
                if (isSpecialClick.isSpecialClick(event)) {
                    return;
                }
                event.preventDefault();
                refine(null);
            }
        }
    }))), items.map(function(item, idx) {
        var isLast = idx === items.length - 1;
        return /*#__PURE__*/ preact.h("li", {
            key: item.label + idx,
            className: instantsearchUiComponents.cx(cssClasses.item, isLast && cssClasses.selectedItem)
        }, /*#__PURE__*/ preact.h(Template.default, _object_spread_props._(_object_spread._({}, templateProps), {
            templateKey: "separator",
            rootTagName: "span",
            rootProps: {
                className: cssClasses.separator,
                'aria-hidden': true
            }
        })), isLast ? item.label : /*#__PURE__*/ preact.h("a", {
            className: cssClasses.link,
            href: createURL(item.value),
            onClick: function onClick(event) {
                if (isSpecialClick.isSpecialClick(event)) {
                    return;
                }
                event.preventDefault();
                refine(item.value);
            }
        }, item.label));
    })));
};

exports.default = Breadcrumb;
