// @rasenganjs/mdx@1.2.0-beta.8/dist/chunk-U2MZHTHK.js
// src/utils/extract-toc.ts
import React from "react";
function extractTOC(markdown) {
  const lines = markdown.split("\n");
  const toc = [];
  lines.forEach((line) => {
    const h2Match = line.match(/^## (.+)/);
    const h3Match = line.match(/^### (.+)/);
    if (h2Match) {
      const title = h2Match[1].trim();
      const anchor = generateAnchor(title);
      toc.push({
        title,
        anchor,
        level: 2,
        children: []
      });
    } else if (h3Match && toc.length > 0) {
      const title = h3Match[1].trim();
      const anchor = generateAnchor(title);
      toc[toc.length - 1].children.push({
        title,
        anchor,
        level: 3
      });
    }
  });
  return toc;
}
var generateAnchor = (title) => {
  if (Array.isArray(title)) {
    const text2 = title.map((item) => {
      if (React.isValidElement(item)) {
        return item.props["children"].toString().trim();
      }
      return item.toString().trim();
    }).join(" ");
    const lastItem = title[title.length - 1];
    if (typeof lastItem === "string") {
      const match1 = text2.match(/\s\[#([^\]]+)\]$/);
      if (match1) {
        return {
          id: match1[1],
          // remove the last element from title
          text: title.slice(0, title.length - 1)
        };
      }
    }
    return {
      id: text2.replace(/\s+/g, "-").toLowerCase(),
      text: title
    };
  }
  if (React.isValidElement(title)) {
    return {
      id: title.props["children"].toString().trim().replace(/\s+/g, "-").toLowerCase(),
      text: title
    };
  }
  const strippedTitle = String(toText(title).trim());
  const match = strippedTitle.match(/\s\[#([^\]]+)\]$/);
  let id;
  let text;
  if (match) {
    id = match[1].trim().replace(/\s+/g, "-").toLowerCase();
    text = strippedTitle.replace(match[0], "").trim();
  } else {
    id = strippedTitle.replace(/\s+/g, "-").toLowerCase();
    text = strippedTitle;
  }
  return { id, text };
};
function toText(input) {
  if (input == null) return "";
  if (typeof input === "string") return input;
  if (Array.isArray(input)) return input.map(toText).join(" ");
  if (input instanceof Uint8Array)
    return new TextDecoder().decode(input);
  try {
    return String(input);
  } catch {
    return "";
  }
}

export {
  extractTOC,
  generateAnchor
};
