// @rasenganjs/mdx@1.2.0-beta.8/dist/plugin.js
import {
  extractTOC
} from "./chunk-U2MZHTHK.js";

// src/utils/plugin.ts
import matter from "gray-matter";

// src/utils/create-filter.ts
import micromatch from "micromatch";
function createFilter(include, exclude) {
  return function(id) {
    if (typeof id !== "string") return false;
    const matcher = micromatch.matcher(include);
    if (exclude) {
      const excluder = micromatch.matcher(exclude);
      return matcher(id) && !excluder(id);
    } else {
      return matcher(id);
    }
  };
}

// src/utils/plugin.ts
import remarkGfm from "remark-gfm";
import rehypeStringify from "rehype-stringify";
import remarkParse from "remark-parse";
import remarkRehype from "remark-rehype";
import rehypePrettyCode from "rehype-pretty-code";
async function plugin(options) {
  const mdx = (await import("@mdx-js/rollup")).default;
  const userConfig = {
    rehypePlugins: options?.rehypePlugins ?? [],
    remarkPlugins: options?.rehypePlugins ?? [],
    code: {
      theme: options?.code?.theme ?? {
        light: "one-light",
        dark: "one-dark-pro"
      },
      keepBackground: options?.code?.keepBackground ?? true
    }
  };
  let config;
  const filter = createFilter("**/*.md?(x)");
  const mdxInstance = mdx({
    remarkPlugins: [remarkParse, remarkGfm, ...userConfig.remarkPlugins],
    rehypePlugins: [
      remarkRehype,
      rehypeStringify,
      [
        rehypePrettyCode,
        {
          theme: userConfig.code.theme,
          keepBackground: userConfig.code.keepBackground
        }
      ],
      // Developer rehype plugin list
      ...userConfig.rehypePlugins
    ]
  });
  return {
    name: "vite-plugin-rasengan-mdx",
    // Apply transformation of the mdx file before other plugins
    enforce: "pre",
    config(config2, env) {
      mdxInstance.config(config2, env);
    },
    /**
     * Stores the resolved Vite configuration for later use.
     *
     * @param resolvedConfig - The resolved Vite configuration object.
     */
    configResolved(resolvedConfig) {
      config = resolvedConfig;
    },
    /**
     * Transforms an MDX file by applying the `@mdx-js/rollup` transformation, extracting frontmatter data, and appending a `metadata` object to the transformed content.
     *
     * @param code - The content of the MDX file.
     * @param id - The ID of the MDX file.
     * @returns An object containing the transformed MDX code and a source map, or `null` if the file is not an MDX file.
     */
    async transform(code, id) {
      if (!filter(id)) {
        return null;
      }
      const { content, data: frontmatter } = matter(code);
      const isTocVisible = frontmatter.toc !== void 0 ? frontmatter.toc : false;
      const toc = extractTOC(content);
      const result = await mdxInstance.transform(content, id);
      const fileName = id.split("/").pop().replace(/.page.mdx?$/, "");
      const metadata = {
        path: frontmatter.path || `/${fileName}`,
        metadata: frontmatter.metadata || {
          title: fileName
        }
      };
      let newCode = result.code;
      newCode = newCode.replace(/export default\s+/, "const MDXContent = ");
      return {
        code: `
          import { MDXRenderer } from "@rasenganjs/mdx";
          import mdxConfig from "virtual:rasengan/mdx-components";

          ${newCode}

          const metadata = ${JSON.stringify(metadata)};
          const toc = ${isTocVisible ? JSON.stringify(toc) : void 0};


          const MDXWrapper = {
            metadata,
            toc,
            type: "MDXPageComponent", // Helps to differentiate PageComponent from MDXPageComponent
            Content: MDXContent, // The content of the MDX file
            Renderer: MDXRenderer, // The renderer component
            config: mdxConfig, // The MDX config coming from mdx-components virtual module
          };


          export default MDXWrapper;
        `,
        map: result.map
      };
    }
  };
}
function loadMDXComponentsPlugin() {
  const virtualModuleId = "virtual:rasengan/mdx-components";
  const resolvedVirtualModuleId = "\0" + virtualModuleId;
  return {
    name: "vite-plugin-rasengan-mdx-components",
    resolveId(id) {
      if (id === virtualModuleId) {
        return resolvedVirtualModuleId;
      }
    },
    async load(id) {
      if (id === resolvedVirtualModuleId) {
        return `
          const modules = import.meta.glob(
            [
              '/mdx-components.{js,jsx,ts,tsx}',
            ],
            { eager: true }
          );
          let config = {};

          const modulesArray = Object.entries(modules);

          if (modulesArray.length > 0) {
            const [filePath, mod] = modulesArray[0];
            config = mod.default;
          }

          export default config;
        `;
      }
    }
  };
}

// src/plugin.ts
var plugin2 = (options) => [
  plugin(options),
  loadMDXComponentsPlugin()
];
var plugin_default = plugin2;
export {
  plugin_default as default
};
