// @rasenganjs/mdx@1.2.0-beta.8/dist/plugin.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/plugin.ts
var plugin_exports = {};
__export(plugin_exports, {
  default: () => plugin_default
});
module.exports = __toCommonJS(plugin_exports);

// src/utils/plugin.ts
var import_gray_matter = __toESM(require("gray-matter"), 1);

// src/utils/create-filter.ts
var import_micromatch = __toESM(require("micromatch"), 1);
function createFilter(include, exclude) {
  return function(id) {
    if (typeof id !== "string") return false;
    const matcher = import_micromatch.default.matcher(include);
    if (exclude) {
      const excluder = import_micromatch.default.matcher(exclude);
      return matcher(id) && !excluder(id);
    } else {
      return matcher(id);
    }
  };
}

// src/utils/plugin.ts
var import_remark_gfm = __toESM(require("remark-gfm"), 1);
var import_rehype_stringify = __toESM(require("rehype-stringify"), 1);
var import_remark_parse = __toESM(require("remark-parse"), 1);
var import_remark_rehype = __toESM(require("remark-rehype"), 1);
var import_rehype_pretty_code = __toESM(require("rehype-pretty-code"), 1);

// src/utils/extract-toc.ts
var import_react = __toESM(require("react"), 1);
function extractTOC(markdown) {
  const lines = markdown.split("\n");
  const toc = [];
  lines.forEach((line) => {
    const h2Match = line.match(/^## (.+)/);
    const h3Match = line.match(/^### (.+)/);
    if (h2Match) {
      const title = h2Match[1].trim();
      const anchor = generateAnchor(title);
      toc.push({
        title,
        anchor,
        level: 2,
        children: []
      });
    } else if (h3Match && toc.length > 0) {
      const title = h3Match[1].trim();
      const anchor = generateAnchor(title);
      toc[toc.length - 1].children.push({
        title,
        anchor,
        level: 3
      });
    }
  });
  return toc;
}
var generateAnchor = (title) => {
  if (Array.isArray(title)) {
    const text2 = title.map((item) => {
      if (import_react.default.isValidElement(item)) {
        return item.props["children"].toString().trim();
      }
      return item.toString().trim();
    }).join(" ");
    const lastItem = title[title.length - 1];
    if (typeof lastItem === "string") {
      const match1 = text2.match(/\s\[#([^\]]+)\]$/);
      if (match1) {
        return {
          id: match1[1],
          // remove the last element from title
          text: title.slice(0, title.length - 1)
        };
      }
    }
    return {
      id: text2.replace(/\s+/g, "-").toLowerCase(),
      text: title
    };
  }
  if (import_react.default.isValidElement(title)) {
    return {
      id: title.props["children"].toString().trim().replace(/\s+/g, "-").toLowerCase(),
      text: title
    };
  }
  const strippedTitle = String(toText(title).trim());
  const match = strippedTitle.match(/\s\[#([^\]]+)\]$/);
  let id;
  let text;
  if (match) {
    id = match[1].trim().replace(/\s+/g, "-").toLowerCase();
    text = strippedTitle.replace(match[0], "").trim();
  } else {
    id = strippedTitle.replace(/\s+/g, "-").toLowerCase();
    text = strippedTitle;
  }
  return { id, text };
};
function toText(input) {
  if (input == null) return "";
  if (typeof input === "string") return input;
  if (Array.isArray(input)) return input.map(toText).join(" ");
  if (input instanceof Uint8Array)
    return new TextDecoder().decode(input);
  try {
    return String(input);
  } catch {
    return "";
  }
}

// src/utils/plugin.ts
async function plugin(options) {
  const mdx = (await import("@mdx-js/rollup")).default;
  const userConfig = {
    rehypePlugins: options?.rehypePlugins ?? [],
    remarkPlugins: options?.rehypePlugins ?? [],
    code: {
      theme: options?.code?.theme ?? {
        light: "one-light",
        dark: "one-dark-pro"
      },
      keepBackground: options?.code?.keepBackground ?? true
    }
  };
  let config;
  const filter = createFilter("**/*.md?(x)");
  const mdxInstance = mdx({
    remarkPlugins: [import_remark_parse.default, import_remark_gfm.default, ...userConfig.remarkPlugins],
    rehypePlugins: [
      import_remark_rehype.default,
      import_rehype_stringify.default,
      [
        import_rehype_pretty_code.default,
        {
          theme: userConfig.code.theme,
          keepBackground: userConfig.code.keepBackground
        }
      ],
      // Developer rehype plugin list
      ...userConfig.rehypePlugins
    ]
  });
  return {
    name: "vite-plugin-rasengan-mdx",
    // Apply transformation of the mdx file before other plugins
    enforce: "pre",
    config(config2, env) {
      mdxInstance.config(config2, env);
    },
    /**
     * Stores the resolved Vite configuration for later use.
     *
     * @param resolvedConfig - The resolved Vite configuration object.
     */
    configResolved(resolvedConfig) {
      config = resolvedConfig;
    },
    /**
     * Transforms an MDX file by applying the `@mdx-js/rollup` transformation, extracting frontmatter data, and appending a `metadata` object to the transformed content.
     *
     * @param code - The content of the MDX file.
     * @param id - The ID of the MDX file.
     * @returns An object containing the transformed MDX code and a source map, or `null` if the file is not an MDX file.
     */
    async transform(code, id) {
      if (!filter(id)) {
        return null;
      }
      const { content, data: frontmatter } = (0, import_gray_matter.default)(code);
      const isTocVisible = frontmatter.toc !== void 0 ? frontmatter.toc : false;
      const toc = extractTOC(content);
      const result = await mdxInstance.transform(content, id);
      const fileName = id.split("/").pop().replace(/.page.mdx?$/, "");
      const metadata = {
        path: frontmatter.path || `/${fileName}`,
        metadata: frontmatter.metadata || {
          title: fileName
        }
      };
      let newCode = result.code;
      newCode = newCode.replace(/export default\s+/, "const MDXContent = ");
      return {
        code: `
          import { MDXRenderer } from "@rasenganjs/mdx";
          import mdxConfig from "virtual:rasengan/mdx-components";

          ${newCode}

          const metadata = ${JSON.stringify(metadata)};
          const toc = ${isTocVisible ? JSON.stringify(toc) : void 0};


          const MDXWrapper = {
            metadata,
            toc,
            type: "MDXPageComponent", // Helps to differentiate PageComponent from MDXPageComponent
            Content: MDXContent, // The content of the MDX file
            Renderer: MDXRenderer, // The renderer component
            config: mdxConfig, // The MDX config coming from mdx-components virtual module
          };


          export default MDXWrapper;
        `,
        map: result.map
      };
    }
  };
}
function loadMDXComponentsPlugin() {
  const virtualModuleId = "virtual:rasengan/mdx-components";
  const resolvedVirtualModuleId = "\0" + virtualModuleId;
  return {
    name: "vite-plugin-rasengan-mdx-components",
    resolveId(id) {
      if (id === virtualModuleId) {
        return resolvedVirtualModuleId;
      }
    },
    async load(id) {
      if (id === resolvedVirtualModuleId) {
        return `
          const modules = import.meta.glob(
            [
              '/mdx-components.{js,jsx,ts,tsx}',
            ],
            { eager: true }
          );
          let config = {};

          const modulesArray = Object.entries(modules);

          if (modulesArray.length > 0) {
            const [filePath, mod] = modulesArray[0];
            config = mod.default;
          }

          export default config;
        `;
      }
    }
  };
}

// src/plugin.ts
var plugin2 = (options) => [
  plugin(options),
  loadMDXComponentsPlugin()
];
var plugin_default = plugin2;
