// instantsearch.js@4.90.0/cjs/components/RelevantSort/RelevantSort.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var preact = require('preact');
var Template = require('../Template/Template.js');

var RelevantSort = function RelevantSort(param) {
    var cssClasses = param.cssClasses, templates = param.templates, isRelevantSorted = param.isRelevantSorted, isVirtualReplica = param.isVirtualReplica, refine = param.refine;
    return isVirtualReplica ? /*#__PURE__*/ preact.h("div", {
        className: cssClasses.root
    }, /*#__PURE__*/ preact.h(Template.default, {
        templateKey: "text",
        templates: templates,
        rootProps: {
            className: cssClasses.text
        },
        data: {
            isRelevantSorted: isRelevantSorted
        }
    }), /*#__PURE__*/ preact.h("button", {
        type: "button",
        className: cssClasses.button,
        onClick: function onClick() {
            if (isRelevantSorted) {
                refine(0);
            } else {
                refine(undefined);
            }
        }
    }, /*#__PURE__*/ preact.h(Template.default, {
        rootTagName: "span",
        templateKey: "button",
        templates: templates,
        data: {
            isRelevantSorted: isRelevantSorted
        }
    }))) : null;
};

exports.default = RelevantSort;
