// instantsearch.js@4.90.0/cjs/connectors/configure-related-items/connectConfigureRelatedItems.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _object_spread = require('@swc/helpers/_/_object_spread');
var _object_spread_props = require('@swc/helpers/_/_object_spread_props');
var _to_consumable_array = require('@swc/helpers/_/_to_consumable_array');
var algoliasearchHelper = require('algoliasearch-helper');
var connectConfigure = require('../configure/connectConfigure.js');
var documentation = require('../../lib/utils/documentation.js');
var getPropertyByPath = require('../../lib/utils/getPropertyByPath.js');

var withUsage = documentation.createDocumentationMessageGenerator({
    name: 'configure-related-items',
    connector: true
});
function createOptionalFilter(param) {
    var attributeName = param.attributeName, attributeValue = param.attributeValue, attributeScore = param.attributeScore;
    return "".concat(attributeName, ":").concat(attributeValue, "<score=").concat(attributeScore || 1, ">");
}
var connectConfigureRelatedItems = function connectConfigureRelatedItems(renderFn, unmountFn) {
    return function(widgetParams) {
        var _ref = widgetParams || {}, hit = _ref.hit, matchingPatterns = _ref.matchingPatterns, _ref_transformSearchParameters = _ref.transformSearchParameters, transformSearchParameters = _ref_transformSearchParameters === void 0 ? function(x) {
            return x;
        } : _ref_transformSearchParameters;
        if (!hit) {
            throw new Error(withUsage('The `hit` option is required.'));
        }
        if (!matchingPatterns) {
            throw new Error(withUsage('The `matchingPatterns` option is required.'));
        }
        var optionalFilters = Object.keys(matchingPatterns).reduce(function(acc, attributeName) {
            var attribute = matchingPatterns[attributeName];
            var attributeValue = getPropertyByPath.getPropertyByPath(hit, attributeName);
            var attributeScore = attribute.score;
            if (Array.isArray(attributeValue)) {
                return _to_consumable_array._(acc).concat([
                    attributeValue.map(function(attributeSubValue) {
                        return createOptionalFilter({
                            attributeName: attributeName,
                            attributeValue: attributeSubValue,
                            attributeScore: attributeScore
                        });
                    })
                ]);
            }
            if (typeof attributeValue === 'string') {
                return _to_consumable_array._(acc).concat([
                    createOptionalFilter({
                        attributeName: attributeName,
                        attributeValue: attributeValue,
                        attributeScore: attributeScore
                    })
                ]);
            }
            return acc;
        }, []);
        var searchParameters = _object_spread._({}, transformSearchParameters(new algoliasearchHelper.SearchParameters({
            sumOrFiltersScores: true,
            facetFilters: [
                "objectID:-".concat(hit.objectID)
            ],
            optionalFilters: optionalFilters
        })));
        var makeWidget = connectConfigure.default(renderFn, unmountFn);
        return _object_spread_props._(_object_spread._({}, makeWidget({
            searchParameters: searchParameters
        })), {
            $$type: 'ais.configureRelatedItems'
        });
    };
};

exports.default = connectConfigureRelatedItems;
