// instantsearch.js@4.90.0/cjs/widgets/current-refinements/current-refinements.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _object_spread = require('@swc/helpers/_/_object_spread');
var _object_spread_props = require('@swc/helpers/_/_object_spread_props');
var instantsearchUiComponents = require('instantsearch-ui-components');
var preact = require('preact');
var CurrentRefinements = require('../../components/CurrentRefinements/CurrentRefinements.js');
var connectCurrentRefinements = require('../../connectors/current-refinements/connectCurrentRefinements.js');
var suit$1 = require('../../lib/suit.js');
var documentation = require('../../lib/utils/documentation.js');
var getContainerNode = require('../../lib/utils/getContainerNode.js');

var withUsage = documentation.createDocumentationMessageGenerator({
    name: 'current-refinements'
});
var suit = suit$1.component('CurrentRefinements');
var renderer = function renderer(param, isFirstRender) {
    var items = param.items, widgetParams = param.widgetParams, canRefine = param.canRefine;
    if (isFirstRender) {
        return;
    }
    var container = widgetParams.container, cssClasses = widgetParams.cssClasses;
    preact.render(/*#__PURE__*/ preact.h(CurrentRefinements.default, {
        cssClasses: cssClasses,
        items: items,
        canRefine: canRefine
    }), container);
};
var currentRefinements = function currentRefinements(widgetParams) {
    var _ref = widgetParams || {}, container = _ref.container, includedAttributes = _ref.includedAttributes, excludedAttributes = _ref.excludedAttributes, tmp = _ref.cssClasses, userCssClasses = tmp === void 0 ? {} : tmp, transformItems = _ref.transformItems;
    if (!container) {
        throw new Error(withUsage('The `container` option is required.'));
    }
    var containerNode = getContainerNode.getContainerNode(container);
    var cssClasses = {
        root: instantsearchUiComponents.cx(suit(), userCssClasses.root),
        noRefinementRoot: instantsearchUiComponents.cx(suit({
            modifierName: 'noRefinement'
        }), userCssClasses.noRefinementRoot),
        list: instantsearchUiComponents.cx(suit({
            descendantName: 'list'
        }), userCssClasses.list),
        item: instantsearchUiComponents.cx(suit({
            descendantName: 'item'
        }), userCssClasses.item),
        label: instantsearchUiComponents.cx(suit({
            descendantName: 'label'
        }), userCssClasses.label),
        category: instantsearchUiComponents.cx(suit({
            descendantName: 'category'
        }), userCssClasses.category),
        categoryLabel: instantsearchUiComponents.cx(suit({
            descendantName: 'categoryLabel'
        }), userCssClasses.categoryLabel),
        delete: instantsearchUiComponents.cx(suit({
            descendantName: 'delete'
        }), userCssClasses.delete)
    };
    var makeWidget = connectCurrentRefinements.default(renderer, function() {
        return preact.render(null, containerNode);
    });
    return _object_spread_props._(_object_spread._({}, makeWidget({
        container: containerNode,
        cssClasses: cssClasses,
        includedAttributes: includedAttributes,
        excludedAttributes: excludedAttributes,
        transformItems: transformItems
    })), {
        $$widgetType: 'ais.currentRefinements'
    });
};

exports.default = currentRefinements;
