// react-instantsearch@7.26.0/dist/es/components/AutocompleteSearch.js
import { _ as _$1 } from '@swc/helpers/cjs/_object_spread.cjs';
import { _ } from '@swc/helpers/cjs/_object_spread_props.cjs';
import { createAutocompleteSearchComponent } from 'instantsearch-ui-components';
import React, { createElement, Fragment } from 'react';

var AutocompleteSearchComponent = createAutocompleteSearchComponent({
    createElement: createElement,
    Fragment: Fragment
});
function AutocompleteSearch(param) {
    var inputProps = param.inputProps, clearQuery = param.clearQuery, onQueryChange = param.onQueryChange, query = param.query, refine = param.refine, isSearchStalled = param.isSearchStalled;
    return /*#__PURE__*/ React.createElement(AutocompleteSearchComponent, {
        inputProps: _(_$1({}, inputProps), {
            onChange: function onChange(event) {
                var value = event.currentTarget.value;
                refine(value);
                onQueryChange === null || onQueryChange === void 0 ? void 0 : onQueryChange(value);
            }
        }),
        onClear: clearQuery,
        query: query,
        isSearchStalled: isSearchStalled
    });
}

export { AutocompleteSearch };
