// react-instantsearch@7.26.0/dist/cjs/components/AutocompleteSearch.js
'use strict';

var _object_spread = require('@swc/helpers/_/_object_spread');
var _object_spread_props = require('@swc/helpers/_/_object_spread_props');
var instantsearchUiComponents = require('instantsearch-ui-components');
var React = require('react');

var AutocompleteSearchComponent = instantsearchUiComponents.createAutocompleteSearchComponent({
    createElement: React.createElement,
    Fragment: React.Fragment
});
function AutocompleteSearch(param) {
    var inputProps = param.inputProps, clearQuery = param.clearQuery, onQueryChange = param.onQueryChange, query = param.query, refine = param.refine, isSearchStalled = param.isSearchStalled;
    return /*#__PURE__*/ React.createElement(AutocompleteSearchComponent, {
        inputProps: _object_spread_props._(_object_spread._({}, inputProps), {
            onChange: function onChange(event) {
                var value = event.currentTarget.value;
                refine(value);
                onQueryChange === null || onQueryChange === void 0 ? void 0 : onQueryChange(value);
            }
        }),
        onClear: clearQuery,
        query: query,
        isSearchStalled: isSearchStalled
    });
}

exports.AutocompleteSearch = AutocompleteSearch;
