// react-instantsearch-core@7.26.0/dist/cjs/hooks/useInstantSearch.js
'use strict';

var _to_consumable_array = require('@swc/helpers/_/_to_consumable_array');
var React = require('react');
var useInstantSearchContext = require('../lib/useInstantSearchContext.js');
var useIsomorphicLayoutEffect = require('../lib/useIsomorphicLayoutEffect.js');
var useSearchResults = require('../lib/useSearchResults.js');
var useSearchState = require('../lib/useSearchState.js');

function useInstantSearch() {
    var catchError = (arguments.length > 0 && arguments[0] !== void 0 ? arguments[0] : {}).catchError;
    var search = useInstantSearchContext.useInstantSearchContext();
    var _useSearchState = useSearchState.useSearchState(), uiState = _useSearchState.uiState, setUiState = _useSearchState.setUiState, indexUiState = _useSearchState.indexUiState, setIndexUiState = _useSearchState.setIndexUiState, renderState = _useSearchState.renderState, indexRenderState = _useSearchState.indexRenderState;
    var _useSearchResults = useSearchResults.useSearchResults(), results = _useSearchResults.results, scopedResults = _useSearchResults.scopedResults;
    var addMiddlewares = React.useCallback(function() {
        for(var _len = arguments.length, middlewares = new Array(_len), _key = 0; _key < _len; _key++){
            middlewares[_key] = arguments[_key];
        }
        var _search;
        (_search = search).use.apply(_search, _to_consumable_array._(middlewares));
        return function() {
            var _search;
            (_search = search).unuse.apply(_search, _to_consumable_array._(middlewares));
        };
    }, [
        search
    ]);
    var refresh = React.useCallback(function() {
        search.refresh();
    }, [
        search
    ]);
    useIsomorphicLayoutEffect.useIsomorphicLayoutEffect(function() {
        if (catchError) {
            var onError = function onError() {};
            search.addListener('error', onError);
            return function() {
                return search.removeListener('error', onError);
            };
        }
        return function() {};
    }, [
        search,
        catchError
    ]);
    return {
        results: results,
        scopedResults: scopedResults,
        uiState: uiState,
        setUiState: setUiState,
        indexUiState: indexUiState,
        setIndexUiState: setIndexUiState,
        renderState: renderState,
        indexRenderState: indexRenderState,
        addMiddlewares: addMiddlewares,
        refresh: refresh,
        status: search.status,
        error: search.error
    };
}

exports.useInstantSearch = useInstantSearch;
