// instantsearch.js@4.90.0/cjs/components/GeoSearchControls/GeoSearchControls.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _object_spread = require('@swc/helpers/_/_object_spread');
var _object_spread_props = require('@swc/helpers/_/_object_spread_props');
var instantsearchUiComponents = require('instantsearch-ui-components');
var preact = require('preact');
var Template = require('../Template/Template.js');
var GeoSearchButton = require('./GeoSearchButton.js');
var GeoSearchToggle = require('./GeoSearchToggle.js');

var GeoSearchControls = function GeoSearchControls(param) {
    var cssClasses = param.cssClasses, enableRefine = param.enableRefine, enableRefineControl = param.enableRefineControl, enableClearMapRefinement = param.enableClearMapRefinement, isRefineOnMapMove = param.isRefineOnMapMove, isRefinedWithMap = param.isRefinedWithMap, hasMapMoveSinceLastRefine = param.hasMapMoveSinceLastRefine, onRefineToggle = param.onRefineToggle, onRefineClick = param.onRefineClick, onClearClick = param.onClearClick, templateProps = param.templateProps;
    return /*#__PURE__*/ preact.h(preact.Fragment, null, enableRefine && /*#__PURE__*/ preact.h("div", null, enableRefineControl && /*#__PURE__*/ preact.h("div", {
        className: cssClasses.control
    }, isRefineOnMapMove || !hasMapMoveSinceLastRefine ? /*#__PURE__*/ preact.h(GeoSearchToggle.default, {
        classNameLabel: instantsearchUiComponents.cx(cssClasses.label, isRefineOnMapMove && cssClasses.selectedLabel),
        classNameInput: cssClasses.input,
        checked: isRefineOnMapMove,
        onToggle: onRefineToggle
    }, /*#__PURE__*/ preact.h(Template.default, _object_spread_props._(_object_spread._({}, templateProps), {
        templateKey: "toggle",
        rootTagName: "span"
    }))) : /*#__PURE__*/ preact.h(GeoSearchButton.default, {
        className: cssClasses.redo,
        disabled: !hasMapMoveSinceLastRefine,
        onClick: onRefineClick
    }, /*#__PURE__*/ preact.h(Template.default, _object_spread_props._(_object_spread._({}, templateProps), {
        templateKey: "redo",
        rootTagName: "span"
    })))), !enableRefineControl && !isRefineOnMapMove && /*#__PURE__*/ preact.h("div", {
        className: cssClasses.control
    }, /*#__PURE__*/ preact.h(GeoSearchButton.default, {
        className: instantsearchUiComponents.cx(cssClasses.redo, !hasMapMoveSinceLastRefine && cssClasses.disabledRedo),
        disabled: !hasMapMoveSinceLastRefine,
        onClick: onRefineClick
    }, /*#__PURE__*/ preact.h(Template.default, _object_spread_props._(_object_spread._({}, templateProps), {
        templateKey: "redo",
        rootTagName: "span"
    })))), enableClearMapRefinement && isRefinedWithMap && /*#__PURE__*/ preact.h(GeoSearchButton.default, {
        className: cssClasses.reset,
        onClick: onClearClick
    }, /*#__PURE__*/ preact.h(Template.default, _object_spread_props._(_object_spread._({}, templateProps), {
        templateKey: "reset",
        rootTagName: "span"
    })))));
};

exports.default = GeoSearchControls;
