// instantsearch.js@4.90.0/cjs/widgets/menu-select/menu-select.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _object_spread = require('@swc/helpers/_/_object_spread');
var _object_spread_props = require('@swc/helpers/_/_object_spread_props');
var instantsearchUiComponents = require('instantsearch-ui-components');
var preact = require('preact');
var MenuSelect = require('../../components/MenuSelect/MenuSelect.js');
var connectMenu = require('../../connectors/menu/connectMenu.js');
var suit$1 = require('../../lib/suit.js');
var defaultTemplates = require('./defaultTemplates.js');
var documentation = require('../../lib/utils/documentation.js');
var getContainerNode = require('../../lib/utils/getContainerNode.js');
var prepareTemplateProps = require('../../lib/templating/prepareTemplateProps.js');

var withUsage = documentation.createDocumentationMessageGenerator({
    name: 'menu-select'
});
var suit = suit$1.component('MenuSelect');
var renderer = function renderer(param) {
    var containerNode = param.containerNode, cssClasses = param.cssClasses, renderState = param.renderState, templates = param.templates;
    return function(param, isFirstRendering) {
        var refine = param.refine, items = param.items, instantSearchInstance = param.instantSearchInstance;
        if (isFirstRendering) {
            renderState.templateProps = prepareTemplateProps.prepareTemplateProps({
                defaultTemplates: defaultTemplates.default,
                templatesConfig: instantSearchInstance.templatesConfig,
                templates: templates
            });
            return;
        }
        preact.render(/*#__PURE__*/ preact.h(MenuSelect.default, {
            cssClasses: cssClasses,
            items: items,
            refine: refine,
            templateProps: renderState.templateProps
        }), containerNode);
    };
};
var menuSelect = function menuSelect(widgetParams) {
    var _ref = widgetParams || {}, container = _ref.container, attribute = _ref.attribute, _ref_sortBy = _ref.sortBy, sortBy = _ref_sortBy === void 0 ? [
        'name:asc'
    ] : _ref_sortBy, _ref_limit = _ref.limit, limit = _ref_limit === void 0 ? 10 : _ref_limit, tmp = _ref.cssClasses, userCssClasses = tmp === void 0 ? {} : tmp, _ref_templates = _ref.templates, templates = _ref_templates === void 0 ? {} : _ref_templates, transformItems = _ref.transformItems;
    if (!container) {
        throw new Error(withUsage('The `container` option is required.'));
    }
    var containerNode = getContainerNode.getContainerNode(container);
    var cssClasses = {
        root: instantsearchUiComponents.cx(suit(), userCssClasses.root),
        noRefinementRoot: instantsearchUiComponents.cx(suit({
            modifierName: 'noRefinement'
        }), userCssClasses.noRefinementRoot),
        select: instantsearchUiComponents.cx(suit({
            descendantName: 'select'
        }), userCssClasses.select),
        option: instantsearchUiComponents.cx(suit({
            descendantName: 'option'
        }), userCssClasses.option)
    };
    var specializedRenderer = renderer({
        containerNode: containerNode,
        cssClasses: cssClasses,
        renderState: {},
        templates: templates
    });
    var makeWidget = connectMenu.default(specializedRenderer, function() {
        return preact.render(null, containerNode);
    });
    return _object_spread_props._(_object_spread._({}, makeWidget({
        attribute: attribute,
        limit: limit,
        sortBy: sortBy,
        transformItems: transformItems
    })), {
        $$widgetType: 'ais.menuSelect'
    });
};

exports.default = menuSelect;
