// instantsearch.js@4.90.0/cjs/components/Highlight/Highlight.js
'use strict';

var _object_spread = require('@swc/helpers/_/_object_spread');
var _object_without_properties = require('@swc/helpers/_/_object_without_properties');
var instantsearchUiComponents = require('instantsearch-ui-components');
var preact = require('preact');
var InternalHighlight = require('../InternalHighlight/InternalHighlight.js');

function Highlight(_0) {
    var _0_classNames = _0.classNames, classNames = _0_classNames === void 0 ? {} : _0_classNames, props = _object_without_properties._(_0, [
        "classNames"
    ]);
    return /*#__PURE__*/ preact.h(InternalHighlight.InternalHighlight, _object_spread._({
        classNames: {
            root: instantsearchUiComponents.cx('ais-Highlight', classNames.root),
            highlighted: instantsearchUiComponents.cx('ais-Highlight-highlighted', classNames.highlighted),
            nonHighlighted: instantsearchUiComponents.cx('ais-Highlight-nonHighlighted', classNames.nonHighlighted),
            separator: instantsearchUiComponents.cx('ais-Highlight-separator', classNames.separator)
        }
    }, props));
}

exports.Highlight = Highlight;
